---
title: "JvRAD"
author: "Raymond Otoo"
date: "2024-04-26"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/Users/raymondotoo/Library/CloudStorage/Box-Box/Lab Research Studies/Raymond/JvRad_Batch_1/Proteomics data/AllenAR_092723/protein_analysis/02_DA_results/protein_combined_results.csv")
```

## R Markdown

#Ref
#https://microsud.github.io/Bacterial-Proteomics-in-R/tutorial_main_doc.html

```{r}
# Install BiocManager if not already installed
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# Use BiocManager to install Bioconductor packages
BiocManager::install("DEP")
BiocManager::install("dplyr")
BiocManager::install("proteus")
BiocManager::install("SummarizedExperiment")

```


```{r}
library("DEP") # for Proteomics analysis
library("dplyr") # for data manipulation
library("pheatmap")
library("SummarizedExperiment")
```

## 01_Data Import and Cleaning:

# - Reads data from a CSV file (combined_results.csv) containing protein information and expression values.
# - Cleans column names and extracts unique protein IDs.
```{r}
jvrad <- read.csv("/Users/raymondotoo/Library/CloudStorage/Box-Box/Lab Research Studies/Raymond/JvRad_Batch_1/Proteomics data/AllenAR_092723/protein_analysis/02_DA_results/protein_combined_results.csv")

colnames(jvrad)

#appropriate cleaning for the gene_symbol column but maintain it for consistency the uniprot_id column which doesnt need this cleaning as this doesnt change anything
jvrad$UniProtIds <- sub(".*=", "", jvrad$uniprot_id)
jvrad$UniProtIds <- sub("\"", "", jvrad$uniprot_id)
jvrad$UniProtIds <- sub("\"$", "", jvrad$uniprot_id)

head(jvrad)
```

```{r}
data_unique <- make_unique(jvrad, "uniprot_id", "uniprot_id", delim =";")
data_unique
```

```{r}
jvrad$name %>% duplicated() %>% any()
```

```{r}
jvrad
```

```{r}
dim(jvrad)
```


## 02_Data Wrangling and Experimental Design:

# - Creates a data frame (experimental_design) containing sample labels, conditions, and replicates.
# - Defines expression columns based on sample names.
```{r}
label <- c("s200", "s201", "s202", "s203", "s204", 
           "s400", "s401", "s402", "s503", "s504")

condition <- c("ir", "ir", "ir", "ir", "ir", 
               "sham", "sham", "sham", "sham", "sham")


replicate <- rep(1:5,2)

label <- as.character(label)
condition <- as.character(condition)


#create the dataframe
experimental_design <- data.frame(label=label, condition=condition, replicate = as.double(replicate))
experimental_design

typeof(experimental_design)
```

```{r}
# Use grep to find the column numbers (indices) that match the specified patterns
#selects columns that starts with s2,s4 or s5. this captures all the samples
exp_columns <- grep("^s2|^s4|^s5", colnames(data_unique))

#used with experimental design
data_se<- make_se(data_unique, exp_columns, experimental_design)
```


```{r}
exp_columns
```

```{r}
data_se
```

## 03_Data Exploration:

# - Generates a mean-standard deviation plot (meanSdPlot) to visualize data distribution across samples.
# - Analyzes the number of proteins quantified in each sample with plot_frequency before and after filtering missing values (filter_missval).
# - Explores protein coverage across samples with plot_coverage.
# - Creates histograms to visualize the distribution of expression values.
```{r}
meanSdPlot(data_se)
```

```{r}
plot_frequency(data_se) # first check the number of proteins not quantified in any of the samples
```

```{r}
data_filt <- filter_missval(data_se, thr = 0)
plot_frequency(data_filt)
```
```{r}
plot_numbers(data_filt)
```
## Protein coverage in all samples.
```{r}
library("ggplot2")
p <- plot_coverage(data_filt) + scale_fill_brewer(palette = "Paired")
p <- p+ geom_text(aes(label=Freq), vjust=1.6, color="white")
p
```


```{r}
hist(2^assay(data_se)[,"ir_1"],n=1000) # on orginal scale using onesample
hist(assay(data_se)[,"ir_1"],n=1000) # log2 scale
plot(density(assay(data_se)[,"ir_1"],na.rm=TRUE))
```


## 04_Normalization:

# - Applies VSN normalization (normalize_vsn) for variance stabilization.
# - Optional: Includes code for median (normalize_median) and quantile normalization (normalizeQuantiles) but these are not applied by default.
```{r}
#Apply VSN normalisation
data_norm <- normalize_vsn(data_filt) # VSN is performed on original scale

#Apply median normalisation
data_norm_med <- data_filt
#assay(data_norm_med)<-proteus::normalizeMedian(assay(data_filt))

#Apply quantile normalisation
data_norm_quant <- data_filt
assay(data_norm_quant)<-limma::normalizeQuantiles(assay(data_filt))
```

```{r}
mds_plot <- meanSdPlot(data_norm) 
```

## Comparison of Raw and Normalized Data:
# - Generates a plot (plot_normalization) to compare the distribution of expression values before and after normalization.
```{r}
# specify colors
sub_cols <- c(ir= "#e63946", sham="blue")
p.norm <- plot_normalization(data_filt, data_norm)
p.norm + scale_fill_manual(values = sub_cols)

```


```{r}
plot_missval(data_filt)
plot_detect(data_filt) # make df such that average of a gene in all conditions in one column and 
```


## 05_Missing Value Imputation:
#- Analyzes missing values with plot_missval and identifies undetected proteins with plot_detect.
#- Imputes missing values using the minimum probability method (impute).
```{r}
data_imp <- impute(data_filt, fun = "MinProb", q = 0.01)
```


## 06_Visualization and Quality Control:
#- Compares the distribution of expression intensities before and after imputation with plot_imputation.
## # Plot intensity distributions before and after imputation
```{r}
plot_imputation(data_norm, data_imp)
```

## 07_Performs Principal Component Analysis (PCA) with plot_pca to visualize sample clustering based on protein expression.
### Use all detected proteins and do a PCA for visualizing if there is clustering of replicates and conditions.
```{r}
library("ggplot2")

sub_cols <- c(ir= "#e63946", sham="blue")

p.pca <- plot_pca(data_imp, x = 1, y = 2, 
                  n = nrow(data_imp@assays), # use all detected proteins
                  point_size = 5,
                  indicate = "condition",
                  label=F) + 
  ggtitle("JvRad") + 
  scale_color_manual("Treatments", values = sub_cols)
p.pca
```


## 08_Differential Enrichment Analysis:
#- Conducts differential analysis (test_diff) using linear models and empirical Bayes statistics and fdrtools.
#- Identifies significantly differentially expressed (DE) proteins based on a fold change threshold and adjusted p-value (add_rejections).
```{r}
data_diff<- test_diff(data_imp,type = "all")
```

## Identify/Mark significant proteins
```{r}
dep <- add_rejections(data_diff, alpha = 0.05, lfc = log2(1.5))
dep
```

##Plot top proteins
```{r}
#make a volcano plot of the selected contrast
plot_volcano(dep, contrast = "ir_vs_sham", label_size = 2, add_names= TRUE)

#plot a number of selcted proteins in on plot
plot_single(dep, proteins = c("A0A087WRT9","A0A0A6YVU8", "S4R1M9", "V9GX34"), type = "centered")

#plot only one protein
plot_single(dep, proteins = "V9GX34", type = "centered")
```
```{r}
cor_matrix <- plot_cor(data_imp, 
                       significant = F, 
                       lower = 0, 
                       upper = 1, 
                       pal = "GnBu",
                       indicate = c("condition", "replicate"), 
                       plot = F)

cor_matrix
```


## 09_Correlation Analysis:
#- Calculates protein-protein correlations (plot_cor) and generates a heatmap with pheatmap for visualization.
```{r}
pheatmap(cor_matrix)
```

```{r}
dep <- add_rejections(data_diff, alpha = 0.05, lfc = 0.0)
```


## Heatmap of Expression Data:
# - Creates a heatmap with hierarchical clustering (plot_heatmap) to visualize protein expression patterns across samples and replicates.
# Allows customization of clustering methods, color limits, and display options.

```{r}
png("jvrad_heatmap.png", width = 10, height = 10, units = 'in', res = 600)

plot_heatmap(dep, 
             type = "centered", 
             kmeans = T, 
             col_limit = 0.1, 
             show_row_names = T,
             column_title = "JvRad Heatmap",
             row_font_size=8,
             indicate = c("condition", "replicate"), 
             clustering_distance = "spearman",
             plot = TRUE)

dev.off()
```


```{r}
#plot comparisons 
p.heat <- plot_heatmap(dep, 
                       type = "contrast", 
                       kmeans = F, 
                       col_limit = 0.3, 
                       show_row_names = T,
                       indicate = c("condition", "replicate"), 
                       show_row_dend= T,
                       row_dend_side = "right", 
                       width = 0.5, 
                       gap = unit(1, "mm"))
```
```{r}
png("jvrad_volcanoPlot.png", width = 10, height = 6, units = 'in', res = 600)
remove_grids <- theme(panel.grid.major = element_blank(), 
                      panel.grid.minor = element_blank(),
                      panel.background = element_blank(), 
                      axis.line = element_line(colour = "black"))

ir_vs_sham <- plot_volcano(dep, 
                                      contrast = "ir_vs_sham", 
                                      label_size = 2, 
                                      add_names = TRUE) + remove_grids 

ir_vs_sham
dev.off()
```

### Plot DA proteins
```{r}
png("jvrad_key_protsSinglePlots.png", width = 10, height = 10, units = 'in', res = 600)
key_prots <- c("P07758","P0DW87","P28661", "P54754", "Q3UV17","Q60766","Q68FL4","Q6GQW0","Q6PDN3","Q6SKR2")

rep_cols <- c("1"="#ffc49b","2"="#e05780","3"="#adb6c4","4"="green","5"="blue" )

plot_single(dep, 
            proteins = key_prots, 
            type = "centered") +  
  scale_color_manual("Replicate", values = rep_cols)

dev.off()
```




## 10_Pathway Enrichment Analysis:
# - Extracts information about significantly DE proteins (get_results).
# - Performs KEGG pathway enrichment analysis with enrichKEGG and visualizes enriched pathways using barplots (barplot) and network plots (cnetplot).
# - Optionally performs additional enrichment analysis methods like MKEGG (enrichMKEGG) and visualizes the

```{r}
data_results <- get_results(dep)
#write.csv(data_results, "./results/L2_comparison_all.csv")
# Number of significant proteins
data_results %>% filter(significant) %>% nrow()
```

```{r}
data_results
```

```{r}
colnames(data_results)
```

```{r}
# Install BiocManager if not already installed
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# Use BiocManager to install Bioconductor packages
BiocManager::install("pathview")
BiocManager::install("enrichplot")
BiocManager::install("DOSE")
```

```{r}
library(clusterProfiler)
library(pathview)
library(enrichplot)
library(DOSE)
```


## Here,demonstrate this using DA proteins between saline 10Gy and SFN 10Gy treatments.
```{r}
jv_rad <- data_results[, c("name","ir_vs_sham_significant", "ir_vs_sham_ratio", "ir_vs_sham_p.adj")]

foldchanges.1 = jv_rad$ir_vs_sham_ratio
names(foldchanges.1) = jv_rad$name

# we use a threshold of -1.2 or + 1.2
gene <- names(foldchanges.1)[abs(foldchanges.1) > 0.00]
gene
```


## KEGG enrichment
```{r}
png("jvrad_keggEnrichPathways.png", width = 10, height = 6, units = 'in', res = 600)

#Identify KEGG pathways that are enriched.
ir_vs_sham_kegg <- enrichKEGG(gene,
                           organism = 'mouse',
                           keyType = "uniprot",
                           pvalueCutoff = 0.05)


#plot results
barplot(ir_vs_sham_kegg,
        drop=TRUE,
        showCategory = 10,
        title = "KEGG Enrichment Pathways",
        font.size = 12)

ir_vs_sham_kegg

ir_vs_sham_kegg[0:100]
dev.off()
```



```{r}
# Install BiocManager if not already installed
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# Use BiocManager to install Bioconductor packages
BiocManager::install("org.Mm.eg.db")
BiocManager::install("GO.db")
BiocManager::install("topGO")
BiocManager::install("KEGG.db")
BiocManager::install("ReactomePA")

```



```{r}
png("jvrad_cnetplot.png", width = 10, height = 10, units = 'in', res = 600)
enrichplot::cnetplot(ir_vs_sham_kegg,categorySize = "pvalue", 
                     foldChange = foldchanges.1, colorEdge= F)
dev.off()
```



```{r}
png("enrichMKEGG.png", width = 10, height = 6, units = 'in', res = 600)
mkk <- enrichMKEGG(gene = gene,
                   organism = 'mmu',
                   pvalueCutoff = 0.05,
                   minGSSize = 5,
                   keyType = "uniprot",
                   qvalueCutoff = 0.05)

barplot(mkk,
        drop=TRUE,
        showCategory = 10,
        title = "KEGG Enrichment Pathways",
        font.size = 12)

head(mkk)
mkk[0:100]
dev.off()
```


```{r}
library(pathview)
pathview(ir_vs_sham_kegg,
         pathway.id = "mmu05014",
         species = "mouse",
         plot.col.key= TRUE,
         res=600)

```

