---
title: "NASA_microbiome_Downstream"
author: "Raymond Otoo"
date: "2024-03-19"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
## https://www.yanh.org/2021/01/01/microbiome-r/



## 1. Files needed
### A. sample-metadata.tsv - a matrix containing the infomation for analysis
### B. rooted-tree.qza - a phylogenetic tree indicating the evolutional similarity of microbial features, potentially used when calculating phylogenetic diversity metrics
### C. table.qza -  a matrix containing the abundances of detected microbial features (OTUs, ASVs, microbial markers)
### D. taxonomy.qza - an array indicating the taxonomic assignment of features
```{r}
# Do not run
# Run this block once to install libraries and dependencies

#Install and load relevant packages: phyloseq, tidyverse, vegan, DESeq2 , ANCOMBC, ComplexHeatmap. The following code are solely written in R. Now you can open the Rstudio and repeat it.
p1 <- c("tidyverse", "vegan", "BiocManager")
p2 <- c("phyloseq", "ANCOMBC", "DESeq2", "ComplexHeatmap")
load_package <- function(p) {
  if (!requireNamespace(p, quietly = TRUE)) {
    ifelse(p %in% p1, 
           install.packages(p, repos = "http://cran.us.r-project.org/"), 
           BiocManager::install(p))
  }
  library(p, character.only = TRUE, quietly = TRUE)
}

invisible(lapply(c(p1,p2), load_package))
```

```{r}
setwd("E:/NASA_microbiome/downstream/")
```
```{r}
setwd("E:/NASA_microbiome/downstream/")

#the metadata
metadata_test <- read.table(file = "sample-metadata.tsv", 
                       sep = "\t", 
                       header = T, 
                       row.names = 1)

metadata_NASA <- read.table(file = "NASA-Metadata.tsv", 
                       sep = "\t", 
                       header = T, 
                       row.names = 1)

metadata_test

metadata_NASA

```


## 2. Build phyloseq project
```{r}
setwd("E:/NASA_microbiome/downstream/")

############### Original Files ###################
## table-no-ecmu-hits-TreatmentGroup.qza
## tree.qza
## taxonomy.qza
## NASA-Metadata.tsv
#################################################

############### After unzipping #################
## feature-table.tsv
## tree.nwk
## taxonomy.tsv
## NASA-Metadata.tsv
#################################################

#the OTU table
otu <- read.table(file = "feature-table.tsv", 
                  sep = "\t", 
                  header = T, 
                  row.names = 1, 
                  skip = 1, 
                  comment.char = "")

# the taxonomy table
taxonomy <- read.table(file = "taxonomy.tsv", 
                       sep = "\t", 
                       header = T ,
                       row.names = 1)



# clean the taxonomy, Greengenes format
tax <- taxonomy %>%
  select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "d__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}


#the metadata
metadata <- read.table(file = "NASA-Metadata.tsv", 
                       sep = "\t", 
                       header = T, 
                       row.names = 1)


#extract suitable info in desired format
OTU = otu_table(as.matrix(otu), taxa_are_rows = TRUE)
TAX = tax_table(as.matrix(tax.clean))
SAMPLE <- sample_data(metadata)
TREE = read_tree("tree.nwk")

tax
tax.clean

# merge the data
#ps <- phyloseq(OTU, TAX, SAMPLE,TREE)

```
```{r}
class(OTU)
```






```{r}
tax.clean
```


```{r}
set.seed(111) # keep result reproductive
ps.rarefied = rarefy_even_depth(ps, rngseed=1, sample.size=1103, replace=F)
```

```{r}
plot_richness(ps.rarefied, x="body.site", measures=c("Observed", "Shannon")) +
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))
```

