---
title: "JvRad_DEG"
author: "Raymond Otoo"
date: "2024-01-26"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/Users/raymondotoo/OneDrive - University of Arkansas for Medical Sciences/MultiOmics/Bulk RNA-seq/JvRad/")
```

## R Markdown
```{r}
R.version
```

```{r}
getwd()
```

```{r}
#setwd("/Users/raymondotoo/OneDrive - UA Little Rock/MultiOmics/Bulk RNA-seq/JvRad/")
```

# Import gene counts table
```{r}
# Install BiocManager if not already installed
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# Use BiocManager to install Bioconductor packages
#BiocManager::install("DESeq2")
#BiocManager::install("ggplot2")
#BiocManager::install("clusterProfiler")
#BiocManager::install("biomaRt")
#BiocManager::install("ReactomePA")
#BiocManager::install("DOSE")
#BiocManager::install("KEGG.db")
#BiocManager::install("org.Mm.eg.db")
#BiocManager::install("org.Hs.eg.db")
#BiocManager::install("pheatmap")
#BiocManager::install("genefilter")
#BiocManager::install("RColorBrewer")
#BiocManager::install("GO.db")
#BiocManager::install("topGO")
#BiocManager::install("dplyr")
#BiocManager::install("gage")
#BiocManager::install("ggsci")
```

```{r}
getwd()
```


```{r}

# - skip first row (general command info)
# - make row names the gene identifiers
countdata <- read.table("final_counts.txt", header = TRUE, skip = 1, row.names = 1)

# Extract the first 14 characters from the original column names
short_names <- substr(colnames(countdata), 1, 17)

# Remove leading ".." from the short names
short_names <- gsub("^\\.\\.", "", short_names)

# Rename columns using the cleaned short names
colnames(countdata) <- short_names

# Remove length/char columns
countdata <- countdata[, -c(1:5)]

# Make sure ID's are correct
head(countdata)

# Display the updated countdata
countdata
```
```{r}
colnames(countdata)
```

```{r}
# Import metadata file
# - make row names the matching sampleID's from the countdata
metadata <- read.delim("metadata_2.txt", row.names = 1)

# Add sampleID's to the mapping file
metadata$sampleid <- row.names(metadata)

# Reorder sampleID's to match featureCounts column order. 
metadata <- metadata[match(colnames(countdata), metadata$sampleid), ]

# Make sure ID's are correct
#head(metadata)
metadata
```
```{r}
library(DESeq2)
# - countData : count dataframe
# - colData : sample metadata in the dataframe with row names as sampleID's
# - design : The design of the comparisons to use. 
#            Use (~) before the name of the column variable to compare
ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData = metadata,
                                 design = ~Group)


# Find differential expressed genes
ddsMat <- DESeq(ddsMat)
```




##Get basic statisics about the number of significant genes
```{r}
# Get results from testing with FDR adjust pvalues
results <- results(ddsMat, pAdjustMethod = "fdr", alpha = 0.5)


# Generate summary of testing. 
summary(results)
```
# Check directionality of the log2 fold changes
```{r}
## Log2 fold change is set as (LoGlu / HiGlu)
## Postive fold changes = Increased in LoGlu
## Negative fold changes = Decreased in LoGlu
mcols(results, use.names = T)
```
#Step 8. Annotate gene symbols
```{r}
# Mouse genome database (Select the correct one)
library(org.Mm.eg.db) 

# Add gene full name
results$description <- mapIds(x = org.Mm.eg.db,
                              keys = row.names(results),
                              column = "GENENAME",
                              keytype = "SYMBOL",
                              multiVals = "first")

# Add gene symbol
results$symbol <- row.names(results)

# Add ENTREZ ID
results$entrez <- mapIds(x = org.Mm.eg.db,
                         keys = row.names(results),
                         column = "ENTREZID",
                         keytype = "SYMBOL",
                         multiVals = "first")

# Add ENSEMBL
results$ensembl <- mapIds(x = org.Mm.eg.db,
                          keys = row.names(results),
                          column = "ENSEMBL",
                          keytype = "SYMBOL",
                          multiVals = "first")

# Subset for only significant genes (q < 0.05)
results_sig <- subset(results, padj < 0.5)
head(results_sig)
```
```{r}
# Write normalized gene counts to a .txt file
write.table(x = as.data.frame(counts(ddsMat), normalized = T), 
            file = 'normalized_counts.txt', 
            sep = '\t', 
            quote = F,
            col.names = NA)

# Write significant normalized gene counts to a .txt file
write.table(x = counts(ddsMat[row.names(results_sig)], normalized = T), 
            file = 'normalized_counts_significant.txt', 
            sep = '\t', 
            quote = F, 
            col.names = NA)

# Write the annotated results table to a .txt file
write.table(x = as.data.frame(results), 
            file = "results_gene_annotated.txt", 
            sep = '\t', 
            quote = F,
            col.names = NA)

# Write significant annotated results table to a .txt file
write.table(x = as.data.frame(results_sig), 
            file = "results_gene_annotated_significant.txt", 
            sep = '\t', 
            quote = F,
            col.names = NA)
```

#Principal Component Analysis (PCA)
```{r}
# Load necessary libraries
library("ggplot2")
library("DESeq2")

# Convert all samples to rlog
ddsMat_rlog <- rlog(ddsMat, blind = FALSE)

# Plot PCA by column variable and save as a PNG file
png(filename = "PCA_plot.png", width = 10, height = 10, units = "in", res = 600)

# Extract the PCA data
pcaData <- plotPCA(ddsMat_rlog, intgroup = "Group", ntop = 500, returnData = TRUE)

# Create the PCA plot
ggplot(pcaData, aes(x = PC1, y = PC2, color = Group)) +
  geom_point(size = 5) +  # Increase point size
  scale_y_continuous(limits = c(-10, 20)) +  # Adjust y-axis limits if necessary
  theme_bw() +  # Use a clean theme
  ggtitle(label = "Principal Component Analysis (PCA)",
          subtitle = "Top 500 most variable genes")

dev.off()

```



# 9b. Heatmap
```{r}
library("pheatmap")
library("RColorBrewer")

png(filename = "Heatmap_forJvRad.png",width = 10, height = 10, units = "in", res = 600)


# Convert all samples to rlog
ddsMat_rlog <- rlog(ddsMat, blind = FALSE)

# Gather 30 significant genes and make matrix
mat <- assay(ddsMat_rlog[row.names(results_sig)])[1:40, ]

# Choose which column variables you want to annotate the columns by.
annotation_col = data.frame(
  Group = factor(colData(ddsMat_rlog)$Group), 
  Replicate = factor(colData(ddsMat_rlog)$Replicate),
  row.names = colData(ddsMat_rlog)$sampleid
)

# Specify colors you want to annotate the columns by.
ann_colors = list(
  Group = c(sham = "lightblue", ir = "purple"),
  Replicate = c(Rep1 = "red", Rep2 = "green",Rep3 = "blue",Rep4 = "forestgreen",Rep5 = "black")
)

# Make Heatmap with pheatmap function.
## See more in documentation for customization
pheatmap(mat = mat, 
         color = colorRampPalette(brewer.pal(9, "YlOrBr"))(255), 
         scale = "row", # Scale genes to Z-score (how many standard deviations)
         annotation_col = annotation_col, # Add multiple annotations to the samples
         annotation_colors = ann_colors,# Change the default colors of the annotations
         fontsize = 6.5, # Make fonts smaller
         cellwidth = 55, # Make the cells wider
         show_colnames = F,
          filename = "Heatmap_forJvRad.png"
        )
```
# 9c. Volcano Plot
```{r}
library("dplyr")

# Gather Log-fold change and FDR-corrected pvalues from DESeq2 results
## - Change pvalues to -log10 (1.3 = 0.05)
data <- data.frame(gene = row.names(results),
                   pval = -log10(results$padj), 
                   lfc = results$log2FoldChange)

# Remove any rows that have NA as an entry
data <- na.omit(data)

# Color the points which are up or down
## If fold-change > 0 and pvalue > 1.3 (Increased significant)
## If fold-change < 0 and pvalue > 1.3 (Decreased significant)
data <- mutate(data, color = case_when(data$lfc > 0 & data$pval > 1.3 ~ "Increased",
                                       data$lfc < 0 & data$pval > 1.3 ~ "Decreased",
                                       data$pval < 1.3 ~ "nonsignificant"))

# Make a basic ggplot2 object with x-y values
vol <- ggplot(data, aes(x = lfc, y = pval, color = color))

# Add ggplot2 layers
vol +   
  ggtitle(label = "Volcano Plot", subtitle = "Colored by fold-change direction") +
  geom_point(size = 2.5, alpha = 0.8, na.rm = T) +
  scale_color_manual(name = "Directionality",
                     values = c(Increased = "#008B00", Decreased = "#CD4F39", nonsignificant = "darkgray")) +
  theme_bw(base_size = 14) + # change overall theme
  theme(legend.position = "right") + # change the legend
  xlab(expression(log[2]("ir" / "sham"))) + # Change X-Axis label
  ylab(expression(-log[10]("adjusted p-value"))) + # Change Y-Axis label
  geom_hline(yintercept = 1.3, colour = "darkgrey") + # Add p-adj value cutoff line
  scale_y_continuous(trans = "log1p") # Scale yaxis due to large p-values
```
```{r}
plotMA(results, ylim = c(-5, 5))
```
```{r}
plotDispEsts(ddsMat)
```
#9f. Single gene plot
```{r}
# Convert all samples to rlog
ddsMat_rlog <- rlog(ddsMat, blind = FALSE)

# Get gene with highest expression
top_gene <- rownames(results)[which.min(results$log2FoldChange)]

# Plot single gene
plotCounts(dds = ddsMat, 
           gene = top_gene, 
           intgroup = "Group", 
           normalized = T, 
           transform = T)
```
# Step 10. Finding Pathways from Differential Expressed Genes
```{r}
# Remove any genes that do not have any entrez identifiers
results_sig_entrez <- subset(results_sig, is.na(entrez) == FALSE)

# Create a matrix of gene log2 fold changes
gene_matrix <- results_sig_entrez$log2FoldChange

# Add the entrezID's as names for each logFC entry
names(gene_matrix) <- results_sig_entrez$entrez

# View the format of the gene matrix
##- Names = ENTREZ ID
##- Values = Log2 Fold changes
head(gene_matrix)

```



# Over-Representation Analysis with ClusterProfiler

# https://learn.gencore.bio.nyu.edu/rna-seq-analysis/over-representation-analysis/

# Over-representation (or enrichment) analysis is a statistical method that determines whether genes from pre-defined sets (ex: those beloging to a specific GO term or KEGG pathway) are present more than would be expected (over-represented) in a subset of your data. In this case, the subset is your set of under or over expressed genes. For example, the fruit fly transcriptome has about 10,000 genes. If 260 genes are categorized as “axon guidance” (2.6% of all genes have category “axon guidance”), and in an experiment we find 1000 genes are differentially expressed and 200 of those genes are in the category “axon guidance” (20% of DE genes have category “axon guidance”), is that significant?

# This Section of the Notebook describes the implementation of over-representation analysis using the clusterProfiler package

# 10b. Enrich genes using the KEGG database
```{r}
library("clusterProfiler")
#library

kegg_enrich <- enrichKEGG(gene = names(gene_matrix),
                          organism = 'mouse',
                          pvalueCutoff = 0.05, 
                          qvalueCutoff = 0.10)

# Plot results
barplot(kegg_enrich, 
        drop = TRUE, 
        showCategory = 10, 
        title = "KEGG Enrichment Pathways",
        font.size = 8)

head(kegg_enrich)
kegg_enrich[0:100]
```


#10b. GO Analysis
#Ref: https://geneontology.org/docs/ontology-documentation/
#1. Ashburner et al. Gene ontology: tool for the unification of biology. Nat Genet. 2000 May;25(1):25-9. DOI: 10.1038/75556 
#2. The Gene Ontology Consortium. The Gene Ontology knowledgebase in 2023. Genetics. 2023 May 4;224(1):iyad031. DOI: 10.1093/genetics/iyad031

#An ontology is a formal representation of a body of knowledge within a given domain. Ontologies usually consist of a set of classes (or terms or concepts) with relations that operate between them. The Gene Ontology (GO) describes our knowledge of the biological domain with respect to three aspects:


```{r}
library("clusterProfiler")

go_enrich <- enrichGO(gene = names(gene_matrix),
                      OrgDb = 'org.Mm.eg.db', 
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.10)

# Plot results
barplot(go_enrich, 
        drop = TRUE, 
        showCategory = 10, 
        title = "GO Biological Pathways",
        font.size = 8)

head(go_enrich)
```

##B. Molecular-level activities performed by gene products. Molecular function terms describe activities that occur at the molecular level, such as “catalysis” or “transport”. GO molecular function terms represent activities rather than the entities (molecules or complexes) that perform the actions, and do not specify where, when, or in what context the action takes place. Molecular functions generally correspond to activities that can be performed by individual gene products (i.e. a protein or RNA), but some activities are performed by molecular complexes composed of multiple gene products. Examples of broad functional terms are catalytic activity and transporter activity; examples of narrower functional terms are adenylate cyclase activity or Toll-like receptor binding. To avoid confusion between gene product names and their molecular functions, GO molecular functions are often appended with the word “activity” (a protein kinase would have the GO molecular function protein kinase activity).
```{r}

# “BP”, “MF”, “CC”, “ALL”

library("clusterProfiler")

go_enrich <- enrichGO(gene = names(gene_matrix),
                      OrgDb = 'org.Mm.eg.db', 
                      readable = T,
                      ont = "MF",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.10)

# Plot results
barplot(go_enrich, 
        drop = TRUE, 
        showCategory = 10, 
        title = "Molecular Function",
        font.size = 8)
```


#c. A location, relative to cellular compartments and structures, occupied by a macromolecular machine. There are two ways in which the gene ontology describes locations of gene products: (1) the cellular anatomical entities, in which a gene product carries out a molecular function. Cellular anatomical entities includes cellular structures such as the plasma membrane and the cytoskeleton, as well as membrane-enclosed cellular compartments such as the mitochondrion, and (2) the stable macromolecular complexes of which they are parts, e.g., the clathrin complex.
```{r}
library("clusterProfiler")

go_enrich <- enrichGO(gene = names(gene_matrix),
                      OrgDb = 'org.Mm.eg.db', 
                      readable = T,
                      ont = "CC",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.10)

# Plot results
barplot(go_enrich, 
        drop = TRUE, 
        showCategory = 10, 
        title = "Cellular Component",
        font.size = 8)
```

# ----------------------------------------------------------------------------To do--------------------------------------------------------------------------------------------------------------
# Go Chord plot
# Ref:https://wencke.github.io/
```{r}
#run and install once
#install.packages('GOplot')
```

#Category Netplot

```{r}
#The cnetplot depicts the linkages of genes and biological concepts (e.g. GO terms or KEGG pathways) as a network (helpful to see which genes are involved in enriched pathways and genes that may belong to multiple annotation categories).

# categorySize can be either 'pvalue' or 'geneNum'
library("clusterProfiler")

png(filename = "cnet_plot.png",width = 10, height = 10, units = "in", res = 600)

cnetplot(go_enrich, categorySize="pvalue", foldChange=gene_matrix)

```

```{r}
## Install and run once
# Install BiocManager if not already installed
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# Use BiocManager to install Bioconductor packages
#BiocManager::install("enrichplot")
```

```{r}
library(enrichplot)
goplot(go_enrich, showCategory = 10)

```


# Step 11. Plotting KEGG Pathways
```{r}

# Install BiocManager if not already installed
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# Use BiocManager to install Bioconductor packages
#BiocManager::install("pathview")

```



#Ref: 
```{r}

#setwd("C:/Users/3073312/OneDrive - University of Arkansas for Medical Sciences/MultiOmics/Omics Data Integration/DEG_outputs")
# Load pathview
library(pathview)

# Plot specific KEGG pathways (with fold change) 
## pathway.id : KEGG pathway identifier
pathview(gene.data = gene_matrix, 
         pathway.id = "mmu04390", 
         species = "mouse")
```


