{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a5722c32-f6e5-4598-a4c4-dcb0f4b5422c",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Setting up environment"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d57b63d2-24c5-44a8-b87f-202f3913a5fc",
   "metadata": {},
   "source": [
    "## A. Installating Miniconda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6653acd6-c26b-4c29-a191-155be122cb2b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "%cd /mnt/e/new_workflow/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "778a543c-887b-4017-917b-2609ddc825d5",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Download the Miniconda3 installer to your home directory (Only for LINUX or Cluster)\n",
    "# !wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4c4465ab-9062-44f1-80db-a7a106012e1f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Run the miniconda installation\n",
    "# !bash miniconda.sh -b -f -p ~/miniconda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a7b52321-ae96-401a-828c-6854537e4a3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add miniconda to the system path\n",
    "# !echo 'PATH=\"$HOME/miniconda/bin:$PATH' >> ~/.bash_profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8d275bb3-5be8-471b-9020-4f02d90b39dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source system file to activate miniconda\n",
    "# !source ~/.bash_profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9bbbc548-5721-4374-9aad-31303df32dfc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # Add bioinformatic channels for downloading required packages\n",
    "# !conda config --add channels conda-forge\n",
    "# !conda config --add channels defaults\n",
    "# !conda config --add channels r\n",
    "# !conda config --add channels bioconda"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2af34f53-debc-4b6e-91d5-bac218334cae",
   "metadata": {},
   "source": [
    "## B. Setting Up the Folder Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c9af44f4-990b-4d2e-8aeb-5914fb9315aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ef88de3d-d321-4a6f-b8f9-5d293788e9e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "01_jvrad_rnaseq_pipeline.ipynb\tREADME.md     example  rerun\t     star_index\n",
      "02_RNAseq_pipeline.ipynb\tREADME_files  genome   results\n",
      "README.Rmd\t\t\tannotation    input    sortmerna_db\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c9baf783-7194-41e7-b67e-fa4dada15ade",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "01_jvrad_rnaseq_pipeline.ipynb\tREADME.md     example  rerun\t     star_index\n",
      "02_RNAseq_pipeline.ipynb\tREADME_files  genome   results\n",
      "README.Rmd\t\t\tannotation    input    sortmerna_db\n"
     ]
    }
   ],
   "source": [
    "!ls -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a0285eb3-94c4-4dff-8c80-a4849a63e45f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Install git (if needed)\n",
    "#!conda install -results/1_initial_qc/t wget --yes\n",
    "\n",
    "# Clone this repository with folder structure into the current working folder\n",
    "# !git clone https://github.com/twbattaglia/RNAseq-workflow new_workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0e1f3015-2f6c-4ce2-a732-59ded3348da9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "01_jvrad_rnaseq_pipeline.ipynb\tREADME.md     example  rerun\t     star_index\n",
      "02_RNAseq_pipeline.ipynb\tREADME_files  genome   results\n",
      "README.Rmd\t\t\tannotation    input    sortmerna_db\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ff5f7a8-3e45-4888-8218-2b0a5e126a69",
   "metadata": {},
   "source": [
    "### download reference genome and annotation for mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "27933605-6d99-441f-8862-b68450586fa6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Download genome fasta file into the genome/ folder\n",
    "\n",
    "## Download genome from ebi\n",
    "# !wget -P genome/ https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M34/GRCm39.genome.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a7387792-102c-428c-8186-b97ab1f05f19",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "910ca22b-420c-42d8-b455-8c19d050c6f6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Download annotation file into the annotation/ folder\n",
    "\n",
    "## Download annotation from ebi\n",
    "# !wget -P annotation/ https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M34/gencode.vM34.annotation.gtf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2505fc33-ad00-493b-9c72-a75b9dda56c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Decompress files for use with tools\n",
    "# !gunzip genome/GRCm39.genome.fa.gz\n",
    "# !gunzip annotation/gencode.vM34.annotation.gtf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e839f78e-7240-4d5f-ab92-9c8ae494daa8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Step 1. Analysing Sequence Quality with FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "80aabda2-a76a-426a-8ddc-e8cc73ee3c78",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# !conda install -c bioconda fastqc --yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cae4e471-1339-485d-baf9-8526376ef881",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d6ef8e7a-cc80-40a5-911e-4363e0117158",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#%cd /home/raotoo/new_workflow/new_workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0e531bf4-8ce0-4af6-a53f-dd5f8b19813c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !gunzip input/*.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f35f027e-70cb-4c11-b127-7dcbb35dc2cf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#-----------------------------------Test code for fastqc for one fastqc file -----------------------------------------\n",
    "# !fastqc \\\n",
    "# -o results/1_initial_qc/ \\\n",
    "# --noextract \\\n",
    "# input/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6bec0d51-483c-4240-9885-77a06f98cb9a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#-----------------------------------Test to run fastqc in batch -----------------------------------------\n",
    "#Run fastqc on all files\n",
    "# !for file in input/*.fastq; do fastqc -o results/1_initial_qc/ --noextract \"$file\"; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "81c65a19-5a47-43ce-bc85-d6369cdfbbd4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/raotoo/miniconda/bin/parallel\n"
     ]
    }
   ],
   "source": [
    "#check is parallel is installed\n",
    "# or Install with \n",
    "#!conda install -c conda-forge parallel --yes\n",
    "!which parallel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3493d72d-ca97-43f0-b11c-8736b0e58e7b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4e24a205-9788-44ae-8a1b-ae2be195f4e5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# %%bash\n",
    "\n",
    "# # Set the number of cores to use\n",
    "# num_cores=8  # Adjust based on your system's capabilities\n",
    "\n",
    "# # Run fastqc in parallel for all files in the input directory\n",
    "# find input/ -name '*.fastq' | parallel -j $num_cores \\\n",
    "# 'fastqc -o results/1_initial_qc/ --noextract {}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "60d69978-21fe-40a1-8238-2f649cb695f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow/results/1_initial_qc\n"
     ]
    }
   ],
   "source": [
    "%cd results/1_initial_qc/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8fe5eb17-a350-4f74-95fe-44aa71b78fec",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_fastqc.html\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_fastqc.zip\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_fastqc.html\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_fastqc.zip\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001_fastqc.html\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001_fastqc.zip\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001_fastqc.html\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001_fastqc.zip\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001_fastqc.html\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001_fastqc.zip\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001_fastqc.html\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001_fastqc.zip\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001_fastqc.html\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001_fastqc.zip\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001_fastqc.html\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001_fastqc.zip\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_fastqc.html\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_fastqc.zip\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_fastqc.html\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_fastqc.zip\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001_fastqc.html\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001_fastqc.zip\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001_fastqc.html\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001_fastqc.zip\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001_fastqc.html\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001_fastqc.zip\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001_fastqc.html\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001_fastqc.zip\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001_fastqc.html\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001_fastqc.zip\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001_fastqc.html\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001_fastqc.zip\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001_fastqc.html\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001_fastqc.zip\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001_fastqc.html\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001_fastqc.zip\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001_fastqc.html\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001_fastqc.zip\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001_fastqc.html\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001_fastqc.zip\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d8fd354f-1bcc-47f1-920d-254a6bec8e5c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# !conda install -c bioconda trim-galore --yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "16f83df0-5f7c-4f1f-aae8-3c73d8ce1db3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.10.4 (main, Mar 31 2022, 08:41:55) [GCC 7.5.0]'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "sys.version "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f8e6dcde-1929-4362-a042-c3eaa334cf2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!conda install -c bioconda trim-galore cutadapt=4.1 --yes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19e12d5d-b99b-4cf9-a19c-14f5308c0485",
   "metadata": {},
   "source": [
    "# Trim Galore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a46f6c39-4ecb-487c-94b4-ad6210eb1e7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#NB: Trim galore could not be install in this env via the cells and was installed in terminal. The path for the installed trim galore \n",
    "#was copied and added to the current directory\n",
    "import os\n",
    "os.environ['PATH'] = '/home/raotoo/miniconda/envs/trimmming/bin:' + os.environ['PATH']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fd335d8c-d8c8-4219-a1ac-de2bf22bed41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow/results/1_initial_qc\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3b69baad-2399-42a4-9f09-59e79da57b30",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "%cd /mnt/e/new_workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "563fb58b-f920-41fa-a21e-f429c1ff7c58",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " USAGE:\n",
      "\n",
      "trim_galore [options] <filename(s)>\n",
      "\n",
      "\n",
      "-h/--help               Print this help message and exits.\n",
      "\n",
      "-v/--version            Print the version information and exits.\n",
      "\n",
      "-q/--quality <INT>      Trim low-quality ends from reads in addition to adapter removal. For\n",
      "                        RRBS samples, quality trimming will be performed first, and adapter\n",
      "                        trimming is carried in a second round. Other files are quality and adapter\n",
      "                        trimmed in a single pass. The algorithm is the same as the one used by BWA\n",
      "                        (Subtract INT from all qualities; compute partial sums from all indices\n",
      "                        to the end of the sequence; cut sequence at the index at which the sum is\n",
      "                        minimal). Default Phred score: 20.\n",
      "\n",
      "--phred33               Instructs Cutadapt to use ASCII+33 quality scores as Phred scores\n",
      "                        (Sanger/Illumina 1.9+ encoding) for quality trimming. Default: ON.\n",
      "\n",
      "--phred64               Instructs Cutadapt to use ASCII+64 quality scores as Phred scores\n",
      "                        (Illumina 1.5 encoding) for quality trimming.\n",
      "\n",
      "--fastqc                Run FastQC in the default mode on the FastQ file once trimming is complete.\n",
      "\n",
      "--fastqc_args \"<ARGS>\"  Passes extra arguments to FastQC. If more than one argument is to be passed\n",
      "                        to FastQC they must be in the form \"arg1 arg2 etc.\". An example would be:\n",
      "                        --fastqc_args \"--nogroup --outdir /home/\". Passing extra arguments will\n",
      "                        automatically invoke FastQC, so --fastqc does not have to be specified\n",
      "                        separately.\n",
      "\n",
      "-a/--adapter <STRING>   Adapter sequence to be trimmed. If not specified explicitly, Trim Galore will\n",
      "                        try to auto-detect whether the Illumina universal, Nextera transposase or Illumina\n",
      "                        small RNA adapter sequence was used. Also see '--illumina', '--nextera' and\n",
      "                        '--small_rna'. If no adapter can be detected within the first 1 million sequences\n",
      "                        of the first file specified or if there is a tie between several adapter sequences,\n",
      "                        Trim Galore defaults to '--illumina' (as long as the Illumina adapter was one of the\n",
      "                        options, else '--nextera' is the default). A single base\n",
      "                        may also be given as e.g. -a A{10}, to be expanded to -a AAAAAAAAAA.\n",
      "\n",
      "                        At a special request, multiple adapters can also be specified like so:\n",
      "                        -a  \" AGCTCCCG -a TTTCATTATAT -a TTTATTCGGATTTAT\"\n",
      "                        -a2 \" AGCTAGCG -a TCTCTTATAT -a TTTCGGATTTAT\", or so:\n",
      "                        -a \"file:../multiple_adapters.fa\"\n",
      "                        -a2 \"file:../different_adapters.fa\"\n",
      "                        Potentially in conjucntion with the parameter \"-n 3\" to trim all adapters. Please note\n",
      "                        that this is NOT needed for standard trimming! \n",
      "                        More Information here: https://github.com/FelixKrueger/TrimGalore/issues/86\n",
      "\n",
      "-a2/--adapter2 <STRING> Optional adapter sequence to be trimmed off read 2 of paired-end files. This\n",
      "                        option requires '--paired' to be specified as well. If the libraries to be trimmed\n",
      "                        are smallRNA then a2 will be set to the Illumina small RNA 5' adapter automatically\n",
      "                        (GATCGTCGGACT). A single base may also be given as e.g. -a2 A{10}, to be expanded\n",
      "                        to -a2 AAAAAAAAAA.\n",
      "\n",
      "--illumina              Adapter sequence to be trimmed is the first 13bp of the Illumina universal adapter\n",
      "                        'AGATCGGAAGAGC' instead of the default auto-detection of adapter sequence.\n",
      "\n",
      "--stranded_illumina     Adapter sequence to be trimmed is the first 13bp of the Illumina stranded mRNA or Total\n",
      "                        RNA adapter 'ACTGTCTCTTATA' instead of the default auto-detection of adapter sequence. \n",
      "                        Note that this sequence resembles the Nextera sequence with an additional A from A-tailing.\n",
      "                        Please also see https://github.com/FelixKrueger/TrimGalore/issues/127 or \n",
      "                        https://support.illumina.com/bulletins/2020/06/trimming-t-overhang-options-for-the-illumina-rna-library-prep-wo.html\n",
      "                        for further information. This sequence is currently NOT included in the adapter auto-detection.\n",
      "\n",
      "--nextera               Adapter sequence to be trimmed is the first 12bp of the Nextera adapter\n",
      "                        'CTGTCTCTTATA' instead of the default auto-detection of adapter sequence.\n",
      "\n",
      "--small_rna             Adapter sequence to be trimmed is the first 12bp of the Illumina Small RNA 3' Adapter\n",
      "                        'TGGAATTCTCGG' instead of the default auto-detection of adapter sequence. Selecting\n",
      "                        to trim smallRNA adapters will also lower the --length value to 18bp. If the smallRNA\n",
      "                        libraries are paired-end then a2 will be set to the Illumina small RNA 5' adapter\n",
      "                        automatically (GATCGTCGGACT) unless -a 2 had been defined explicitly.\n",
      "\n",
      "--consider_already_trimmed <INT>     During adapter auto-detection, the limit set by <INT> allows the user to \n",
      "                        set a threshold up to which the file is considered already adapter-trimmed. If no adapter\n",
      "                        sequence exceeds this threshold, no additional adapter trimming will be performed (technically,\n",
      "                        the adapter is set to '-a X'). Quality trimming is still performed as usual.\n",
      "                        Default: NOT SELECTED (i.e. normal auto-detection precedence rules apply).                     \n",
      "\n",
      "--max_length <INT>      Discard reads that are longer than <INT> bp after trimming. This is only advised for\n",
      "                        smallRNA sequencing to remove non-small RNA sequences.\n",
      "\n",
      "\n",
      "--stringency <INT>      Overlap with adapter sequence required to trim a sequence. Defaults to a\n",
      "                        very stringent setting of 1, i.e. even a single bp of overlapping sequence\n",
      "                        will be trimmed off from the 3' end of any read.\n",
      "\n",
      "-e <ERROR RATE>         Maximum allowed error rate (no. of errors divided by the length of the matching\n",
      "                        region) (default: 0.1)\n",
      "\n",
      "--gzip                  Compress the output file with GZIP. If the input files are GZIP-compressed\n",
      "                        the output files will automatically be GZIP compressed as well. As of v0.2.8 the\n",
      "                        compression will take place on the fly.\n",
      "\n",
      "--dont_gzip             Output files won't be compressed with GZIP. This option overrides --gzip.\n",
      "\n",
      "--length <INT>          Discard reads that became shorter than length INT because of either\n",
      "                        quality or adapter trimming. A value of '0' effectively disables\n",
      "                        this behaviour. Default: 20 bp.\n",
      "\n",
      "                        For paired-end files, both reads of a read-pair need to be longer than\n",
      "                        <INT> bp to be printed out to validated paired-end files (see option --paired).\n",
      "                        If only one read became too short there is the possibility of keeping such\n",
      "                        unpaired single-end reads (see --retain_unpaired). Default pair-cutoff: 20 bp.\n",
      "\n",
      "--max_n COUNT           The total number of Ns a read may contain before it will be removed altogether.\n",
      "                        In a paired-end setting, either read exceeding this limit will result in the entire\n",
      "                        pair being removed from the trimmed output files. If COUNT is a number between 0 and 1,\n",
      "                        it is interpreted as a fraction of the read length.\n",
      "\n",
      "--trim-n                Removes Ns from either side of the read. This option does currently not work in RRBS mode.\n",
      "\n",
      "-o/--output_dir <DIR>   If specified all output will be written to this directory instead of the current\n",
      "                        directory. If the directory doesn't exist it will be created for you.\n",
      "\n",
      "--no_report_file        If specified no report file will be generated.\n",
      "\n",
      "--suppress_warn         If specified any output to STDOUT or STDERR will be suppressed.\n",
      "\n",
      "--clip_R1 <int>         Instructs Trim Galore to remove <int> bp from the 5' end of read 1 (or single-end\n",
      "                        reads). This may be useful if the qualities were very poor, or if there is some\n",
      "                        sort of unwanted bias at the 5' end. Default: OFF.\n",
      "\n",
      "--clip_R2 <int>         Instructs Trim Galore to remove <int> bp from the 5' end of read 2 (paired-end reads\n",
      "                        only). This may be useful if the qualities were very poor, or if there is some sort\n",
      "                        of unwanted bias at the 5' end. For paired-end BS-Seq, it is recommended to remove\n",
      "                        the first few bp because the end-repair reaction may introduce a bias towards low\n",
      "                        methylation. Please refer to the M-bias plot section in the Bismark User Guide for\n",
      "                        some examples. Default: OFF.\n",
      "\n",
      "--three_prime_clip_R1 <int>     Instructs Trim Galore to remove <int> bp from the 3' end of read 1 (or single-end\n",
      "                        reads) AFTER adapter/quality trimming has been performed. This may remove some unwanted\n",
      "                        bias from the 3' end that is not directly related to adapter sequence or basecall quality.\n",
      "                        Default: OFF.\n",
      "\n",
      "--three_prime_clip_R2 <int>     Instructs Trim Galore to remove <int> bp from the 3' end of read 2 AFTER\n",
      "                        adapter/quality trimming has been performed. This may remove some unwanted bias from\n",
      "                        the 3' end that is not directly related to adapter sequence or basecall quality.\n",
      "                        Default: OFF.\n",
      "\n",
      "--2colour/--nextseq INT This enables the option '--nextseq-trim=3'CUTOFF' within Cutadapt, which will set a quality\n",
      "                        cutoff (that is normally given with -q instead), but qualities of G bases are ignored.\n",
      "                        This trimming is in common for the NextSeq- and NovaSeq-platforms, where basecalls without\n",
      "                        any signal are called as high-quality G bases. This is mutually exlusive with '-q INT'.\n",
      "\n",
      "\n",
      "--path_to_cutadapt </path/to/cutadapt>     You may use this option to specify a path to the Cutadapt executable,\n",
      "                        e.g. /my/home/cutadapt-1.7.1/bin/cutadapt. Else it is assumed that Cutadapt is in\n",
      "                        the PATH.\n",
      "\n",
      "--basename <PREFERRED_NAME>\tUse PREFERRED_NAME as the basename for output files, instead of deriving the filenames from\n",
      "                        the input files. Single-end data would be called PREFERRED_NAME_trimmed.fq(.gz), or\n",
      "                        PREFERRED_NAME_val_1.fq(.gz) and PREFERRED_NAME_val_2.fq(.gz) for paired-end data. --basename\n",
      "                        only works when 1 file (single-end) or 2 files (paired-end) are specified, but not for longer lists.\n",
      "\n",
      "-j/--cores INT          Number of cores to be used for trimming [default: 1]. For Cutadapt to work with multiple cores, it\n",
      "                        requires Python 3 as well as parallel gzip (pigz) installed on the system. Trim Galore attempts to detect\n",
      "                        the version of Python used by calling Cutadapt. If Python 2 is detected, --cores is set to 1. If the Python\n",
      "                        version cannot be detected, Python 3 is assumed and we let Cutadapt handle potential issues itself.\n",
      "                        \n",
      "                        If pigz cannot be detected on your system, Trim Galore reverts to using gzip compression. Please note\n",
      "                        that gzip compression will slow down multi-core processes so much that it is hardly worthwhile, please \n",
      "                        see: https://github.com/FelixKrueger/TrimGalore/issues/16#issuecomment-458557103 for more info).\n",
      "\t\t\t\t\t\t\n",
      "                        Actual core usage: It should be mentioned that the actual number of cores used is a little convoluted.\n",
      "                        Assuming that Python 3 is used and pigz is installed, --cores 2 would use 2 cores to read the input\n",
      "                        (probably not at a high usage though), 2 cores to write to the output (at moderately high usage), and \n",
      "                        2 cores for Cutadapt itself + 2 additional cores for Cutadapt (not sure what they are used for) + 1 core\n",
      "                        for Trim Galore itself. So this can be up to 9 cores, even though most of them won't be used at 100% for\n",
      "                        most of the time. Paired-end processing uses twice as many cores for the validation (= writing out) step.\n",
      "                        --cores 4 would then be: 4 (read) + 4 (write) + 4 (Cutadapt) + 2 (extra Cutadapt) +\t1 (Trim Galore) = 15.\n",
      "\n",
      "                        It seems that --cores 4 could be a sweet spot, anything above has diminishing returns.\n",
      "\t\t\t\n",
      "\n",
      "\n",
      "SPECIFIC TRIMMING - without adapter/quality trimming\n",
      "\n",
      "--hardtrim5 <int>       Instead of performing adapter-/quality trimming, this option will simply hard-trim sequences\n",
      "                        to <int> bp at the 5'-end. Once hard-trimming of files is complete, Trim Galore will exit.\n",
      "                        Hard-trimmed output files will end in .<int>_5prime.fq(.gz). Here is an example:\n",
      "\n",
      "                        before:         CCTAAGGAAACAAGTACACTCCACACATGCATAAAGGAAATCAAATGTTATTTTTAAGAAAATGGAAAAT\n",
      "                        --hardtrim5 20: CCTAAGGAAACAAGTACACT\n",
      "\n",
      "--hardtrim3 <int>       Instead of performing adapter-/quality trimming, this option will simply hard-trim sequences\n",
      "                        to <int> bp at the 3'-end. Once hard-trimming of files is complete, Trim Galore will exit.\n",
      "                        Hard-trimmed output files will end in .<int>_3prime.fq(.gz). Here is an example:\n",
      "\n",
      "                        before:         CCTAAGGAAACAAGTACACTCCACACATGCATAAAGGAAATCAAATGTTATTTTTAAGAAAATGGAAAAT\n",
      "                        --hardtrim3 20:                                                   TTTTTAAGAAAATGGAAAAT\n",
      "\n",
      "--clock                 In this mode, reads are trimmed in a specific way that is currently used for the Mouse\n",
      "                        Epigenetic Clock (see here: Multi-tissue DNA methylation age predictor in mouse, Stubbs et al.,\n",
      "                        Genome Biology, 2017 18:68 https://doi.org/10.1186/s13059-017-1203-5). Following this, Trim Galore\n",
      "                        will exit.\n",
      "\n",
      "                        In it's current implementation, the dual-UMI RRBS reads come in the following format:\n",
      "\n",
      "                        Read 1  5' UUUUUUUU CAGTA FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF TACTG UUUUUUUU 3'\n",
      "                        Read 2  3' UUUUUUUU GTCAT FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF ATGAC UUUUUUUU 5'\n",
      "\n",
      "                        Where UUUUUUUU is a random 8-mer unique molecular identifier (UMI), CAGTA is a constant region,\n",
      "                        and FFFFFFF... is the actual RRBS-Fragment to be sequenced. The UMIs for Read 1 (R1) and\n",
      "                        Read 2 (R2), as well as the fixed sequences (F1 or F2), are written into the read ID and\n",
      "                        removed from the actual sequence. Here is an example:\n",
      "\n",
      "                        R1: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 1:N:0: CGATGTTT\n",
      "                            ATCTAGTTCAGTACGGTGTTTTCGAATTAGAAAAATATGTATAGAGGAAATAGATATAAAGGCGTATTCGTTATTG\n",
      "                        R2: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 3:N:0: CGATGTTT\n",
      "                            CAATTTTGCAGTACAAAAATAATACCTCCTCTATTTATCCAAAATCACAAAAAACCACCCACTTAACTTTCCCTAA\n",
      "\n",
      "                        R1: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 1:N:0: CGATGTTT:R1:ATCTAGTT:R2:CAATTTTG:F1:CAGT:F2:CAGT\n",
      "                                         CGGTGTTTTCGAATTAGAAAAATATGTATAGAGGAAATAGATATAAAGGCGTATTCGTTATTG\n",
      "                        R2: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 3:N:0: CGATGTTT:R1:ATCTAGTT:R2:CAATTTTG:F1:CAGT:F2:CAGT\n",
      "                                         CAAAAATAATACCTCCTCTATTTATCCAAAATCACAAAAAACCACCCACTTAACTTTCCCTAA\n",
      "\n",
      "                        Following clock trimming, the resulting files (.clock_UMI.R1.fq(.gz) and .clock_UMI.R2.fq(.gz))\n",
      "                        should be adapter- and quality trimmed with Trim Galore as usual. In addition, reads need to be trimmed\n",
      "                        by 15bp from their 3' end to get rid of potential UMI and fixed sequences. The command is:\n",
      "\n",
      "                        trim_galore --paired --three_prime_clip_R1 15 --three_prime_clip_R2 15 *.clock_UMI.R1.fq.gz *.clock_UMI.R2.fq.gz\n",
      "\n",
      "                        Following this, reads should be aligned with Bismark and deduplicated with UmiBam\n",
      "                        in '--dual_index' mode (see here: https://github.com/FelixKrueger/Umi-Grinder). UmiBam recognises\n",
      "                        the UMIs within this pattern: R1:(ATCTAGTT):R2:(CAATTTTG): as (UMI R1) and (UMI R2).\n",
      "\n",
      "--polyA                 This is a new, still experimental, trimming mode to identify and remove poly-A tails from sequences.\n",
      "                        When --polyA is selected, Trim Galore attempts to identify from the first supplied sample whether\n",
      "                        sequences contain more often a stretch of either 'AAAAAAAAAA' or 'TTTTTTTTTT'. This determines\n",
      "                        if Read 1 of a paired-end end file, or single-end files, are trimmed for PolyA or PolyT. In case of\n",
      "                        paired-end sequencing, Read2 is trimmed for the complementary base from the start of the reads. The\n",
      "                        auto-detection uses a default of A{20} for Read1 (3'-end trimming) and T{150} for Read2 (5'-end trimming).\n",
      "                        These values may be changed manually using the options -a and -a2.\n",
      "\n",
      "                        In addition to trimming the sequences, white spaces are replaced with _ and it records in the read ID\n",
      "                        how many bases were trimmed so it can later be used to identify PolyA trimmed sequences. This is currently done\n",
      "                        by writing tags to both the start (\"32:A:\") and end (\"_PolyA:32\") of the reads in the following example:\n",
      "\n",
      "                        @READ-ID:1:1102:22039:36996 1:N:0:CCTAATCC\n",
      "                        GCCTAAGGAAACAAGTACACTCCACACATGCATAAAGGAAATCAAATGTTATTTTTAAGAAAATGGAAAATAAAAACTTTATAAACACCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      "\n",
      "                        @32:A:READ-ID:1:1102:22039:36996_1:N:0:CCTAATCC_PolyA:32\n",
      "                        GCCTAAGGAAACAAGTACACTCCACACATGCATAAAGGAAATCAAATGTTATTTTTAAGAAAATGGAAAATAAAAACTTTATAAACACC\n",
      "\n",
      "                        PLEASE NOTE: The poly-A trimming mode expects that sequences were both adapter and quality trimmed\n",
      "                        before looking for Poly-A tails, and it is the user's responsibility to carry out an initial round of\n",
      "                        trimming. The following sequence:\n",
      " \n",
      "                        1) trim_galore file.fastq.gz\n",
      "                        2) trim_galore --polyA file_trimmed.fq.gz\n",
      "                        3) zcat file_trimmed_trimmed.fq.gz | grep -A 3 PolyA | grep -v ^-- > PolyA_trimmed.fastq\n",
      "\n",
      "                        Will 1) trim qualities and Illumina adapter contamination, 2) find and remove PolyA contamination.\n",
      "                        Finally, if desired, 3) will specifically find PolyA trimmed sequences to a specific FastQ file of your choice.\n",
      "\n",
      "--implicon              This is a special mode of operation for paired-end data, such as required for the IMPLICON method, where a UMI sequence\n",
      "                        is getting transferred from the start of Read 2 to the readID of both reads. Following this, Trim Galore will exit.\n",
      "\n",
      "                        In it's current implementation, the UMI carrying reads come in the following format:\n",
      "\n",
      "                        Read 1  5' FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF 3'\n",
      "                        Read 2  3' UUUUUUUUFFFFFFFFFFFFFFFFFFFFFFFFFFFF 5'\n",
      "\n",
      "                        Where UUUUUUUU is a random 8-mer unique molecular identifier (UMI) and FFFFFFF... is the actual fragment to be\n",
      "                        sequenced. The UMI of Read 2 (R2) is written into the read ID of both reads and removed from the actual sequence.\n",
      "                        Here is an example:\n",
      "\n",
      "                        R1: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 1:N:0: CGATGTTT\n",
      "                            ATCTAGTTCAGTACGGTGTTTTCGAATTAGAAAAATATGTATAGAGGAAATAGATATAAAGGCGTATTCGTTATTG\n",
      "                        R2: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 3:N:0: CGATGTTT\n",
      "                            CAATTTTGCAGTACAAAAATAATACCTCCTCTATTTATCCAAAATCACAAAAAACCACCCACTTAACTTTCCCTAA\n",
      "                        \n",
      "                        After --implicon trimming:\n",
      "                        R1: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 1:N:0: CGATGTTT:CAATTTTG\n",
      "                            ATCTAGTTCAGTACGGTGTTTTCGAATTAGAAAAATATGTATAGAGGAAATAGATATAAAGGCGTATTCGTTATTG\n",
      "                        R2: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 3:N:0: CGATGTTT:CAATTTTG\n",
      "                                    CAGTACAAAAATAATACCTCCTCTATTTATCCAAAATCACAAAAAACCACCCACTTAACTTTCCCTAA\n",
      "\n",
      "RRBS-specific options (MspI digested material):\n",
      "\n",
      "--rrbs                  Specifies that the input file was an MspI digested RRBS sample (recognition\n",
      "                        site: CCGG). Single-end or Read 1 sequences (paired-end) which were adapter-trimmed\n",
      "                        will have a further 2 bp removed from their 3' end. Sequences which were merely\n",
      "                        trimmed because of poor quality will not be shortened further. Read 2 of paired-end\n",
      "                        libraries will in addition have the first 2 bp removed from the 5' end (by setting\n",
      "                        '--clip_r2 2'). This is to avoid using artificial methylation calls from the filled-in\n",
      "                        cytosine positions close to the 3' MspI site in sequenced fragments.\n",
      "                        This option is not recommended for users of the Tecan Ovation RRBS Methyl-Seq with TrueMethyl\n",
      "                        oxBS 1-16 kit (see below).\n",
      "\n",
      "--non_directional       Selecting this option for non-directional RRBS libraries will screen\n",
      "                        quality-trimmed sequences for 'CAA' or 'CGA' at the start of the read\n",
      "                        and, if found, removes the first two basepairs. Like with the option\n",
      "                        '--rrbs' this avoids using cytosine positions that were filled-in\n",
      "                        during the end-repair step. '--non_directional' requires '--rrbs' to\n",
      "                        be specified as well. Note that this option does not set '--clip_r2 2' in\n",
      "                        paired-end mode.\n",
      "\n",
      "--keep                  Keep the quality trimmed intermediate file. Default: off, which means\n",
      "                        the temporary file is being deleted after adapter trimming. Only has\n",
      "                        an effect for RRBS samples since other FastQ files are not trimmed\n",
      "                        for poor qualities separately.\n",
      "\n",
      "\n",
      "Note for RRBS using the Tecan Ovation RRBS Methyl-Seq with TrueMethyl oxBS 1-16 kit:\n",
      "\n",
      "Owing to the fact that the Tecan Ovation RRBS kit attaches a varying number of nucleotides (0-3) after each MspI\n",
      "site Trim Galore should be run WITHOUT the option --rrbs. This trimming is accomplished in a subsequent\n",
      "diversity trimming step afterwards (see their manual).\n",
      "\n",
      "\n",
      "\n",
      "Note for RRBS using MseI:\n",
      "\n",
      "If your DNA material was digested with MseI (recognition motif: TTAA) instead of MspI it is NOT necessary\n",
      "to specify --rrbs or --non_directional since virtually all reads should start with the sequence\n",
      "'TAA', and this holds true for both directional and non-directional libraries. As the end-repair of 'TAA'\n",
      "restricted sites does not involve any cytosines it does not need to be treated especially. Instead, simply\n",
      "run Trim Galore! in the standard (i.e. non-RRBS) mode.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Paired-end specific options:\n",
      "\n",
      "--paired                This option performs length trimming of quality/adapter/RRBS trimmed reads for\n",
      "                        paired-end files. To pass the validation test, both sequences of a sequence pair\n",
      "                        are required to have a certain minimum length which is governed by the option\n",
      "                        --length (see above). If only one read passes this length threshold the\n",
      "                        other read can be rescued (see option --retain_unpaired). Using this option lets\n",
      "                        you discard too short read pairs without disturbing the sequence-by-sequence order\n",
      "                        of FastQ files which is required by many aligners.\n",
      "\n",
      "                        Trim Galore! expects paired-end files to be supplied in a pairwise fashion, e.g.\n",
      "                        file1_1.fq file1_2.fq SRR2_1.fq.gz SRR2_2.fq.gz ... .\n",
      "\n",
      "\n",
      "--retain_unpaired       If only one of the two paired-end reads became too short, the longer\n",
      "                        read will be written to either '.unpaired_1.fq' or '.unpaired_2.fq'\n",
      "                        output files. The length cutoff for unpaired single-end reads is\n",
      "                        governed by the parameters -r1/--length_1 and -r2/--length_2. Default: OFF.\n",
      "\n",
      "-r1/--length_1 <INT>    Unpaired single-end read length cutoff needed for read 1 to be written to\n",
      "                        '.unpaired_1.fq' output file. These reads may be mapped in single-end mode.\n",
      "                        Default: 35 bp.\n",
      "\n",
      "-r2/--length_2 <INT>    Unpaired single-end read length cutoff needed for read 2 to be written to\n",
      "                        '.unpaired_2.fq' output file. These reads may be mapped in single-end mode.\n",
      "                        Default: 35 bp.\n",
      "\n",
      "Last modified on 02 02 2023.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Run Trim Galore\n",
    "!trim_galore --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "de656748-f939-45c8-bcc3-3b26c01eddc4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)\n",
      "Copyright (C) 2020 Free Software Foundation, Inc.\n",
      "License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n",
      "\n",
      "This is free software; you are free to change and redistribute it.\n",
      "There is NO WARRANTY, to the extent permitted by law.\n"
     ]
    }
   ],
   "source": [
    "!bash --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6a62170e-e7e5-4daa-a2b6-24536534c746",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# %%bash\n",
    "\n",
    "# # Set the number of cores to use\n",
    "# num_cores=8  # You can adjust this based on your system's capabilities\n",
    "\n",
    "# # Run Trim Galore! in parallel for all files in the folder\n",
    "# find input/ -name '*.fastq' | parallel -j $num_cores \\\n",
    "# 'trim_galore \\\n",
    "# --quality 20 \\\n",
    "# --fastqc \\\n",
    "# --length 25 \\\n",
    "# --output_dir results/2_trimmed_output/ \\\n",
    "# {}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "89efc690-3529-4b82-9b86-a0549b67ac5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "01_jvrad_rnaseq_pipeline.ipynb\tREADME.md     example  rerun\t     star_index\n",
      "02_RNAseq_pipeline.ipynb\tREADME_files  genome   results\n",
      "README.Rmd\t\t\tannotation    input    sortmerna_db\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "57fb50e5-7597-4af7-909e-32227b2bc0be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "c6ad3b66-2112-457f-b162-a91f5b802728",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow/results/2_trimmed_output\n"
     ]
    }
   ],
   "source": [
    "%cd results/2_trimmed_output/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "78e2cb25-e007-4b02-a451-5c0501b521b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001_trimmed_fastqc.zip\n",
      "sortmerna_db\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33df7cc8-e755-445b-ba39-dab59cdac50e",
   "metadata": {},
   "source": [
    "## 3. Remove rRNA with SortMeRNA (Optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "73baef55-f54e-4916-91ff-880479c64950",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "%cd /mnt/e/new_workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "ed8b3b86-e78e-4f57-90a6-8c17df9810b8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# !conda install -c bioconda sortmerna --yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "b4023d62-744a-44e2-98dd-4d5eb79dc58a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Generating index\n",
    "# Download the sortmerna package (~2min) into sortmerna_db folder\n",
    "# !wget -P sortmerna_db https://github.com/biocore/sortmerna/archive/2.1b.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f651e208-02c9-4375-8232-347300beb9f0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)\n",
      "Copyright (C) 2020 Free Software Foundation, Inc.\n",
      "License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n",
      "\n",
      "This is free software; you are free to change and redistribute it.\n",
      "There is NO WARRANTY, to the extent permitted by law.\n"
     ]
    }
   ],
   "source": [
    "!bash --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "338c8ad5-58d6-4cbe-b7d8-86a502267056",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "53c748ff-d3ce-4b75-9539-a43467295924",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Channels:',\n",
       " ' - conda-forge',\n",
       " ' - bioconda',\n",
       " ' - r',\n",
       " ' - defaults',\n",
       " 'Platform: linux-64',\n",
       " 'Collecting package metadata (repodata.json): ...working... done',\n",
       " 'Solving environment: ...working... done',\n",
       " '',\n",
       " '# All requested packages already installed.',\n",
       " '']"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!conda install -c conda-forge unzip --yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "3abbdd51-a8fd-445c-9718-381e1e317cc8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Decompress folder \n",
    "# !unzip sortmerna_db/2.1b.zip -d sortmerna_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3faf89ff-1da0-4009-925e-a6575dcb40ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Move the database into the correct folder\n",
    "# !mv sortmerna_db/sortmerna-2.1b/rRNA_databases/ sortmerna_db/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "6a58c9c0-110f-4010-a4b5-c26ff543e5a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Remove unnecessary folders\n",
    "# !rm sortmerna_db/2.1b.zip\n",
    "# !rm -r sortmerna_db/sortmerna-2.1b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "4ea25cd7-c51a-4be6-b5e0-8e6d64c23eea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "d13dec6f-008c-402c-98ce-fa8fbc91cd58",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[process:1393] === Options processing starts ... ===\n",
      "Found value: sortmerna\n",
      "Found flag: -h\n",
      "[process:1483] Processing option: h with value: \n",
      "\n",
      "  Program:      SortMeRNA version 4.3.6\n",
      "  Copyright:    2016-2020 Clarity Genomics BVBA:\n",
      "                Turnhoutseweg 30, 2340 Beerse, Belgium\n",
      "                2014-2016 Knight Lab:\n",
      "                Department of Pediatrics, UCSD, La Jolla\n",
      "                2012-2014 Bonsai Bioinformatics Research Group:\n",
      "                LIFL, University Lille 1, CNRS UMR 8022, INRIA Nord-Europe\n",
      "  Disclaimer:   SortMeRNA comes with ABSOLUTELY NO WARRANTY; without even the\n",
      "                implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n",
      "                See the GNU Lesser General Public License for more details.\n",
      "  Contributors: Jenya Kopylova   jenya.kopylov@gmail.com\n",
      "                Laurent No      laurent.noe@lifl.fr\n",
      "                Pierre Pericard  pierre.pericard@lifl.fr\n",
      "                Daniel McDonald  wasade@gmail.com\n",
      "                Mikal Salson    mikael.salson@lifl.fr\n",
      "                Hlne Touzet    helene.touzet@lifl.fr\n",
      "                Rob Knight       robknight@ucsd.edu\n",
      "\n",
      "  Usage:   sortmerna -ref FILE [-ref FILE] -reads FWD_READS [-reads REV_READS] [OPTIONS]:\n",
      "  -------------------------------------------------------------------------------------------------------------\n",
      "  | option            type-format           description                                          default      |\n",
      "  -------------------------------------------------------------------------------------------------------------\n",
      "\n",
      "    [REQUIRED]\n",
      "    --ref             PATH        Required  Reference file (FASTA) absolute or relative path.\n",
      "\n",
      "       Use mutliple times, once per a reference file\n",
      "\n",
      "\n",
      "    --reads           PATH        Required  Raw reads file (FASTA/FASTQ/FASTA.GZ/FASTQ.GZ).\n",
      "\n",
      "       Use twice for files with paired reads.\n",
      "       The file extensions are Not important. The program automatically\n",
      "       recognizes the file format as flat/compressed, fasta/fastq\n",
      "\n",
      "\n",
      "\n",
      "    [COMMON]\n",
      "    --workdir         PATH        Optional  Workspace directory                         USRDIR/sortmerna/run/\n",
      "\n",
      "       Default structure: WORKDIR/\n",
      "                              idx/   (References index)\n",
      "                              kvdb/  (Key-value storage for alignments)\n",
      "                              out/   (processing output)\n",
      "                              readb/ (pre-processed reads/index)\n",
      "\n",
      "\n",
      "    --kvdb            PATH        Optional  Directory for Key-value database            WORKDIR/kvdb\n",
      "\n",
      "       KVDB is used for storing the alignment results.\n",
      "\n",
      "\n",
      "    --idx-dir         PATH        Optional  Directory for storing Reference index.      WORKDIR/idx\n",
      "\n",
      "\n",
      "    --readb           PATH        Optional  Storage for pre-processed reads             WORKDIR/readb/\n",
      "\n",
      "       Directory storing the split reads, or the random access index of compressed reads\n",
      "\n",
      "\n",
      "    --fastx           BOOL        Optional  Output aligned reads into FASTA/FASTQ file\n",
      "    --sam             BOOL        Optional  Output SAM alignment for aligned reads.\n",
      "\n",
      "\n",
      "    --SQ              BOOL        Optional  Add SQ tags to the SAM file\n",
      "\n",
      "\n",
      "    --blast           STR         Optional  output alignments in various Blast-like formats\n",
      "\n",
      "       Sample values: '0'                    - pairwise\n",
      "                      '1'                    - tabular (Blast - m 8 format)\n",
      "                      '1 cigar'              - tabular + column for CIGAR\n",
      "                      '1 cigar qcov'         - tabular + columns for CIGAR and query coverage\n",
      "                      '1 cigar qcov qstrand' - tabular + columns for CIGAR, query coverage,\n",
      "                                               and strand\n",
      "\n",
      "\n",
      "    --aligned         STR/BOOL    Optional  Aligned reads file prefix [dir/][pfx]       WORKDIR/out/aligned\n",
      "\n",
      "       Directory and file prefix for aligned output i.e. each\n",
      "       output file goes into the specified directory with the given prefix.\n",
      "       The appropriate extension: (fasta|fastq|blast|sam|etc) is automatically added.\n",
      "       Both 'dir' and 'pfx' are optional.\n",
      "       The 'dir' can be a relative or an absolute path.\n",
      "       If 'dir' is not specified, the output is created in the WORKDIR/out/\n",
      "       If 'pfx' is not specified, the prefix 'aligned' is used\n",
      "       Examples:\n",
      "       '-aligned $MYDIR/dir_1/dir_2/1' -> $MYDIR/dir_1/dir_2/1.fasta\n",
      "       '-aligned dir_1/apfx'           -> $PWD/dir_1/apfx.fasta\n",
      "       '-aligned dir_1/'               -> $PWD/aligned.fasta\n",
      "       '-aligned apfx'                 -> $PWD/apfx.fasta\n",
      "       '-aligned  (no argument)'       -> WORKDIR/out/aligned.fasta\n",
      "\n",
      "\n",
      "    --other           STR/BOOL    Optional  Non-aligned reads file prefix [dir/][pfx]   WORKDIR/out/other\n",
      "\n",
      "       Directory and file prefix for non-aligned output i.e. each\n",
      "       output file goes into the specified directory with the given prefix.\n",
      "       The appropriate extension: (fasta|fastq|blast|sam|etc) is automatically added.\n",
      "       Must be used with 'fastx'.\n",
      "       Both 'dir' and 'pfx' are optional.\n",
      "       The 'dir' can be a relative or an absolute path.\n",
      "       If 'dir' is not specified, the output is created in the WORKDIR/out/\n",
      "       If 'pfx' is not specified, the prefix 'other' is used\n",
      "       Examples:\n",
      "       '-other $MYDIR/dir_1/dir_2/1' -> $MYDIR/dir_1/dir_2/1.fasta\n",
      "       '-other dir_1/apfx'           -> $PWD/dir_1/apfx.fasta\n",
      "       '-other dir_1/'               -> $PWD/dir_1/other.fasta\n",
      "       '-other apfx'                 -> $PWD/apfx.fasta\n",
      "       '-other  (no argument)'       -> aligned_out/other.fasta\n",
      "                                        i.e. the same output directory\n",
      "                                        as used for aligned output\n",
      "\n",
      "\n",
      "    --num_alignments  INT         Optional  Positive integer (INT >=0).\n",
      "\n",
      "       If used with '-no-best' reports first INT alignments per read reaching\n",
      "       E-value threshold, which allows to lower the CPU time and memory use.\n",
      "       Otherwise outputs INT best alignments.\n",
      "       If INT = 0, all alignments are output\n",
      "\n",
      "\n",
      "    --no-best         BOOL        Optional  Disable best alignments search                          False\n",
      "\n",
      "       The 'best' alignment is the highest scoring alignment out of All alignments of a read,\n",
      "       and the read can potentially be aligned (reaching E-value threshold) to multiple reference\n",
      "       sequences.\n",
      "       By default the program searches for best alignments i.e. performs an exhaustive search\n",
      "       over all references. Using '-no-best' will make the program to search just\n",
      "       the first N alignments, where N is set using '-num_alignments' i.e. 1 by default.\n",
      "\n",
      "\n",
      "    --min_lis         INT         Optional  Search only alignments that have the LIS                2\n",
      "                                            of at least N seeds long\n",
      "\n",
      "       LIS stands for Longest Increasing Subsequence. It is computed using seeds, which\n",
      "       are k-mers common to the read and the reference sequence. Sorted sequences of such seeds\n",
      "       are used to filter the candidate references prior performing the Smith-Waterman alignment.\n",
      "\n",
      "\n",
      "    --print_all_reads BOOL        Optional  Output null alignment strings for non-aligned reads     False\n",
      "                                            to SAM and/or BLAST tabular files\n",
      "\n",
      "    --paired          BOOL        Optional  Flags paired reads                                      False\n",
      "\n",
      "        If a single reads file is provided, use this option to indicate\n",
      "        the file contains interleaved paired reads when neither\n",
      "        'paired_in' | 'paired_out' | 'out2' | 'sout' are specified.\n",
      "\n",
      "\n",
      "    --paired_in       BOOL        Optional  Flags the paired-end reads as Aligned,                  False\n",
      "                                            when either of them is Aligned.\n",
      "\n",
      "        With this option both reads are output into Aligned FASTA/Q file\n",
      "        Must be used with 'fastx'.\n",
      "        Mutually exclusive with 'paired_out'.\n",
      "\n",
      "\n",
      "    --paired_out      BOOL        Optional  Flags the paired-end reads as Non-aligned,              False\n",
      "                                            when either of them is non-aligned.\n",
      "\n",
      "        With this option both reads are output into Non-Aligned FASTA/Q file\n",
      "        Must be used with 'fastx'.\n",
      "        Mutually exclusive with 'paired_in'.\n",
      "\n",
      "\n",
      "    --out2            BOOL        Optional  Output paired reads into separate files.                False\n",
      "\n",
      "       Must be used with 'fastx'.\n",
      "       If a single reads file is provided, this options implies interleaved paired reads\n",
      "       When used with 'sout', four (4) output files for aligned reads will be generated:\n",
      "       'aligned-paired-fwd, aligned-paired-rev, aligned-singleton-fwd, aligned-singleton-rev'.\n",
      "       If 'other' option is also used, eight (8) output files will be generated.\n",
      "\n",
      "\n",
      "    --sout            BOOL        Optional  Separate paired and singleton aligned reads.            False\n",
      "\n",
      "       To be used with 'fastx'.\n",
      "       If a single reads file is provided, this options implies interleaved paired reads\n",
      "       Cannot be used with 'paired_in' | 'paired_out'\n",
      "\n",
      "\n",
      "    --zip-out         STR/BOOL    Optional  Controls the output compression                        '-1'\n",
      "\n",
      "       By default the report files are produced in the same format as the input i.e.\n",
      "       if the reads files are compressed (gz), the output is also compressed.\n",
      "       The default behaviour can be overriden by using '-zip-out'.\n",
      "       The possible values: '1/true/t/yes/y'\n",
      "                            '0/false/f/no/n'\n",
      "                            '-1' (the same format as input - default)\n",
      "       The values are Not case sensitive i.e. 'Yes, YES, yEs, Y, y' are all OK\n",
      "       Examples:\n",
      "       '-reads freads.gz -zip-out n' : generate flat output when the input is compressed\n",
      "       '-reads freads.flat -zip-out' : compress the output when the input files are flat\n",
      "\n",
      "\n",
      "    --match           INT         Optional  SW score (positive integer) for a match.                2\n",
      "\n",
      "    --mismatch        INT         Optional  SW penalty (negative integer) for a mismatch.          -3\n",
      "\n",
      "    --gap_open        INT         Optional  SW penalty (positive integer) for introducing a gap.    5\n",
      "\n",
      "    --gap_ext         INT         Optional  SW penalty (positive integer) for extending a gap.      2\n",
      "\n",
      "    -e                DOUBLE      Optional  E-value threshold.                                      1\n",
      "\n",
      "       Defines the 'statistical significance' of a local alignment.\n",
      "       Exponentially correllates with the Minimal Alignment score.\n",
      "       Higher E-values (100, 1000, ...) cause More reads to Pass the alignment threshold\n",
      "\n",
      "\n",
      "    -F                BOOL        Optional  Search only the forward strand.                         False\n",
      "\n",
      "    -N                BOOL        Optional  SW penalty for ambiguous letters (N's) scored\n",
      "                                            as --mismatch\n",
      "\n",
      "    -R                BOOL        Optional  Search only the reverse-complementary strand.           False\n",
      "\n",
      "\n",
      "    [OTU_PICKING]\n",
      "    --id              INT         Optional  %%id similarity threshold (the alignment                0.97\n",
      "                                            must still pass the E-value threshold).\n",
      "\n",
      "    --coverage        INT         Optional  %%query coverage threshold (the alignment must          0.97\n",
      "                                            still pass the E-value threshold)\n",
      "\n",
      "    --de_novo_otu     BOOL        Optional  Output FASTA file with 'de novo' reads                  False\n",
      "\n",
      "       Read is 'de novo' if its alignment score passes E-value threshold, but both the identity\n",
      "       '-id', and the '-coverage' are below their corresponding thresholds\n",
      "       i.e. ID < %%id and COV < %%cov\n",
      "\n",
      "\n",
      "    --otu_map         BOOL        Optional  Output OTU map (input to QIIME's make_otu_table.py).    False\n",
      "                                            Cannot be used with 'no-best because\n",
      "                                            the grouping is done around the best alignment'\n",
      "\n",
      "\n",
      "    [ADVANCED]\n",
      "    --passes          INT,INT,INT Optional  Three intervals at which to place the seed on           L,L/2,3\n",
      "                                             the read (L is the seed length)\n",
      "\n",
      "    --edges           INT         Optional  Number (or percent if INT followed by %% sign) of       4\n",
      "                                            nucleotides to add to each edge of the read\n",
      "                                            prior to SW local alignment\n",
      "\n",
      "    --num_seeds       BOOL        Optional  Number of seeds matched before searching                2\n",
      "                                            for candidate LIS\n",
      "\n",
      "    --full_search     INT         Optional  Search for all 0-error and 1-error seed                 False\n",
      "                                            matches in the index rather than stopping\n",
      "                                            after finding a 0-error match (<1%% gain in\n",
      "                                            sensitivity with up four-fold decrease in speed)\n",
      "\n",
      "    --pid             BOOL        Optional  Add pid to output file names.                           False\n",
      "\n",
      "    -a                INT         Optional  DEPRECATED in favour of '-threads'. Number of           numCores\n",
      "                                            processing threads to use.\n",
      "                                            Automatically redirects to '-threads'\n",
      "\n",
      "    --threads         INT         Optional  Number of Processing threads to use                     2\n",
      "\n",
      "\n",
      "    [INDEXING]\n",
      "    --index           INT         Optional  Build reference database index                          2\n",
      "\n",
      "       By default when this option is not used, the program checks the reference index and\n",
      "       builds it if not already existing.\n",
      "       This can be changed by using '-index' as follows:\n",
      "       '-index 0' - skip indexing. If the index does not exist, the program will terminate\n",
      "                                and warn to build the index prior performing the alignment\n",
      "       '-index 1' - only perform the indexing and terminate\n",
      "       '-index 2' - the default behaviour, the same as when not using this option at all\n",
      "\n",
      "\n",
      "    -L                DOUBLE      Optional  Indexing: seed length.                                  18\n",
      "\n",
      "    -m                DOUBLE      Optional  Indexing: the amount of memory (in Mbytes) for          3072\n",
      "                                            building the index.\n",
      "\n",
      "    -v                BOOL        Optional  Produce verbose output when building the index          True\n",
      "\n",
      "    --interval        INT         Optional  Indexing: Positive integer: index every Nth L-mer in    1\n",
      "                                            the reference database e.g. '-interval 2'.\n",
      "\n",
      "    --max_pos         INT         Optional  Indexing: maximum (integer) number of positions to      1000\n",
      "                                            store for each unique L-mer.\n",
      "                                            If 0 - all positions are stored.\n",
      "\n",
      "\n",
      "    [HELP]\n",
      "    -h                BOOL        Optional  Print help information\n",
      "\n",
      "    --version         BOOL        Optional  Print SortMeRNA version number\n",
      "\n",
      "\n",
      "    [DEVELOPER]\n",
      "    --dbg_put_db      BOOL        Optional  \n",
      "    --cmd             BOOL        Optional  Launch an interactive session (command prompt)          False\n",
      "\n",
      "    --task            INT         Optional  Processing Task                                         4\n",
      "\n",
      "       Possible values: 0 - align. Only perform alignment\n",
      "                        1 - post-processing (log writing)\n",
      "                        2 - generate reports\n",
      "                        3 - align and post-process\n",
      "                        4 - all\n",
      "\n",
      "\n",
      "    --dbg-level       INT         Optional  Debug level                                             0\n",
      "\n",
      "      Controls verbosity of the execution trace. Default value of 0 corresponds to\n",
      "      the least verbose output.\n",
      "      The highest value currently is 2.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Help\n",
    "!sortmerna -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "95eb3884-9344-4a7b-8f35-1d99fd812201",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "#####---------------------------------------------Run in terminal------------------------------------------------------------\n",
    "# Save the location of all the databases into one folder\n",
    "# sortmernaREF=sortmerna_db/rRNA_databases/silva-arc-16s-id95.fasta,sortmerna_db/index/silva-arc-16s-id95:\\\n",
    "# sortmerna_db/rRNA_databases/silva-arc-23s-id98.fasta,sortmerna_db/index/silva-arc-23s-id98:\\\n",
    "# sortmerna_db/rRNA_databases/silva-bac-16s-id90.fasta,sortmerna_db/index/silva-bac-16s-id95:\\\n",
    "# sortmerna_db/rRNA_databases/silva-bac-23s-id98.fasta,sortmerna_db/index/silva-bac-23s-id98:\\\n",
    "# sortmerna_db/rRNA_databases/silva-euk-18s-id95.fasta,sortmerna_db/index/silva-euk-18s-id95:\\\n",
    "# sortmerna_db/rRNA_databases/silva-euk-28s-id98.fasta,sortmerna_db/index/silva-euk-28s-id98"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "9068a4c3-dd1f-49be-9398-8e0d0e901312",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------------Run in terminal---------------------------------------------------------------------\n",
    "# Run the indexing command (~8 minutes)\n",
    "# !indexdb_rna --ref $sortmernaREF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "f1d64ca4-afbc-44f2-a0ab-12e3b7601246",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  Program:     SortMeRNA version 2.1, 01/02/2016\n",
      "  Copyright:   2012-16 Bonsai Bioinformatics Research Group:\n",
      "               LIFL, University Lille 1, CNRS UMR 8022, INRIA Nord-Europe\n",
      "               2014-16 Knight Lab, Department of Pediatrics, UCSD, La Jolla,\n",
      "  Disclaimer:  SortMeRNA comes with ABSOLUTELY NO WARRANTY; without even the\n",
      "               implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n",
      "               See the GNU Lesser General Public License for more details.\n",
      "  Contact:     Evguenia Kopylova, jenya.kopylov@gmail.com \n",
      "               Laurent No, laurent.noe@lifl.fr\n",
      "               Hlne Touzet, helene.touzet@lifl.fr\n",
      "\n",
      "\n",
      "  usage:   ./indexdb_rna --ref db.fasta,db.idx [OPTIONS]:\n",
      "\n",
      "  --------------------------------------------------------------------------------------------------------\n",
      "  | parameter        value           description                                                 default |\n",
      "  --------------------------------------------------------------------------------------------------------\n",
      "     \u001b[1m--ref\u001b[0m           \u001b[4mSTRING,STRING\u001b[0m   FASTA reference file, index file                            \u001b[0;32mmandatory\u001b[0m\n",
      "                                      (ex. --ref /path/to/file1.fasta,/path/to/index1)\n",
      "                                       If passing multiple reference sequence files, separate\n",
      "                                       them by ':',\n",
      "                                      (ex. --ref /path/to/file1.fasta,/path/to/index1:/path/to/file2.fasta,path/to/index2)\n",
      "   [OPTIONS]:\n",
      "     \u001b[1m--tmpdir\u001b[0m        \u001b[4mSTRING\u001b[0m          directory where to write temporary files\n",
      "     \u001b[1m-m\u001b[0m              \u001b[4mINT\u001b[0m             the amount of memory (in Mbytes) for building the index     \u001b[4m3072\u001b[0m \n",
      "     \u001b[1m-L\u001b[0m              \u001b[4mINT\u001b[0m             seed length                                                 \u001b[4m18\u001b[0m\n",
      "     \u001b[1m--max_pos\u001b[0m       \u001b[4mINT\u001b[0m             maximum number of positions to store for each unique L-mer  \u001b[4m10000\u001b[0m\n",
      "                                      (setting --max_pos 0 will store all positions)\n",
      "     \u001b[1m-v\u001b[0m              \u001b[4mBOOL\u001b[0m            verbose\n",
      "     \u001b[1m-h\u001b[0m              \u001b[4mBOOL\u001b[0m            help\t\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!indexdb_rna -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "6e816183-af45-434e-a865-e78a4bc07c22",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "fb5537ee-5266-4b42-a52b-24631843147d",
   "metadata": {},
   "outputs": [],
   "source": [
    "######------------------------------Error so not used --------------------------------------------------- \n",
    "# Save variable of rRNA databases\n",
    "# # Save the location of all the databases into one folder\n",
    "# sortmernaREF=\"sortmerna_db/rRNA_databases/silva-arc-16s-id95.fasta,sortmerna_db/index/silva-arc-16s-id95:\\\n",
    "# sortmerna_db/rRNA_databases/silva-arc-23s-id98.fasta,sortmerna_db/index/silva-arc-23s-id98:\\\n",
    "# sortmerna_db/rRNA_databases/silva-bac-16s-id90.fasta,sortmerna_db/index/silva-bac-16s-id95:\\\n",
    "# sortmerna_db/rRNA_databases/silva-bac-23s-id98.fasta,sortmerna_db/index/silva-bac-23s-id98:\\\n",
    "# sortmerna_db/rRNA_databases/silva-euk-18s-id95.fasta,sortmerna_db/index/silva-euk-18s-id95:\\\n",
    "# sortmerna_db/rRNA_databases/silva-euk-28s-id98.fasta,sortmerna_db/index/silva-euk-28s-id98\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8d078be-c062-4e42-a403-903be62f2217",
   "metadata": {},
   "source": [
    "## 4. Alignment with star"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "f6a89a10-d070-4ac6-82e5-947c70071e12",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Channels:\n",
      " - bioconda\n",
      " - r\n",
      " - defaults\n",
      " - conda-forge\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/raotoo/miniconda\n",
      "\n",
      "  added / updated specs:\n",
      "    - star\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    htslib-1.18                |       h81da01d_0         2.4 MB  bioconda\n",
      "    libdeflate-1.19            |       hd590300_0          66 KB  conda-forge\n",
      "    libstdcxx-ng-13.2.0        |       h7e041cc_3         3.7 MB  conda-forge\n",
      "    libzlib-1.2.13             |       hd590300_5          60 KB  conda-forge\n",
      "    star-2.7.11a               |       h0033a41_0         8.1 MB  bioconda\n",
      "    zlib-1.2.13                |       hd590300_5          91 KB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        14.4 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  htslib             bioconda/linux-64::htslib-1.18-h81da01d_0 \n",
      "  libdeflate         conda-forge/linux-64::libdeflate-1.19-hd590300_0 \n",
      "  libzlib            conda-forge/linux-64::libzlib-1.2.13-hd590300_5 \n",
      "  star               bioconda/linux-64::star-2.7.11a-h0033a41_0 \n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "  libstdcxx-ng       pkgs/main::libstdcxx-ng-11.2.0-h12345~ --> conda-forge::libstdcxx-ng-13.2.0-h7e041cc_3 \n",
      "  zlib                    pkgs/main::zlib-1.2.13-h5eee18b_0 --> conda-forge::zlib-1.2.13-hd590300_5 \n",
      "\n",
      "The following packages will be SUPERSEDED by a higher-priority channel:\n",
      "\n",
      "  certifi            conda-forge/noarch::certifi-2023.11.1~ --> pkgs/main/linux-64::certifi-2023.11.17-py311h06a4308_0 \n",
      "  conda              conda-forge::conda-23.11.0-py311h38be~ --> pkgs/main::conda-23.11.0-py311h06a4308_0 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages:\n",
      "star-2.7.11a         | 8.1 MB    |                                       |   0% \n",
      "libstdcxx-ng-13.2.0  | 3.7 MB    |                                       |   0% \u001b[A\n",
      "\n",
      "htslib-1.18          | 2.4 MB    |                                       |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "zlib-1.2.13          | 91 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libdeflate-1.19      | 66 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libzlib-1.2.13       | 60 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "htslib-1.18          | 2.4 MB    | #9                                    |   5% \u001b[A\u001b[A\n",
      "\n",
      "htslib-1.18          | 2.4 MB    | #####################2                |  57% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "star-2.7.11a         | 8.1 MB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libdeflate-1.19      | 66 KB     | #########                             |  24% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "libstdcxx-ng-13.2.0  | 3.7 MB    | 1                                     |   0% \u001b[A\n",
      "star-2.7.11a         | 8.1 MB    | ##########################5           |  72% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libzlib-1.2.13       | 60 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libzlib-1.2.13       | 60 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "zlib-1.2.13          | 91 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "zlib-1.2.13          | 91 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libdeflate-1.19      | 66 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libdeflate-1.19      | 66 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "libstdcxx-ng-13.2.0  | 3.7 MB    | ##################################### | 100% \u001b[A\n",
      "\n",
      "htslib-1.18          | 2.4 MB    | ##################################### | 100% \u001b[A\u001b[A\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\n",
      "                                                                                \u001b[A\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n"
     ]
    }
   ],
   "source": [
    "!conda install -c bioconda star --yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "d68202b1-6f64-4039-ac37-f30f8129be75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --runMode genomeGenerate --genomeDir star_index --genomeFastaFiles genome/GRCm39.genome.fa --sjdbGTFfile annotation/gencode.vM34.annotation.gtf --runThreadN 4\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 10:36:50 ..... started STAR run\n",
      "Jan 22 10:36:50 ... starting to generate Genome files\n",
      "Jan 22 10:37:33 ..... processing annotations GTF\n",
      "Jan 22 10:37:54 ... starting to sort Suffix Array. This may take a long time...\n",
      "Jan 22 10:38:03 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Jan 22 11:05:25 ... loading chunks from disk, packing SA...\n",
      "Jan 22 11:06:26 ... finished generating suffix array\n",
      "Jan 22 11:06:26 ... generating Suffix Array index\n",
      "Jan 22 11:10:05 ... completed Suffix Array index\n",
      "Jan 22 11:10:06 ..... inserting junctions into the genome indices\n",
      "Jan 22 11:13:37 ... writing Genome to disk ...\n",
      "Jan 22 11:13:49 ... writing Suffix Array to disk ...\n",
      "Jan 22 11:15:26 ... writing SAindex to disk\n",
      "Jan 22 11:15:36 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "######################################Creates and Index of the reference genome###############################################\n",
    "####---------------------------------------------  Run only once-----------------------------------------------------------------\n",
    "# This can take up to 30 minutes to complete\n",
    "!STAR \\\n",
    "--runMode genomeGenerate \\\n",
    "--genomeDir star_index \\\n",
    "--genomeFastaFiles genome/* \\\n",
    "--sjdbGTFfile annotation/* \\\n",
    "--runThreadN 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "5a594bac-aa76-409c-90a8-ff693dbe6371",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: STAR  [options]... --genomeDir /path/to/genome/index/   --readFilesIn R1.fq R2.fq\n",
      "Spliced Transcripts Alignment to a Reference (c) Alexander Dobin, 2009-2022\n",
      "\n",
      "STAR version=2.7.11a\n",
      "STAR compilation time,server,dir=2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "For more details see:\n",
      "<https://github.com/alexdobin/STAR>\n",
      "<https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf>\n",
      "### versions\n",
      "versionGenome           2.7.4a\n",
      "    string: earliest genome index version compatible with this STAR release. Please do not change this value!\n",
      "\n",
      "### Parameter Files\n",
      "parametersFiles          -\n",
      "    string: name of a user-defined parameters file, \"-\": none. Can only be defined on the command line.\n",
      "\n",
      "### System\n",
      "sysShell            -\n",
      "    string: path to the shell binary, preferably bash, e.g. /bin/bash.\n",
      "                    - ... the default shell is executed, typically /bin/sh. This was reported to fail on some Ubuntu systems - then you need to specify path to bash.\n",
      "\n",
      "### Run Parameters\n",
      "runMode                         alignReads\n",
      "    string: type of the run.\n",
      "                                alignReads             ... map reads\n",
      "                                genomeGenerate         ... generate genome files\n",
      "                                inputAlignmentsFromBAM ... input alignments from BAM. Presently only works with --outWigType and --bamRemoveDuplicates options.\n",
      "                                liftOver               ... lift-over of GTF files (--sjdbGTFfile) between genome assemblies using chain file(s) from --genomeChainFiles.\n",
      "                                soloCellFiltering  </path/to/raw/count/dir/>   </path/to/output/prefix>    ... STARsolo cell filtering (\"calling\") without remapping, followed by the path to raw count directory and output (filtered) prefix\n",
      "\n",
      "runThreadN                      1\n",
      "    int: number of threads to run STAR\n",
      "\n",
      "runDirPerm                      User_RWX\n",
      "    string: permissions for the directories created at the run-time.\n",
      "                                User_RWX ... user-read/write/execute\n",
      "                                All_RWX  ... all-read/write/execute (same as chmod 777)\n",
      "\n",
      "runRNGseed                      777\n",
      "    int: random number generator seed.\n",
      "\n",
      "\n",
      "### Genome Parameters\n",
      "genomeDir                   ./GenomeDir/\n",
      "    string: path to the directory where genome files are stored (for --runMode alignReads) or will be generated (for --runMode generateGenome)\n",
      "\n",
      "genomeLoad                NoSharedMemory\n",
      "    string: mode of shared memory usage for the genome files. Only used with --runMode alignReads.\n",
      "                          LoadAndKeep     ... load genome into shared and keep it in memory after run\n",
      "                          LoadAndRemove   ... load genome into shared but remove it after run\n",
      "                          LoadAndExit     ... load genome into shared memory and exit, keeping the genome in memory for future runs\n",
      "                          Remove          ... do not map anything, just remove loaded genome from memory\n",
      "                          NoSharedMemory  ... do not use shared memory, each job will have its own private copy of the genome\n",
      "\n",
      "genomeFastaFiles            -\n",
      "    string(s): path(s) to the fasta files with the genome sequences, separated by spaces. These files should be plain text FASTA files, they *cannot* be zipped.\n",
      "                            Required for the genome generation (--runMode genomeGenerate). Can also be used in the mapping (--runMode alignReads) to add extra (new) sequences to the genome (e.g. spike-ins).\n",
      "\n",
      "genomeChainFiles            -\n",
      "    string: chain files for genomic liftover. Only used with --runMode liftOver .\n",
      "\n",
      "genomeFileSizes             0\n",
      "    uint(s)>0: genome files exact sizes in bytes. Typically, this should not be defined by the user.\n",
      "    \n",
      "genomeTransformOutput       None\n",
      "    string(s)               which output to transform back to original genome\n",
      "                            SAM     ... SAM/BAM alignments\n",
      "                            SJ      ... splice junctions (SJ.out.tab)\n",
      "                            Quant   ... quantifications (from --quantMode option)\n",
      "                            None    ... no transformation of the output        \n",
      "\n",
      "genomeChrSetMitochondrial   chrM M MT\n",
      "    string(s)               names of the mitochondrial chromosomes. Presently only used for STARsolo statisics output/\n",
      "\n",
      "### Genome Indexing Parameters - only used with --runMode genomeGenerate\n",
      "genomeChrBinNbits           18\n",
      "    int: =log2(chrBin), where chrBin is the size of the bins for genome storage: each chromosome will occupy an integer number of bins. For a genome with large number of contigs, it is recommended to scale this parameter as min(18, log2[max(GenomeLength/NumberOfReferences,ReadLength)]).\n",
      "\n",
      "genomeSAindexNbases         14\n",
      "    int: length (bases) of the SA pre-indexing string. Typically between 10 and 15. Longer strings will use much more memory, but allow faster searches. For small genomes, the parameter --genomeSAindexNbases must be scaled down to min(14, log2(GenomeLength)/2 - 1).\n",
      "\n",
      "genomeSAsparseD             1\n",
      "    int>0: suffux array sparsity, i.e. distance between indices: use bigger numbers to decrease needed RAM at the cost of mapping speed reduction\n",
      "\n",
      "genomeSuffixLengthMax       -1\n",
      "    int: maximum length of the suffixes, has to be longer than read length. -1 = infinite.\n",
      "    \n",
      "genomeTransformType         None\n",
      "    string: type of genome transformation\n",
      "                            None       ... no transformation\n",
      "                            Haploid    ... replace reference alleles with alternative alleles from VCF file (e.g. consensus allele)\n",
      "                            Diploid    ... create two haplotypes for each chromosome listed in VCF file, for genotypes 1|2, assumes perfect phasing (e.g. personal genome)\n",
      "\n",
      "genomeTransformVCF          -\n",
      "    string: path to VCF file for genome transformation\n",
      "\n",
      "\n",
      "    \n",
      "#####UnderDevelopment_begin : not supported - do not use\n",
      "genomeType                  Full\n",
      "    string: type of genome to generate\n",
      "                            Full                ... full (normal) genome\n",
      "                            Transcriptome       ... genome consists of transcript sequences\n",
      "                            SuperTransriptome   ... genome consists of superTranscript sequences\n",
      "#####UnderDevelopment_end\n",
      "\n",
      "# DEPRECATED: please use --genomeTransformVCF and --genomeTransformType options instead.\n",
      "#genomeConsensusFile         -\n",
      "#    string: VCF file with consensus SNPs (i.e. alternative allele is the major (AF>0.5) allele)\n",
      "# DEPRECATED \n",
      "\n",
      "\n",
      "\n",
      "### Splice Junctions Database\n",
      "sjdbFileChrStartEnd                     -\n",
      "    string(s): path to the files with genomic coordinates (chr <tab> start <tab> end <tab> strand) for the splice junction introns. Multiple files can be supplied wand will be concatenated.\n",
      "\n",
      "sjdbGTFfile                             -\n",
      "    string: path to the GTF file with annotations\n",
      "\n",
      "sjdbGTFchrPrefix                        -\n",
      "    string: prefix for chromosome names in a GTF file (e.g. 'chr' for using ENSMEBL annotations with UCSC genomes)\n",
      "\n",
      "sjdbGTFfeatureExon                      exon\n",
      "    string: feature type in GTF file to be used as exons for building transcripts\n",
      "\n",
      "sjdbGTFtagExonParentTranscript          transcript_id\n",
      "    string: GTF attribute name for parent transcript ID (default \"transcript_id\" works for GTF files)\n",
      "\n",
      "sjdbGTFtagExonParentGene                gene_id\n",
      "    string: GTF attribute name for parent gene ID (default \"gene_id\" works for GTF files)\n",
      "\n",
      "sjdbGTFtagExonParentGeneName            gene_name\n",
      "    string(s): GTF attrbute name for parent gene name\n",
      "\n",
      "sjdbGTFtagExonParentGeneType            gene_type gene_biotype\n",
      "    string(s): GTF attrbute name for parent gene type\n",
      "\n",
      "sjdbOverhang                            100\n",
      "    int>0: length of the donor/acceptor sequence on each side of the junctions, ideally = (mate_length - 1)\n",
      "\n",
      "sjdbScore                               2\n",
      "    int: extra alignment score for alignments that cross database junctions\n",
      "\n",
      "sjdbInsertSave                          Basic\n",
      "    string: which files to save when sjdb junctions are inserted on the fly at the mapping step\n",
      "                    Basic ... only small junction / transcript files\n",
      "                    All   ... all files including big Genome, SA and SAindex - this will create a complete genome directory\n",
      "\n",
      "### Variation parameters\n",
      "varVCFfile                              -\n",
      "    string: path to the VCF file that contains variation data. The 10th column should contain the genotype information, e.g. 0/1\n",
      "\n",
      "### Input Files\n",
      "inputBAMfile                -\n",
      "    string: path to BAM input file, to be used with --runMode inputAlignmentsFromBAM\n",
      "\n",
      "### Read Parameters\n",
      "readFilesType               Fastx\n",
      "    string: format of input read files\n",
      "                            Fastx       ... FASTA or FASTQ\n",
      "                            SAM SE      ... SAM or BAM single-end reads; for BAM use --readFilesCommand samtools view\n",
      "                            SAM PE      ... SAM or BAM paired-end reads; for BAM use --readFilesCommand samtools view\n",
      "                            \n",
      "readFilesSAMattrKeep        All\n",
      "    string(s): for --readFilesType SAM SE/PE, which SAM tags to keep in the output BAM, e.g.: --readFilesSAMtagsKeep RG PL\n",
      "                            All     ... keep all tags\n",
      "                            None    ... do not keep any tags\n",
      "\n",
      "readFilesIn                 Read1 Read2\n",
      "    string(s): paths to files that contain input read1 (and, if needed,  read2)\n",
      "\n",
      "readFilesManifest           -\n",
      "    string: path to the \"manifest\" file with the names of read files. The manifest file should contain 3 tab-separated columns:\n",
      "            paired-end reads: read1_file_name $tab$ read2_file_name $tab$ read_group_line.\n",
      "            single-end reads: read1_file_name $tab$ -               $tab$ read_group_line.\n",
      "            Spaces, but not tabs are allowed in file names.\n",
      "            If read_group_line does not start with ID:, it can only contain one ID field, and ID: will be added to it.\n",
      "            If read_group_line starts with ID:, it can contain several fields separated by $tab$, and all fields will be be copied verbatim into SAM @RG header line.\n",
      "\n",
      "readFilesPrefix             -\n",
      "    string: prefix for the read files names, i.e. it will be added in front of the strings in --readFilesIn\n",
      "\n",
      "readFilesCommand             -\n",
      "    string(s): command line to execute for each of the input file. This command should generate FASTA or FASTQ text and send it to stdout\n",
      "               For example: zcat - to uncompress .gz files, bzcat - to uncompress .bz2 files, etc.\n",
      "\n",
      "readMapNumber               -1\n",
      "    int: number of reads to map from the beginning of the file\n",
      "                            -1: map all reads\n",
      "\n",
      "readMatesLengthsIn          NotEqual\n",
      "    string: Equal/NotEqual - lengths of names,sequences,qualities for both mates are the same  / not the same. NotEqual is safe in all situations.\n",
      "\n",
      "readNameSeparator           /\n",
      "    string(s): character(s) separating the part of the read names that will be trimmed in output (read name after space is always trimmed)\n",
      "\n",
      "readQualityScoreBase        33\n",
      "    int>=0: number to be subtracted from the ASCII code to get Phred quality score\n",
      "\n",
      "### Read Clipping\n",
      "\n",
      "clipAdapterType             Hamming\n",
      "    string:                 adapter clipping type\n",
      "                            Hamming ... adapter clipping based on Hamming distance, with the number of mismatches controlled by --clip5pAdapterMMp\n",
      "                            CellRanger4 ... 5p and 3p adapter clipping similar to CellRanger4. Utilizes Opal package by Martin oi: https://github.com/Martinsos/opal\n",
      "                            None ... no adapter clipping, all other clip* parameters are disregarded\n",
      "                            \n",
      "clip3pNbases                 0\n",
      "    int(s): number(s) of bases to clip from 3p of each mate. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip3pAdapterSeq            -\n",
      "    string(s): adapter sequences to clip from 3p of each mate.  If one value is given, it will be assumed the same for both mates.\n",
      "                            polyA ... polyA sequence with the length equal to read length\n",
      "\n",
      "clip3pAdapterMMp            0.1\n",
      "    double(s): max proportion of mismatches for 3p adapter clipping for each mate.  If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip3pAfterAdapterNbases    0\n",
      "    int(s): number of bases to clip from 3p of each mate after the adapter clipping. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip5pNbases                 0\n",
      "    int(s): number(s) of bases to clip from 5p of each mate. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "#####UnderDevelopment_begin : not supported - do not use   \n",
      "clip5pAdapterSeq            -\n",
      "    string(s): adapter sequences to clip from 5p of each mate, separated by space.\n",
      "\n",
      "clip5pAdapterMMp            0.1\n",
      "    double(s): max proportion of mismatches for 5p adapter clipping for each mate, separated by space\n",
      "\n",
      "clip5pAfterAdapterNbases    0\n",
      "    int(s): number of bases to clip from 5p of each mate after the adapter clipping, separated by space.\n",
      "#####UnderDevelopment_end\n",
      "\n",
      "### Limits\n",
      "limitGenomeGenerateRAM               31000000000\n",
      "    int>0: maximum available RAM (bytes) for genome generation\n",
      "\n",
      "limitIObufferSize                    30000000 50000000\n",
      "    int>0: max available buffers size (bytes) for input/output, per thread\n",
      "\n",
      "limitOutSAMoneReadBytes              100000\n",
      "    int>0: max size of the SAM record (bytes) for one read. Recommended value: >(2*(LengthMate1+LengthMate2+100)*outFilterMultimapNmax\n",
      "\n",
      "limitOutSJoneRead                    1000\n",
      "    int>0: max number of junctions for one read (including all multi-mappers)\n",
      "\n",
      "limitOutSJcollapsed                  1000000\n",
      "    int>0: max number of collapsed junctions\n",
      "\n",
      "limitBAMsortRAM                         0\n",
      "    int>=0: maximum available RAM (bytes) for sorting BAM. If =0, it will be set to the genome index size. 0 value can only be used with --genomeLoad NoSharedMemory option.\n",
      "\n",
      "limitSjdbInsertNsj                     1000000\n",
      "    int>=0: maximum number of junction to be inserted to the genome on the fly at the mapping stage, including those from annotations and those detected in the 1st step of the 2-pass run\n",
      "\n",
      "limitNreadsSoft                        -1\n",
      "    int: soft limit on the number of reads\n",
      "\n",
      "### Output: general\n",
      "outFileNamePrefix               ./\n",
      "    string: output files name prefix (including full or relative path). Can only be defined on the command line.\n",
      "\n",
      "outTmpDir                       -\n",
      "    string: path to a directory that will be used as temporary by STAR. All contents of this directory will be removed!\n",
      "            - the temp directory will default to outFileNamePrefix_STARtmp\n",
      "\n",
      "outTmpKeep                      None\n",
      "    string: whether to keep the tempporary files after STAR runs is finished\n",
      "                                None ... remove all temporary files\n",
      "                                All .. keep all files\n",
      "\n",
      "outStd                          Log\n",
      "    string: which output will be directed to stdout (standard out)\n",
      "                                Log                    ... log messages\n",
      "                                SAM                    ... alignments in SAM format (which normally are output to Aligned.out.sam file), normal standard output will go into Log.std.out\n",
      "                                BAM_Unsorted           ... alignments in BAM format, unsorted. Requires --outSAMtype BAM Unsorted\n",
      "                                BAM_SortedByCoordinate ... alignments in BAM format, sorted by coordinate. Requires --outSAMtype BAM SortedByCoordinate\n",
      "                                BAM_Quant              ... alignments to transcriptome in BAM format, unsorted. Requires --quantMode TranscriptomeSAM\n",
      "\n",
      "outReadsUnmapped                None\n",
      "   string: output of unmapped and partially mapped (i.e. mapped only one mate of a paired end read) reads in separate file(s).\n",
      "                                None    ... no output\n",
      "                                Fastx   ... output in separate fasta/fastq files, Unmapped.out.mate1/2\n",
      "\n",
      "outQSconversionAdd              0\n",
      "   int: add this number to the quality score (e.g. to convert from Illumina to Sanger, use -31)\n",
      "\n",
      "outMultimapperOrder             Old_2.4\n",
      "    string: order of multimapping alignments in the output files\n",
      "                                Old_2.4             ... quasi-random order used before 2.5.0\n",
      "                                Random              ... random order of alignments for each multi-mapper. Read mates (pairs) are always adjacent, all alignment for each read stay together. This option will become default in the future releases.\n",
      "\n",
      "### Output: SAM and BAM\n",
      "outSAMtype                      SAM\n",
      "    strings: type of SAM/BAM output\n",
      "                                1st word:\n",
      "                                BAM  ... output BAM without sorting\n",
      "                                SAM  ... output SAM without sorting\n",
      "                                None ... no SAM/BAM output\n",
      "                                2nd, 3rd:\n",
      "                                Unsorted           ... standard unsorted\n",
      "                                SortedByCoordinate ... sorted by coordinate. This option will allocate extra memory for sorting which can be specified by --limitBAMsortRAM.\n",
      "\n",
      "outSAMmode                      Full\n",
      "    string: mode of SAM output\n",
      "                                None ... no SAM output\n",
      "                                Full ... full SAM output\n",
      "                                NoQS ... full SAM but without quality scores\n",
      "\n",
      "outSAMstrandField               None\n",
      "    string: Cufflinks-like strand field flag\n",
      "                                None        ... not used\n",
      "                                intronMotif ... strand derived from the intron motif. This option changes the output alignments: reads with inconsistent and/or non-canonical introns are filtered out.\n",
      "\n",
      "outSAMattributes                Standard\n",
      "    string: a string of desired SAM attributes, in the order desired for the output SAM. Tags can be listed in any combination/order.\n",
      "                                ***Presets:\n",
      "                                None        ... no attributes\n",
      "                                Standard    ... NH HI AS nM\n",
      "                                All         ... NH HI AS nM NM MD jM jI MC ch                                                                    \n",
      "                                ***Alignment:\n",
      "                                NH          ... number of loci the reads maps to: =1 for unique mappers, >1 for multimappers. Standard SAM tag.\n",
      "                                HI          ... multiple alignment index, starts with --outSAMattrIHstart (=1 by default). Standard SAM tag.\n",
      "                                AS          ... local alignment score, +1/-1 for matches/mismateches, score* penalties for indels and gaps. For PE reads, total score for two mates. Stadnard SAM tag.\n",
      "                                nM          ... number of mismatches. For PE reads, sum over two mates.\n",
      "                                NM          ... edit distance to the reference (number of mismatched + inserted + deleted bases) for each mate. Standard SAM tag.\n",
      "                                MD          ... string encoding mismatched and deleted reference bases (see standard SAM specifications). Standard SAM tag.\n",
      "                                jM          ... intron motifs for all junctions (i.e. N in CIGAR): 0: non-canonical; 1: GT/AG, 2: CT/AC, 3: GC/AG, 4: CT/GC, 5: AT/AC, 6: GT/AT. If splice junctions database is used, and a junction is annotated, 20 is added to its motif value.\n",
      "                                jI          ... start and end of introns for all junctions (1-based).\n",
      "                                XS          ... alignment strand according to --outSAMstrandField.\n",
      "                                MC          ... mate's CIGAR string. Standard SAM tag.\n",
      "                                ch          ... marks all segment of all chimeric alingments for --chimOutType WithinBAM output.\n",
      "                                cN          ... number of bases clipped from the read ends: 5' and 3'\n",
      "                                ***Variation:\n",
      "                                vA          ... variant allele\n",
      "                                vG          ... genomic coordinate of the variant overlapped by the read.\n",
      "                                vW          ... 1 - alignment passes WASP filtering; 2,3,4,5,6,7 - alignment does not pass WASP filtering. Requires --waspOutputMode SAMtag.\n",
      "                                ha          ... haplotype (1/2) when mapping to the diploid genome. Requires genome generated with --genomeTransformType Diploid .                               \n",
      "                                ***STARsolo:\n",
      "                                CR CY UR UY ... sequences and quality scores of cell barcodes and UMIs for the solo* demultiplexing.\n",
      "                                GX GN       ... gene ID and gene name for unique-gene reads.\n",
      "                                gx gn       ... gene IDs and gene names for unique- and multi-gene reads.\n",
      "                                CB UB       ... error-corrected cell barcodes and UMIs for solo* demultiplexing. Requires --outSAMtype BAM SortedByCoordinate.\n",
      "                                sM          ... assessment of CB and UMI.\n",
      "                                sS          ... sequence of the entire barcode (CB,UMI,adapter).\n",
      "                                sQ          ... quality of the entire barcode.\n",
      "                                sF          ... type of feature overlap and number of features for each alignment\n",
      "                                ***Unsupported/undocumented:\n",
      "                                rB          ... alignment block read/genomic coordinates.\n",
      "                                vR          ... read coordinate of the variant.\n",
      "\n",
      "outSAMattrIHstart               1\n",
      "    int>=0:                     start value for the IH attribute. 0 may be required by some downstream software, such as Cufflinks or StringTie.\n",
      "\n",
      "outSAMunmapped                  None\n",
      "    string(s): output of unmapped reads in the SAM format\n",
      "                                1st word:\n",
      "                                None   ... no output\n",
      "                                Within ... output unmapped reads within the main SAM file (i.e. Aligned.out.sam)\n",
      "                                2nd word:\n",
      "                                KeepPairs ... record unmapped mate for each alignment, and, in case of unsorted output, keep it adjacent to its mapped mate. Only affects multi-mapping reads.\n",
      "\n",
      "outSAMorder                     Paired\n",
      "    string: type of sorting for the SAM output\n",
      "                                Paired: one mate after the other for all paired alignments\n",
      "                                PairedKeepInputOrder: one mate after the other for all paired alignments, the order is kept the same as in the input FASTQ files\n",
      "\n",
      "outSAMprimaryFlag        OneBestScore\n",
      "    string: which alignments are considered primary - all others will be marked with 0x100 bit in the FLAG\n",
      "                                OneBestScore ... only one alignment with the best score is primary\n",
      "                                AllBestScore ... all alignments with the best score are primary\n",
      "\n",
      "outSAMreadID            Standard\n",
      "    string: read ID record type\n",
      "                                Standard ... first word (until space) from the FASTx read ID line, removing /1,/2 from the end\n",
      "                                Number   ... read number (index) in the FASTx file\n",
      "\n",
      "outSAMmapqUnique        255\n",
      "    int: 0 to 255: the MAPQ value for unique mappers\n",
      "\n",
      "outSAMflagOR           0\n",
      "    int: 0 to 65535: sam FLAG will be bitwise OR'd with this value, i.e. FLAG=FLAG | outSAMflagOR. This is applied after all flags have been set by STAR, and after outSAMflagAND. Can be used to set specific bits that are not set otherwise.\n",
      "\n",
      "outSAMflagAND           65535\n",
      "    int: 0 to 65535: sam FLAG will be bitwise AND'd with this value, i.e. FLAG=FLAG & outSAMflagOR. This is applied after all flags have been set by STAR, but before outSAMflagOR. Can be used to unset specific bits that are not set otherwise.\n",
      "\n",
      "outSAMattrRGline        -\n",
      "    string(s): SAM/BAM read group line. The first word contains the read group identifier and must start with \"ID:\", e.g. --outSAMattrRGline ID:xxx CN:yy \"DS:z z z\".\n",
      "            xxx will be added as RG tag to each output alignment. Any spaces in the tag values have to be double quoted.\n",
      "            Comma separated RG lines correspons to different (comma separated) input files in --readFilesIn. Commas have to be surrounded by spaces, e.g.\n",
      "            --outSAMattrRGline ID:xxx , ID:zzz \"DS:z z\" , ID:yyy DS:yyyy\n",
      "\n",
      "outSAMheaderHD          -\n",
      "    strings: @HD (header) line of the SAM header\n",
      "\n",
      "outSAMheaderPG          -\n",
      "    strings: extra @PG (software) line of the SAM header (in addition to STAR)\n",
      "\n",
      "outSAMheaderCommentFile -\n",
      "    string: path to the file with @CO (comment) lines of the SAM header\n",
      "\n",
      "outSAMfilter            None\n",
      "    string(s): filter the output into main SAM/BAM files\n",
      "                        KeepOnlyAddedReferences ... only keep the reads for which all alignments are to the extra reference sequences added with --genomeFastaFiles at the mapping stage.\n",
      "                        KeepAllAddedReferences ...  keep all alignments to the extra reference sequences added with --genomeFastaFiles at the mapping stage.\n",
      "\n",
      "\n",
      "outSAMmultNmax          -1\n",
      "    int: max number of multiple alignments for a read that will be output to the SAM/BAM files. Note that if this value is not equal to -1, the top scoring alignment will be output first\n",
      "                        -1 ... all alignments (up to --outFilterMultimapNmax) will be output\n",
      "\n",
      "outSAMtlen              1\n",
      "    int: calculation method for the TLEN field in the SAM/BAM files\n",
      "                        1 ... leftmost base of the (+)strand mate to rightmost base of the (-)mate. (+)sign for the (+)strand mate\n",
      "                        2 ... leftmost base of any mate to rightmost base of any mate. (+)sign for the mate with the leftmost base. This is different from 1 for overlapping mates with protruding ends\n",
      "\n",
      "outBAMcompression       1\n",
      "    int: -1 to 10  BAM compression level, -1=default compression (6?), 0=no compression, 10=maximum compression\n",
      "\n",
      "outBAMsortingThreadN    0\n",
      "    int: >=0: number of threads for BAM sorting. 0 will default to min(6,--runThreadN).\n",
      "\n",
      "outBAMsortingBinsN      50\n",
      "    int: >0:  number of genome bins fo coordinate-sorting\n",
      "\n",
      "### BAM processing\n",
      "bamRemoveDuplicatesType  -\n",
      "    string: mark duplicates in the BAM file, for now only works with (i) sorted BAM fed with inputBAMfile, and (ii) for paired-end alignments only\n",
      "                        -                       ... no duplicate removal/marking\n",
      "                        UniqueIdentical         ... mark all multimappers, and duplicate unique mappers. The coordinates, FLAG, CIGAR must be identical\n",
      "                        UniqueIdenticalNotMulti  ... mark duplicate unique mappers but not multimappers.\n",
      "\n",
      "bamRemoveDuplicatesMate2basesN   0\n",
      "    int>0: number of bases from the 5' of mate 2 to use in collapsing (e.g. for RAMPAGE)\n",
      "\n",
      "### Output Wiggle\n",
      "outWigType          None\n",
      "    string(s): type of signal output, e.g. \"bedGraph\" OR \"bedGraph read1_5p\". Requires sorted BAM: --outSAMtype BAM SortedByCoordinate .\n",
      "                    1st word:\n",
      "                    None       ... no signal output\n",
      "                    bedGraph   ... bedGraph format\n",
      "                    wiggle     ... wiggle format\n",
      "                    2nd word:\n",
      "                    read1_5p   ... signal from only 5' of the 1st read, useful for CAGE/RAMPAGE etc\n",
      "                    read2      ... signal from only 2nd read\n",
      "\n",
      "outWigStrand        Stranded\n",
      "    string: strandedness of wiggle/bedGraph output\n",
      "                    Stranded   ...  separate strands, str1 and str2\n",
      "                    Unstranded ...  collapsed strands\n",
      "\n",
      "outWigReferencesPrefix    -\n",
      "    string: prefix matching reference names to include in the output wiggle file, e.g. \"chr\", default \"-\" - include all references\n",
      "\n",
      "outWigNorm              RPM\n",
      "    string: type of normalization for the signal\n",
      "                        RPM    ... reads per million of mapped reads\n",
      "                        None   ... no normalization, \"raw\" counts\n",
      "\n",
      "### Output Filtering\n",
      "outFilterType                   Normal\n",
      "    string: type of filtering\n",
      "                                Normal  ... standard filtering using only current alignment\n",
      "                                BySJout ... keep only those reads that contain junctions that passed filtering into SJ.out.tab\n",
      "\n",
      "outFilterMultimapScoreRange     1\n",
      "    int: the score range below the maximum score for multimapping alignments\n",
      "\n",
      "outFilterMultimapNmax           10\n",
      "    int: maximum number of loci the read is allowed to map to. Alignments (all of them) will be output only if the read maps to no more loci than this value.\n",
      "         Otherwise no alignments will be output, and the read will be counted as \"mapped to too many loci\" in the Log.final.out .\n",
      "\n",
      "outFilterMismatchNmax           10\n",
      "    int: alignment will be output only if it has no more mismatches than this value.\n",
      "\n",
      "outFilterMismatchNoverLmax      0.3\n",
      "    real: alignment will be output only if its ratio of mismatches to *mapped* length is less than or equal to this value.\n",
      "\n",
      "outFilterMismatchNoverReadLmax  1.0\n",
      "    real: alignment will be output only if its ratio of mismatches to *read* length is less than or equal to this value.\n",
      "\n",
      "\n",
      "outFilterScoreMin               0\n",
      "    int: alignment will be output only if its score is higher than or equal to this value.\n",
      "\n",
      "outFilterScoreMinOverLread      0.66\n",
      "    real: same as outFilterScoreMin, but  normalized to read length (sum of mates' lengths for paired-end reads)\n",
      "\n",
      "outFilterMatchNmin              0\n",
      "    int: alignment will be output only if the number of matched bases is higher than or equal to this value.\n",
      "\n",
      "outFilterMatchNminOverLread     0.66\n",
      "    real: sam as outFilterMatchNmin, but normalized to the read length (sum of mates' lengths for paired-end reads).\n",
      "\n",
      "outFilterIntronMotifs           None\n",
      "    string: filter alignment using their motifs\n",
      "                None                           ... no filtering\n",
      "                RemoveNoncanonical             ... filter out alignments that contain non-canonical junctions\n",
      "                RemoveNoncanonicalUnannotated  ... filter out alignments that contain non-canonical unannotated junctions when using annotated splice junctions database. The annotated non-canonical junctions will be kept.\n",
      "\n",
      "outFilterIntronStrands          RemoveInconsistentStrands\n",
      "    string: filter alignments\n",
      "                RemoveInconsistentStrands      ... remove alignments that have junctions with inconsistent strands\n",
      "                None                           ... no filtering\n",
      "\n",
      "### Output splice junctions (SJ.out.tab)\n",
      "outSJtype                       Standard\n",
      "    string: type of splice junction output\n",
      "                                Standard    ... standard SJ.out.tab output\n",
      "                                None        ... no splice junction output\n",
      "\n",
      "### Output Filtering: Splice Junctions\n",
      "outSJfilterReads                All\n",
      "    string: which reads to consider for collapsed splice junctions output\n",
      "                                All     ... all reads, unique- and multi-mappers\n",
      "                                Unique  ... uniquely mapping reads only\n",
      "\n",
      "outSJfilterOverhangMin          30  12  12  12\n",
      "    4 integers:    minimum overhang length for splice junctions on both sides for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterCountUniqueMin       3   1   1   1\n",
      "    4 integers: minimum uniquely mapping read count per junction for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                Junctions are output if one of outSJfilterCountUniqueMin OR outSJfilterCountTotalMin conditions are satisfied\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterCountTotalMin     3   1   1   1\n",
      "    4 integers: minimum total (multi-mapping+unique) read count per junction for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                Junctions are output if one of outSJfilterCountUniqueMin OR outSJfilterCountTotalMin conditions are satisfied\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterDistToOtherSJmin     10  0   5   10\n",
      "    4 integers>=0: minimum allowed distance to other junctions' donor/acceptor\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterIntronMaxVsReadN        50000 100000 200000\n",
      "    N integers>=0: maximum gap allowed for junctions supported by 1,2,3,,,N reads\n",
      "                                i.e. by default junctions supported by 1 read can have gaps <=50000b, by 2 reads: <=100000b, by 3 reads: <=200000. by >=4 reads any gap <=alignIntronMax\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "### Scoring\n",
      "scoreGap                     0\n",
      "    int: splice junction penalty (independent on intron motif)\n",
      "\n",
      "scoreGapNoncan               -8\n",
      "    int: non-canonical junction penalty (in addition to scoreGap)\n",
      "\n",
      "scoreGapGCAG                 -4\n",
      "    GC/AG and CT/GC junction penalty (in addition to scoreGap)\n",
      "\n",
      "scoreGapATAC                 -8\n",
      "    AT/AC  and GT/AT junction penalty  (in addition to scoreGap)\n",
      "\n",
      "scoreGenomicLengthLog2scale   -0.25\n",
      "    extra score logarithmically scaled with genomic length of the alignment: scoreGenomicLengthLog2scale*log2(genomicLength)\n",
      "\n",
      "scoreDelOpen                 -2\n",
      "    deletion open penalty\n",
      "\n",
      "scoreDelBase                 -2\n",
      "    deletion extension penalty per base (in addition to scoreDelOpen)\n",
      "\n",
      "scoreInsOpen                 -2\n",
      "    insertion open penalty\n",
      "\n",
      "scoreInsBase                 -2\n",
      "    insertion extension penalty per base (in addition to scoreInsOpen)\n",
      "\n",
      "scoreStitchSJshift           1\n",
      "    maximum score reduction while searching for SJ boundaries in the stitching step\n",
      "\n",
      "\n",
      "### Alignments and Seeding\n",
      "\n",
      "seedSearchStartLmax             50\n",
      "    int>0: defines the search start point through the read - the read is split into pieces no longer than this value\n",
      "\n",
      "seedSearchStartLmaxOverLread    1.0\n",
      "    real: seedSearchStartLmax normalized to read length (sum of mates' lengths for paired-end reads)\n",
      "\n",
      "seedSearchLmax       0\n",
      "    int>=0: defines the maximum length of the seeds, if =0 seed length is not limited\n",
      "\n",
      "seedMultimapNmax      10000\n",
      "    int>0: only pieces that map fewer than this value are utilized in the stitching procedure\n",
      "\n",
      "seedPerReadNmax       1000\n",
      "    int>0: max number of seeds per read\n",
      "\n",
      "seedPerWindowNmax     50\n",
      "    int>0: max number of seeds per window\n",
      "\n",
      "seedNoneLociPerWindow    10\n",
      "    int>0: max number of one seed loci per window\n",
      "\n",
      "seedSplitMin                12\n",
      "    int>0: min length of the seed sequences split by Ns or mate gap\n",
      "\n",
      "seedMapMin              5\n",
      "    int>0: min length of seeds to be mapped\n",
      "\n",
      "alignIntronMin              21\n",
      "    minimum intron size: genomic gap is considered intron if its length>=alignIntronMin, otherwise it is considered Deletion\n",
      "\n",
      "alignIntronMax              0\n",
      "    maximum intron size, if 0, max intron size will be determined by (2^winBinNbits)*winAnchorDistNbins\n",
      "\n",
      "alignMatesGapMax            0\n",
      "    maximum gap between two mates, if 0, max intron gap will be determined by (2^winBinNbits)*winAnchorDistNbins\n",
      "\n",
      "alignSJoverhangMin          5\n",
      "    int>0: minimum overhang (i.e. block size) for spliced alignments\n",
      "\n",
      "alignSJstitchMismatchNmax   0 -1 0 0\n",
      "    4*int>=0: maximum number of mismatches for stitching of the splice junctions (-1: no limit).\n",
      "                            (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif.\n",
      "\n",
      "alignSJDBoverhangMin        3\n",
      "    int>0: minimum overhang (i.e. block size) for annotated (sjdb) spliced alignments\n",
      "\n",
      "alignSplicedMateMapLmin     0\n",
      "    int>0: minimum mapped length for a read mate that is spliced\n",
      "\n",
      "alignSplicedMateMapLminOverLmate 0.66\n",
      "    real>0: alignSplicedMateMapLmin normalized to mate length\n",
      "\n",
      "alignWindowsPerReadNmax     10000\n",
      "    int>0: max number of windows per read\n",
      "\n",
      "alignTranscriptsPerWindowNmax       100\n",
      "    int>0: max number of transcripts per window\n",
      "\n",
      "alignTranscriptsPerReadNmax               10000\n",
      "    int>0: max number of different alignments per read to consider\n",
      "\n",
      "alignEndsType           Local\n",
      "    string: type of read ends alignment\n",
      "                        Local             ... standard local alignment with soft-clipping allowed\n",
      "                        EndToEnd          ... force end-to-end read alignment, do not soft-clip\n",
      "                        Extend5pOfRead1   ... fully extend only the 5p of the read1, all other ends: local alignment\n",
      "                        Extend5pOfReads12 ... fully extend only the 5p of the both read1 and read2, all other ends: local alignment\n",
      "\n",
      "alignEndsProtrude       0    ConcordantPair\n",
      "    int, string:        allow protrusion of alignment ends, i.e. start (end) of the +strand mate downstream of the start (end) of the -strand mate\n",
      "                        1st word: int: maximum number of protrusion bases allowed\n",
      "                        2nd word: string:\n",
      "                                            ConcordantPair ... report alignments with non-zero protrusion as concordant pairs\n",
      "                                            DiscordantPair ... report alignments with non-zero protrusion as discordant pairs\n",
      "\n",
      "alignSoftClipAtReferenceEnds    Yes\n",
      "    string: allow the soft-clipping of the alignments past the end of the chromosomes\n",
      "                                Yes ... allow\n",
      "                                No  ... prohibit, useful for compatibility with Cufflinks\n",
      "\n",
      "alignInsertionFlush     None\n",
      "    string: how to flush ambiguous insertion positions\n",
      "                        None    ... insertions are not flushed\n",
      "                        Right   ... insertions are flushed to the right\n",
      "\n",
      "### Paired-End reads\n",
      "peOverlapNbasesMin          0\n",
      "    int>=0:             minimum number of overlap bases to trigger mates merging and realignment. Specify >0 value to switch on the \"merginf of overlapping mates\" algorithm.\n",
      "\n",
      "peOverlapMMp                0.01\n",
      "    real, >=0 & <1:     maximum proportion of mismatched bases in the overlap area\n",
      "\n",
      "### Windows, Anchors, Binning\n",
      "\n",
      "winAnchorMultimapNmax           50\n",
      "    int>0: max number of loci anchors are allowed to map to\n",
      "\n",
      "winBinNbits                     16\n",
      "    int>0: =log2(winBin), where winBin is the size of the bin for the windows/clustering, each window will occupy an integer number of bins.\n",
      "\n",
      "winAnchorDistNbins              9\n",
      "    int>0: max number of bins between two anchors that allows aggregation of anchors into one window\n",
      "\n",
      "winFlankNbins                   4\n",
      "    int>0: log2(winFlank), where win Flank is the size of the left and right flanking regions for each window\n",
      "\n",
      "winReadCoverageRelativeMin      0.5\n",
      "    real>=0: minimum relative coverage of the read sequence by the seeds in a window, for STARlong algorithm only.\n",
      "\n",
      "winReadCoverageBasesMin      0\n",
      "    int>0: minimum number of bases covered by the seeds in a window , for STARlong algorithm only.\n",
      "\n",
      "### Chimeric Alignments\n",
      "chimOutType                 Junctions\n",
      "    string(s): type of chimeric output\n",
      "                            Junctions       ... Chimeric.out.junction\n",
      "                            SeparateSAMold  ... output old SAM into separate Chimeric.out.sam file\n",
      "                            WithinBAM       ... output into main aligned BAM files (Aligned.*.bam)\n",
      "                            WithinBAM HardClip  ... (default) hard-clipping in the CIGAR for supplemental chimeric alignments (default if no 2nd word is present)\n",
      "                            WithinBAM SoftClip  ... soft-clipping in the CIGAR for supplemental chimeric alignments\n",
      "\n",
      "chimSegmentMin              0\n",
      "    int>=0: minimum length of chimeric segment length, if ==0, no chimeric output\n",
      "\n",
      "chimScoreMin                0\n",
      "    int>=0: minimum total (summed) score of the chimeric segments\n",
      "\n",
      "chimScoreDropMax            20\n",
      "    int>=0: max drop (difference) of chimeric score (the sum of scores of all chimeric segments) from the read length\n",
      "\n",
      "chimScoreSeparation         10\n",
      "    int>=0: minimum difference (separation) between the best chimeric score and the next one\n",
      "\n",
      "chimScoreJunctionNonGTAG    -1\n",
      "    int: penalty for a non-GT/AG chimeric junction\n",
      "\n",
      "chimJunctionOverhangMin     20\n",
      "    int>=0: minimum overhang for a chimeric junction\n",
      "\n",
      "chimSegmentReadGapMax       0\n",
      "    int>=0: maximum gap in the read sequence between chimeric segments\n",
      "\n",
      "chimFilter                  banGenomicN\n",
      "    string(s): different filters for chimeric alignments\n",
      "                            None ... no filtering\n",
      "                            banGenomicN ... Ns are not allowed in the genome sequence around the chimeric junction\n",
      "\n",
      "chimMainSegmentMultNmax        10\n",
      "    int>=1: maximum number of multi-alignments for the main chimeric segment. =1 will prohibit multimapping main segments.\n",
      "\n",
      "chimMultimapNmax                    0\n",
      "    int>=0: maximum number of chimeric multi-alignments\n",
      "                                0 ... use the old scheme for chimeric detection which only considered unique alignments\n",
      "\n",
      "chimMultimapScoreRange          1\n",
      "    int>=0: the score range for multi-mapping chimeras below the best chimeric score. Only works with --chimMultimapNmax > 1\n",
      "\n",
      "chimNonchimScoreDropMin         20\n",
      "    int>=0: to trigger chimeric detection, the drop in the best non-chimeric alignment score with respect to the read length has to be greater than this value\n",
      "\n",
      "chimOutJunctionFormat           0\n",
      "    int: formatting type for the Chimeric.out.junction file\n",
      "                                0 ... no comment lines/headers\n",
      "                                1 ... comment lines at the end of the file: command line and Nreads: total, unique/multi-mapping\n",
      "\n",
      "### Quantification of Annotations\n",
      "quantMode                   -\n",
      "    string(s): types of quantification requested\n",
      "                            -                ... none\n",
      "                            TranscriptomeSAM ... output SAM/BAM alignments to transcriptome into a separate file\n",
      "                            GeneCounts       ... count reads per gene\n",
      "\n",
      "quantTranscriptomeBAMcompression    1       1\n",
      "    int: -2 to 10  transcriptome BAM compression level\n",
      "                            -2  ... no BAM output\n",
      "                            -1  ... default compression (6?)\n",
      "                             0  ... no compression\n",
      "                             10 ... maximum compression\n",
      "\n",
      "quantTranscriptomeBan       IndelSoftclipSingleend\n",
      "    string: prohibit various alignment type\n",
      "                            IndelSoftclipSingleend  ... prohibit indels, soft clipping and single-end alignments - compatible with RSEM\n",
      "                            Singleend               ... prohibit single-end alignments\n",
      "\n",
      "### 2-pass Mapping\n",
      "twopassMode                 None\n",
      "    string: 2-pass mapping mode.\n",
      "                            None        ... 1-pass mapping\n",
      "                            Basic       ... basic 2-pass mapping, with all 1st pass junctions inserted into the genome indices on the fly\n",
      "\n",
      "twopass1readsN              -1\n",
      "    int: number of reads to process for the 1st step. Use very large number (or default -1) to map all reads in the first step.\n",
      "\n",
      "\n",
      "### WASP parameters\n",
      "waspOutputMode              None\n",
      "    string: WASP allele-specific output type. This is re-implementation of the original WASP mappability filtering by Bryce van de Geijn, Graham McVicker, Yoav Gilad & Jonathan K Pritchard. Please cite the original WASP paper: Nature Methods 12, 10611063 (2015), https://www.nature.com/articles/nmeth.3582 .\n",
      "                            SAMtag      ... add WASP tags to the alignments that pass WASP filtering\n",
      "\n",
      "### STARsolo (single cell RNA-seq) parameters\n",
      "soloType                    None\n",
      "    string(s): type of single-cell RNA-seq\n",
      "                            CB_UMI_Simple   ... (a.k.a. Droplet) one UMI and one Cell Barcode of fixed length in read2, e.g. Drop-seq and 10X Chromium.\n",
      "                            CB_UMI_Complex  ... multiple Cell Barcodes of varying length, one UMI of fixed length and one adapter sequence of fixed length are allowed in read2 only (e.g. inDrop, ddSeq).\n",
      "                            CB_samTagOut    ... output Cell Barcode as CR and/or CB SAm tag. No UMI counting. --readFilesIn cDNA_read1 [cDNA_read2 if paired-end] CellBarcode_read . Requires --outSAMtype BAM Unsorted [and/or SortedByCoordinate]\n",
      "                            SmartSeq        ... Smart-seq: each cell in a separate FASTQ (paired- or single-end), barcodes are corresponding read-groups, no UMI sequences, alignments deduplicated according to alignment start and end (after extending soft-clipped bases)\n",
      "\n",
      "soloCBtype                  Sequence\n",
      "    string: cell barcode type\n",
      "                            Sequence: cell barcode is a sequence (standard option)\n",
      "                            String: cell barcode is an arbitrary string\n",
      "\n",
      "soloCBwhitelist             -\n",
      "    string(s): file(s) with whitelist(s) of cell barcodes. Only --soloType CB_UMI_Complex allows more than one whitelist file.\n",
      "                            None            ... no whitelist: all cell barcodes are allowed\n",
      "\n",
      "soloCBstart                 1\n",
      "    int>0: cell barcode start base\n",
      "\n",
      "soloCBlen                   16\n",
      "    int>0: cell barcode length\n",
      "\n",
      "soloUMIstart                17\n",
      "    int>0: UMI start base\n",
      "\n",
      "soloUMIlen                  10\n",
      "    int>0: UMI length\n",
      "\n",
      "soloBarcodeReadLength       1\n",
      "    int: length of the barcode read\n",
      "                            1   ... equal to sum of soloCBlen+soloUMIlen\n",
      "                            0   ... not defined, do not check\n",
      "\n",
      "soloBarcodeMate             0                            \n",
      "    int: identifies which read mate contains the barcode (CB+UMI) sequence\n",
      "                            0   ... barcode sequence is on separate read, which should always be the last file in the --readFilesIn listed\n",
      "                            1   ... barcode sequence is a part of mate 1\n",
      "                            2   ... barcode sequence is a part of mate 2\n",
      "\n",
      "soloCBposition              -\n",
      "    strings(s)              position of Cell Barcode(s) on the barcode read.\n",
      "                            Presently only works with --soloType CB_UMI_Complex, and barcodes are assumed to be on Read2.\n",
      "                            Format for each barcode: startAnchor_startPosition_endAnchor_endPosition\n",
      "                            start(end)Anchor defines the Anchor Base for the CB: 0: read start; 1: read end; 2: adapter start; 3: adapter end\n",
      "                            start(end)Position is the 0-based position with of the CB start(end) with respect to the Anchor Base\n",
      "                            String for different barcodes are separated by space.\n",
      "                            Example: inDrop (Zilionis et al, Nat. Protocols, 2017):\n",
      "                            --soloCBposition  0_0_2_-1  3_1_3_8\n",
      "\n",
      "soloUMIposition             -\n",
      "    string                  position of the UMI on the barcode read, same as soloCBposition\n",
      "                            Example: inDrop (Zilionis et al, Nat. Protocols, 2017):\n",
      "                            --soloCBposition  3_9_3_14\n",
      "\n",
      "soloAdapterSequence         -\n",
      "    string:                 adapter sequence to anchor barcodes. Only one adapter sequence is allowed.\n",
      "\n",
      "soloAdapterMismatchesNmax   1\n",
      "    int>0:                  maximum number of mismatches allowed in adapter sequence.\n",
      "\n",
      "soloCBmatchWLtype           1MM_multi\n",
      "    string:                 matching the Cell Barcodes to the WhiteList\n",
      "                            Exact                           ... only exact matches allowed\n",
      "                            1MM                             ... only one match in whitelist with 1 mismatched base allowed. Allowed CBs have to have at least one read with exact match.\n",
      "                            1MM_multi                       ... multiple matches in whitelist with 1 mismatched base allowed, posterior probability calculation is used choose one of the matches.\n",
      "                                                                Allowed CBs have to have at least one read with exact match. This option matches best with CellRanger 2.2.0\n",
      "                            1MM_multi_pseudocounts          ... same as 1MM_Multi, but pseudocounts of 1 are added to all whitelist barcodes.\n",
      "                            1MM_multi_Nbase_pseudocounts    ... same as 1MM_multi_pseudocounts, multimatching to WL is allowed for CBs with N-bases. This option matches best with CellRanger >= 3.0.0\n",
      "                            EditDist_2                    ... allow up to edit distance of 3 fpr each of the barcodes. May include one deletion + one insertion. Only works with --soloType CB_UMI_Complex. Matches to multiple passlist barcdoes are not allowed. Similar to ParseBio Split-seq pipeline. \n",
      "\n",
      "soloInputSAMattrBarcodeSeq  -\n",
      "    string(s):              when inputting reads from a SAM file (--readsFileType SAM SE/PE), these SAM attributes mark the barcode sequence (in proper order).\n",
      "                            For instance, for 10X CellRanger or STARsolo BAMs, use --soloInputSAMattrBarcodeSeq CR UR .\n",
      "                            This parameter is required when running STARsolo with input from SAM. \n",
      "    \n",
      "soloInputSAMattrBarcodeQual  -\n",
      "    string(s):              when inputting reads from a SAM file (--readsFileType SAM SE/PE), these SAM attributes mark the barcode qualities (in proper order).\n",
      "                            For instance, for 10X CellRanger or STARsolo BAMs, use --soloInputSAMattrBarcodeQual CY UY .\n",
      "                            If this parameter is '-' (default), the quality 'H' will be assigned to all bases.\n",
      "\n",
      "soloStrand                  Forward\n",
      "    string: strandedness of the solo libraries:\n",
      "                            Unstranded  ... no strand information\n",
      "                            Forward     ... read strand same as the original RNA molecule\n",
      "                            Reverse     ... read strand opposite to the original RNA molecule\n",
      "\n",
      "soloFeatures                Gene\n",
      "    string(s): genomic features for which the UMI counts per Cell Barcode are collected\n",
      "                            Gene            ... genes: reads match the gene transcript\n",
      "                            SJ              ... splice junctions: reported in SJ.out.tab\n",
      "                            GeneFull        ... full gene (pre-mRNA): count all reads overlapping genes' exons and introns\n",
      "                            GeneFull_ExonOverIntron ... full gene (pre-mRNA): count all reads overlapping genes' exons and introns: prioritize 100% overlap with exons\n",
      "                            GeneFull_Ex50pAS        ... full gene (pre-RNA): count all reads overlapping genes' exons and introns: prioritize >50% overlap with exons. Do not count reads with 100% exonic overlap in the antisense direction.\n",
      "\n",
      "#####UnderDevelopment_begin : not supported - do not use\n",
      "                            Transcript3p    ... quantification of transcript for 3' protocols\n",
      "#####UnderDevelopment_end\n",
      "\n",
      "soloMultiMappers            Unique\n",
      "    string(s): counting method for reads mapping to multiple genes           \n",
      "                            Unique     ... count only reads that map to unique genes\n",
      "                            Uniform    ... uniformly distribute multi-genic UMIs to all genes\n",
      "                            Rescue     ... distribute UMIs proportionally to unique+uniform counts (~ first iteration of EM)\n",
      "                            PropUnique ... distribute UMIs proportionally to unique mappers, if present, and uniformly if not.\n",
      "                            EM         ... multi-gene UMIs are distributed using Expectation Maximization algorithm\n",
      "\n",
      "soloUMIdedup                1MM_All\n",
      "    string(s):              type of UMI deduplication (collapsing) algorithm\n",
      "                            1MM_All                     ... all UMIs with 1 mismatch distance to each other are collapsed (i.e. counted once).\n",
      "                            1MM_Directional_UMItools    ... follows the \"directional\" method from the UMI-tools by Smith, Heger and Sudbery (Genome Research 2017).\n",
      "                            1MM_Directional             ... same as 1MM_Directional_UMItools, but with more stringent criteria for duplicate UMIs\n",
      "                            Exact                       ... only exactly matching UMIs are collapsed.\n",
      "                            NoDedup                     ... no deduplication of UMIs, count all reads.\n",
      "                            1MM_CR                      ... CellRanger2-4 algorithm for 1MM UMI collapsing.\n",
      "\n",
      "soloUMIfiltering            -\n",
      "    string(s)               type of UMI filtering (for reads uniquely mapping to genes)\n",
      "                            -                  ... basic filtering: remove UMIs with N and homopolymers (similar to CellRanger 2.2.0).\n",
      "                            MultiGeneUMI       ... basic + remove lower-count UMIs that map to more than one gene.\n",
      "                            MultiGeneUMI_All   ... basic + remove all UMIs that map to more than one gene.\n",
      "                            MultiGeneUMI_CR    ... basic + remove lower-count UMIs that map to more than one gene, matching CellRanger > 3.0.0 .\n",
      "                                                   Only works with --soloUMIdedup 1MM_CR\n",
      "                                                \n",
      "soloOutFileNames            Solo.out/          features.tsv barcodes.tsv        matrix.mtx\n",
      "    string(s)               file names for STARsolo output:\n",
      "                            file_name_prefix   gene_names   barcode_sequences   cell_feature_count_matrix\n",
      "\n",
      "soloCellFilter              CellRanger2.2 3000 0.99 10\n",
      "    string(s):              cell filtering type and parameters\n",
      "                            None            ... do not output filtered cells\n",
      "                            TopCells        ... only report top cells by UMI count, followed by the exact number of cells\n",
      "                            CellRanger2.2   ... simple filtering of CellRanger 2.2. \n",
      "                                                Can be followed by numbers: number of expected cells, robust maximum percentile for UMI count, maximum to minimum ratio for UMI count\n",
      "                                                The harcoded values are from CellRanger: nExpectedCells=3000;  maxPercentile=0.99;  maxMinRatio=10\n",
      "                            EmptyDrops_CR   ... EmptyDrops filtering in CellRanger flavor. Please cite the original EmptyDrops paper: A.T.L Lun et al, Genome Biology, 20, 63 (2019): https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y\n",
      "                                                Can be followed by 10 numeric parameters:  nExpectedCells   maxPercentile   maxMinRatio   indMin   indMax   umiMin   umiMinFracMedian   candMaxN   FDR   simN \n",
      "                                                The harcoded values are from CellRanger:             3000            0.99            10    45000    90000      500               0.01      20000  0.01  10000\n",
      "\n",
      "soloOutFormatFeaturesGeneField3    \"Gene Expression\"\n",
      "    string(s):                field 3 in the Gene features.tsv file. If \"-\", then no 3rd field is output.\n",
      "\n",
      "soloCellReadStats           None\n",
      "    string:                 Output reads statistics for each CB\n",
      "                            Standard    ... standard output\n",
      "\n",
      "#####UnderDevelopment_begin : not supported - do not use\n",
      "soloClusterCBfile           -\n",
      "    string:                 file containing the cluster information for cell barcodes, two columns: CB cluster_index. Only used with --soloFeatures Transcript3p\n",
      "#####UnderDevelopment_end\n"
     ]
    }
   ],
   "source": [
    "!STAR -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "3a454761-a36e-44d6-bf0c-a3366ca4f192",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "cf59cb39-a625-4fe8-99eb-13f3761860b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "README.Rmd  README_files  example  input    sortmerna_db\n",
      "README.md   annotation\t  genome   results  star_index\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "da70c8f4-dc07-49b8-8914-5d6a9cf9ba3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow/results\n"
     ]
    }
   ],
   "source": [
    "%cd results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "b3f2d084-4f00-4976-8d7e-09e749b87dbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1_initial_qc\t  3_rRNA\t       5_final_counts\n",
      "2_trimmed_output  4_aligned_sequences  6_multiQC\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "a0792ae4-bb85-450e-a205-f5941b89ebb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow/results/2_trimmed_output\n"
     ]
    }
   ],
   "source": [
    "%cd 2_trimmed_output/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "f5b18407-9b82-450c-bde0-754e2bc1927e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001_trimmed.fq\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001_trimmed_fastqc.zip\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001.fastq_trimming_report.txt\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001_trimmed.fq\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001_trimmed_fastqc.html\n",
      "JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001_trimmed_fastqc.zip\n",
      "sortmerna_db\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "37cb7747-c29e-46a5-be38-88ccd9545338",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow/results/2_trimmed_output\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "ab357254-332d-457c-81c8-be98d52608ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow/results\n"
     ]
    }
   ],
   "source": [
    "%cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "876231a6-b039-45a3-b733-b47cc61ee342",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "%cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16e4e890-f222-4763-bf97-fede0151ebaf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_trimmed.fq --runThreadN 4 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 12:40:06 ..... started STAR run\n",
      "Jan 22 12:40:06 ..... loading genome\n"
     ]
    }
   ],
   "source": [
    "# ###--------------------------------------- Test Run: Runs one pair of read-----------------------------------------------------------\n",
    "# %%bash\n",
    "\n",
    "# # Set the path to the STAR genome index\n",
    "# genome_dir=\"/mnt/e/new_workflow/star_index/\"  # Replace with the actual path\n",
    "\n",
    "# # Set the path to the output directory\n",
    "# output_dir=\"/mnt/e/new_workflow/results/4_aligned_sequences/\"\n",
    "\n",
    "# # Run STAR for the specific pair of reads\n",
    "# STAR \\\n",
    "#     --genomeDir $genome_dir \\\n",
    "#     --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_trimmed.fq \\\n",
    "#                  /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_trimmed.fq \\\n",
    "#     --runThreadN 4 \\\n",
    "#     --outSAMtype BAM SortedByCoordinate \\\n",
    "#     --quantMode GeneCounts \\\n",
    "#     --outFileNamePrefix \"${output_dir}/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b3befa60-122c-4253-992a-ababe4953bd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/raotoo/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3d2265de-132c-4164-91b1-63df0fa29012",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the path to the STAR genome index\n",
    "genome_dir=\"/mnt/e/new_workflow/star_index/\"\n",
    "\n",
    "# Set the path to the output directory\n",
    "output_dir=\"/mnt/e/new_workflow/results/4_aligned_sequences/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a5e5d9dd-fcb5-40cf-ac9f-6902a4985b71",
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm -r /mnt/e/new_workflow/results/3_star_output/*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7dd78e98-29ff-4f57-8220-cf997f76c060",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!find /mnt/e/new_workflow/results/4_aligned_sequences/ -type f ! -name 'aligned_bam' ! -name 'aligned_log' -exec rm {} +"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1a60f99c-8ccf-4042-90fe-75193eef45f6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_trimmed.fq --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 16:53:19 ..... started STAR run\n",
      "Jan 22 16:53:19 ..... loading genome\n",
      "Jan 22 16:54:01 ..... started mapping\n",
      "Jan 22 17:09:21 ..... finished mapping\n",
      "Jan 22 17:09:24 ..... started sorting BAM\n",
      "Jan 22 17:09:24 ..... finished successfully\n",
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001_trimmed.fq --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 17:09:26 ..... started STAR run\n",
      "Jan 22 17:09:26 ..... loading genome\n",
      "Jan 22 17:12:02 ..... started mapping\n",
      "Jan 22 17:25:18 ..... finished mapping\n",
      "Jan 22 17:25:20 ..... started sorting BAM\n",
      "Jan 22 17:25:21 ..... finished successfully\n",
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001_trimmed.fq --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 17:25:23 ..... started STAR run\n",
      "Jan 22 17:25:23 ..... loading genome\n",
      "Jan 22 17:28:00 ..... started mapping\n",
      "Jan 22 17:42:52 ..... finished mapping\n",
      "Jan 22 17:42:55 ..... started sorting BAM\n",
      "Jan 22 17:42:56 ..... finished successfully\n",
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001_trimmed.fq --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 17:42:57 ..... started STAR run\n",
      "Jan 22 17:42:57 ..... loading genome\n",
      "Jan 22 17:45:34 ..... started mapping\n",
      "Jan 22 17:59:58 ..... finished mapping\n",
      "Jan 22 18:00:00 ..... started sorting BAM\n",
      "Jan 22 18:00:01 ..... finished successfully\n",
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 18:00:03 ..... started STAR run\n",
      "Jan 22 18:00:03 ..... loading genome\n",
      "Jan 22 18:02:39 ..... started mapping\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "EXITING because of FATAL ERROR in reads input: quality string length is not equal to sequence length\n",
      "@VH01214:38:AACNY7YM5:1:2611:27680:56073\n",
      "+\n",
      "CCCCCCCCCCCCCCCCCCCCCCCCCC;CCCCCCCCCCCCCCC;CCCCCCCCCCCCCCCCC-CCCCCCCCCCCCCCCCCCCCCCCC\n",
      "SOLUTION: fix your fastq file\n",
      "\n",
      "Jan 22 18:17:11 ...... FATAL ERROR, exiting\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001_trimmed.fq --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 18:17:13 ..... started STAR run\n",
      "Jan 22 18:17:13 ..... loading genome\n",
      "Jan 22 18:19:50 ..... started mapping\n",
      "Jan 22 18:37:09 ..... finished mapping\n",
      "Jan 22 18:37:12 ..... started sorting BAM\n",
      "Jan 22 18:37:14 ..... finished successfully\n",
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001_trimmed.fq --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 18:37:16 ..... started STAR run\n",
      "Jan 22 18:37:16 ..... loading genome\n",
      "Jan 22 18:39:53 ..... started mapping\n",
      "Jan 22 18:57:04 ..... finished mapping\n",
      "Jan 22 18:57:07 ..... started sorting BAM\n",
      "Jan 22 18:57:09 ..... finished successfully\n",
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001_trimmed.fq --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 18:57:10 ..... started STAR run\n",
      "Jan 22 18:57:10 ..... loading genome\n",
      "Jan 22 18:59:47 ..... started mapping\n",
      "Jan 22 19:14:37 ..... finished mapping\n",
      "Jan 22 19:14:40 ..... started sorting BAM\n",
      "Jan 22 19:14:42 ..... finished successfully\n",
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001_trimmed.fq --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 19:14:43 ..... started STAR run\n",
      "Jan 22 19:14:43 ..... loading genome\n",
      "Jan 22 19:17:21 ..... started mapping\n",
      "Jan 22 19:27:44 ..... finished mapping\n",
      "Jan 22 19:27:47 ..... started sorting BAM\n",
      "Jan 22 19:27:50 ..... finished successfully\n",
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001_trimmed.fq /mnt/e/new_workflow/results/2_trimmed_output/JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001_trimmed.fq --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/results/4_aligned_sequences//JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 22 19:27:52 ..... started STAR run\n",
      "Jan 22 19:27:52 ..... loading genome\n",
      "Jan 22 19:30:29 ..... started mapping\n",
      "Jan 22 19:37:08 ..... finished mapping\n",
      "Jan 22 19:37:11 ..... started sorting BAM\n",
      "Jan 22 19:37:12 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "# Set the path to the STAR genome index\n",
    "genome_dir = \"/mnt/e/new_workflow/star_index/\"\n",
    "\n",
    "# Set the path to the output directory\n",
    "output_dir = \"/mnt/e/new_workflow/results/4_aligned_sequences/\"\n",
    "\n",
    "# List of file names\n",
    "file_names = [\n",
    "    \"JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R1_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_R2_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R1_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_R2_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R1_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_R2_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R1_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_R2_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R1_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_R2_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R1_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_R2_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R1_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_R2_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R1_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_R2_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R1_001_trimmed.fq\",\n",
    "    \"JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_R2_001_trimmed.fq\"\n",
    "]\n",
    "\n",
    "\n",
    "# Run STAR for each pair of reads\n",
    "for i in range(0, len(file_names), 2):\n",
    "    file_R1 = file_names[i]\n",
    "    file_R2 = file_names[i + 1]\n",
    "\n",
    "    # Extract the sample name from the file name\n",
    "    sample_name = file_R1.split(\"_R1_001_trimmed.fq\")[0]\n",
    "\n",
    "    # Run STAR\n",
    "    cmd = f\"STAR --genomeDir {genome_dir} --readFilesIn /mnt/e/new_workflow/results/2_trimmed_output/{file_R1} /mnt/e/new_workflow/results/2_trimmed_output/{file_R2} --runThreadN 12 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix {output_dir}/{sample_name}_STAR_\"\n",
    "    subprocess.run(cmd, shell=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b0e8ee99-04d2-4c95-9609-8239cd827b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gunzip rerun/*.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "9b9127c3-012e-4857-a303-a2e640dbc979",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " USAGE:\n",
      "\n",
      "trim_galore [options] <filename(s)>\n",
      "\n",
      "\n",
      "-h/--help               Print this help message and exits.\n",
      "\n",
      "-v/--version            Print the version information and exits.\n",
      "\n",
      "-q/--quality <INT>      Trim low-quality ends from reads in addition to adapter removal. For\n",
      "                        RRBS samples, quality trimming will be performed first, and adapter\n",
      "                        trimming is carried in a second round. Other files are quality and adapter\n",
      "                        trimmed in a single pass. The algorithm is the same as the one used by BWA\n",
      "                        (Subtract INT from all qualities; compute partial sums from all indices\n",
      "                        to the end of the sequence; cut sequence at the index at which the sum is\n",
      "                        minimal). Default Phred score: 20.\n",
      "\n",
      "--phred33               Instructs Cutadapt to use ASCII+33 quality scores as Phred scores\n",
      "                        (Sanger/Illumina 1.9+ encoding) for quality trimming. Default: ON.\n",
      "\n",
      "--phred64               Instructs Cutadapt to use ASCII+64 quality scores as Phred scores\n",
      "                        (Illumina 1.5 encoding) for quality trimming.\n",
      "\n",
      "--fastqc                Run FastQC in the default mode on the FastQ file once trimming is complete.\n",
      "\n",
      "--fastqc_args \"<ARGS>\"  Passes extra arguments to FastQC. If more than one argument is to be passed\n",
      "                        to FastQC they must be in the form \"arg1 arg2 etc.\". An example would be:\n",
      "                        --fastqc_args \"--nogroup --outdir /home/\". Passing extra arguments will\n",
      "                        automatically invoke FastQC, so --fastqc does not have to be specified\n",
      "                        separately.\n",
      "\n",
      "-a/--adapter <STRING>   Adapter sequence to be trimmed. If not specified explicitly, Trim Galore will\n",
      "                        try to auto-detect whether the Illumina universal, Nextera transposase or Illumina\n",
      "                        small RNA adapter sequence was used. Also see '--illumina', '--nextera' and\n",
      "                        '--small_rna'. If no adapter can be detected within the first 1 million sequences\n",
      "                        of the first file specified or if there is a tie between several adapter sequences,\n",
      "                        Trim Galore defaults to '--illumina' (as long as the Illumina adapter was one of the\n",
      "                        options, else '--nextera' is the default). A single base\n",
      "                        may also be given as e.g. -a A{10}, to be expanded to -a AAAAAAAAAA.\n",
      "\n",
      "                        At a special request, multiple adapters can also be specified like so:\n",
      "                        -a  \" AGCTCCCG -a TTTCATTATAT -a TTTATTCGGATTTAT\"\n",
      "                        -a2 \" AGCTAGCG -a TCTCTTATAT -a TTTCGGATTTAT\", or so:\n",
      "                        -a \"file:../multiple_adapters.fa\"\n",
      "                        -a2 \"file:../different_adapters.fa\"\n",
      "                        Potentially in conjucntion with the parameter \"-n 3\" to trim all adapters. Please note\n",
      "                        that this is NOT needed for standard trimming! \n",
      "                        More Information here: https://github.com/FelixKrueger/TrimGalore/issues/86\n",
      "\n",
      "-a2/--adapter2 <STRING> Optional adapter sequence to be trimmed off read 2 of paired-end files. This\n",
      "                        option requires '--paired' to be specified as well. If the libraries to be trimmed\n",
      "                        are smallRNA then a2 will be set to the Illumina small RNA 5' adapter automatically\n",
      "                        (GATCGTCGGACT). A single base may also be given as e.g. -a2 A{10}, to be expanded\n",
      "                        to -a2 AAAAAAAAAA.\n",
      "\n",
      "--illumina              Adapter sequence to be trimmed is the first 13bp of the Illumina universal adapter\n",
      "                        'AGATCGGAAGAGC' instead of the default auto-detection of adapter sequence.\n",
      "\n",
      "--stranded_illumina     Adapter sequence to be trimmed is the first 13bp of the Illumina stranded mRNA or Total\n",
      "                        RNA adapter 'ACTGTCTCTTATA' instead of the default auto-detection of adapter sequence. \n",
      "                        Note that this sequence resembles the Nextera sequence with an additional A from A-tailing.\n",
      "                        Please also see https://github.com/FelixKrueger/TrimGalore/issues/127 or \n",
      "                        https://support.illumina.com/bulletins/2020/06/trimming-t-overhang-options-for-the-illumina-rna-library-prep-wo.html\n",
      "                        for further information. This sequence is currently NOT included in the adapter auto-detection.\n",
      "\n",
      "--nextera               Adapter sequence to be trimmed is the first 12bp of the Nextera adapter\n",
      "                        'CTGTCTCTTATA' instead of the default auto-detection of adapter sequence.\n",
      "\n",
      "--small_rna             Adapter sequence to be trimmed is the first 12bp of the Illumina Small RNA 3' Adapter\n",
      "                        'TGGAATTCTCGG' instead of the default auto-detection of adapter sequence. Selecting\n",
      "                        to trim smallRNA adapters will also lower the --length value to 18bp. If the smallRNA\n",
      "                        libraries are paired-end then a2 will be set to the Illumina small RNA 5' adapter\n",
      "                        automatically (GATCGTCGGACT) unless -a 2 had been defined explicitly.\n",
      "\n",
      "--consider_already_trimmed <INT>     During adapter auto-detection, the limit set by <INT> allows the user to \n",
      "                        set a threshold up to which the file is considered already adapter-trimmed. If no adapter\n",
      "                        sequence exceeds this threshold, no additional adapter trimming will be performed (technically,\n",
      "                        the adapter is set to '-a X'). Quality trimming is still performed as usual.\n",
      "                        Default: NOT SELECTED (i.e. normal auto-detection precedence rules apply).                     \n",
      "\n",
      "--max_length <INT>      Discard reads that are longer than <INT> bp after trimming. This is only advised for\n",
      "                        smallRNA sequencing to remove non-small RNA sequences.\n",
      "\n",
      "\n",
      "--stringency <INT>      Overlap with adapter sequence required to trim a sequence. Defaults to a\n",
      "                        very stringent setting of 1, i.e. even a single bp of overlapping sequence\n",
      "                        will be trimmed off from the 3' end of any read.\n",
      "\n",
      "-e <ERROR RATE>         Maximum allowed error rate (no. of errors divided by the length of the matching\n",
      "                        region) (default: 0.1)\n",
      "\n",
      "--gzip                  Compress the output file with GZIP. If the input files are GZIP-compressed\n",
      "                        the output files will automatically be GZIP compressed as well. As of v0.2.8 the\n",
      "                        compression will take place on the fly.\n",
      "\n",
      "--dont_gzip             Output files won't be compressed with GZIP. This option overrides --gzip.\n",
      "\n",
      "--length <INT>          Discard reads that became shorter than length INT because of either\n",
      "                        quality or adapter trimming. A value of '0' effectively disables\n",
      "                        this behaviour. Default: 20 bp.\n",
      "\n",
      "                        For paired-end files, both reads of a read-pair need to be longer than\n",
      "                        <INT> bp to be printed out to validated paired-end files (see option --paired).\n",
      "                        If only one read became too short there is the possibility of keeping such\n",
      "                        unpaired single-end reads (see --retain_unpaired). Default pair-cutoff: 20 bp.\n",
      "\n",
      "--max_n COUNT           The total number of Ns a read may contain before it will be removed altogether.\n",
      "                        In a paired-end setting, either read exceeding this limit will result in the entire\n",
      "                        pair being removed from the trimmed output files. If COUNT is a number between 0 and 1,\n",
      "                        it is interpreted as a fraction of the read length.\n",
      "\n",
      "--trim-n                Removes Ns from either side of the read. This option does currently not work in RRBS mode.\n",
      "\n",
      "-o/--output_dir <DIR>   If specified all output will be written to this directory instead of the current\n",
      "                        directory. If the directory doesn't exist it will be created for you.\n",
      "\n",
      "--no_report_file        If specified no report file will be generated.\n",
      "\n",
      "--suppress_warn         If specified any output to STDOUT or STDERR will be suppressed.\n",
      "\n",
      "--clip_R1 <int>         Instructs Trim Galore to remove <int> bp from the 5' end of read 1 (or single-end\n",
      "                        reads). This may be useful if the qualities were very poor, or if there is some\n",
      "                        sort of unwanted bias at the 5' end. Default: OFF.\n",
      "\n",
      "--clip_R2 <int>         Instructs Trim Galore to remove <int> bp from the 5' end of read 2 (paired-end reads\n",
      "                        only). This may be useful if the qualities were very poor, or if there is some sort\n",
      "                        of unwanted bias at the 5' end. For paired-end BS-Seq, it is recommended to remove\n",
      "                        the first few bp because the end-repair reaction may introduce a bias towards low\n",
      "                        methylation. Please refer to the M-bias plot section in the Bismark User Guide for\n",
      "                        some examples. Default: OFF.\n",
      "\n",
      "--three_prime_clip_R1 <int>     Instructs Trim Galore to remove <int> bp from the 3' end of read 1 (or single-end\n",
      "                        reads) AFTER adapter/quality trimming has been performed. This may remove some unwanted\n",
      "                        bias from the 3' end that is not directly related to adapter sequence or basecall quality.\n",
      "                        Default: OFF.\n",
      "\n",
      "--three_prime_clip_R2 <int>     Instructs Trim Galore to remove <int> bp from the 3' end of read 2 AFTER\n",
      "                        adapter/quality trimming has been performed. This may remove some unwanted bias from\n",
      "                        the 3' end that is not directly related to adapter sequence or basecall quality.\n",
      "                        Default: OFF.\n",
      "\n",
      "--2colour/--nextseq INT This enables the option '--nextseq-trim=3'CUTOFF' within Cutadapt, which will set a quality\n",
      "                        cutoff (that is normally given with -q instead), but qualities of G bases are ignored.\n",
      "                        This trimming is in common for the NextSeq- and NovaSeq-platforms, where basecalls without\n",
      "                        any signal are called as high-quality G bases. This is mutually exlusive with '-q INT'.\n",
      "\n",
      "\n",
      "--path_to_cutadapt </path/to/cutadapt>     You may use this option to specify a path to the Cutadapt executable,\n",
      "                        e.g. /my/home/cutadapt-1.7.1/bin/cutadapt. Else it is assumed that Cutadapt is in\n",
      "                        the PATH.\n",
      "\n",
      "--basename <PREFERRED_NAME>\tUse PREFERRED_NAME as the basename for output files, instead of deriving the filenames from\n",
      "                        the input files. Single-end data would be called PREFERRED_NAME_trimmed.fq(.gz), or\n",
      "                        PREFERRED_NAME_val_1.fq(.gz) and PREFERRED_NAME_val_2.fq(.gz) for paired-end data. --basename\n",
      "                        only works when 1 file (single-end) or 2 files (paired-end) are specified, but not for longer lists.\n",
      "\n",
      "-j/--cores INT          Number of cores to be used for trimming [default: 1]. For Cutadapt to work with multiple cores, it\n",
      "                        requires Python 3 as well as parallel gzip (pigz) installed on the system. Trim Galore attempts to detect\n",
      "                        the version of Python used by calling Cutadapt. If Python 2 is detected, --cores is set to 1. If the Python\n",
      "                        version cannot be detected, Python 3 is assumed and we let Cutadapt handle potential issues itself.\n",
      "                        \n",
      "                        If pigz cannot be detected on your system, Trim Galore reverts to using gzip compression. Please note\n",
      "                        that gzip compression will slow down multi-core processes so much that it is hardly worthwhile, please \n",
      "                        see: https://github.com/FelixKrueger/TrimGalore/issues/16#issuecomment-458557103 for more info).\n",
      "\t\t\t\t\t\t\n",
      "                        Actual core usage: It should be mentioned that the actual number of cores used is a little convoluted.\n",
      "                        Assuming that Python 3 is used and pigz is installed, --cores 2 would use 2 cores to read the input\n",
      "                        (probably not at a high usage though), 2 cores to write to the output (at moderately high usage), and \n",
      "                        2 cores for Cutadapt itself + 2 additional cores for Cutadapt (not sure what they are used for) + 1 core\n",
      "                        for Trim Galore itself. So this can be up to 9 cores, even though most of them won't be used at 100% for\n",
      "                        most of the time. Paired-end processing uses twice as many cores for the validation (= writing out) step.\n",
      "                        --cores 4 would then be: 4 (read) + 4 (write) + 4 (Cutadapt) + 2 (extra Cutadapt) +\t1 (Trim Galore) = 15.\n",
      "\n",
      "                        It seems that --cores 4 could be a sweet spot, anything above has diminishing returns.\n",
      "\t\t\t\n",
      "\n",
      "\n",
      "SPECIFIC TRIMMING - without adapter/quality trimming\n",
      "\n",
      "--hardtrim5 <int>       Instead of performing adapter-/quality trimming, this option will simply hard-trim sequences\n",
      "                        to <int> bp at the 5'-end. Once hard-trimming of files is complete, Trim Galore will exit.\n",
      "                        Hard-trimmed output files will end in .<int>_5prime.fq(.gz). Here is an example:\n",
      "\n",
      "                        before:         CCTAAGGAAACAAGTACACTCCACACATGCATAAAGGAAATCAAATGTTATTTTTAAGAAAATGGAAAAT\n",
      "                        --hardtrim5 20: CCTAAGGAAACAAGTACACT\n",
      "\n",
      "--hardtrim3 <int>       Instead of performing adapter-/quality trimming, this option will simply hard-trim sequences\n",
      "                        to <int> bp at the 3'-end. Once hard-trimming of files is complete, Trim Galore will exit.\n",
      "                        Hard-trimmed output files will end in .<int>_3prime.fq(.gz). Here is an example:\n",
      "\n",
      "                        before:         CCTAAGGAAACAAGTACACTCCACACATGCATAAAGGAAATCAAATGTTATTTTTAAGAAAATGGAAAAT\n",
      "                        --hardtrim3 20:                                                   TTTTTAAGAAAATGGAAAAT\n",
      "\n",
      "--clock                 In this mode, reads are trimmed in a specific way that is currently used for the Mouse\n",
      "                        Epigenetic Clock (see here: Multi-tissue DNA methylation age predictor in mouse, Stubbs et al.,\n",
      "                        Genome Biology, 2017 18:68 https://doi.org/10.1186/s13059-017-1203-5). Following this, Trim Galore\n",
      "                        will exit.\n",
      "\n",
      "                        In it's current implementation, the dual-UMI RRBS reads come in the following format:\n",
      "\n",
      "                        Read 1  5' UUUUUUUU CAGTA FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF TACTG UUUUUUUU 3'\n",
      "                        Read 2  3' UUUUUUUU GTCAT FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF ATGAC UUUUUUUU 5'\n",
      "\n",
      "                        Where UUUUUUUU is a random 8-mer unique molecular identifier (UMI), CAGTA is a constant region,\n",
      "                        and FFFFFFF... is the actual RRBS-Fragment to be sequenced. The UMIs for Read 1 (R1) and\n",
      "                        Read 2 (R2), as well as the fixed sequences (F1 or F2), are written into the read ID and\n",
      "                        removed from the actual sequence. Here is an example:\n",
      "\n",
      "                        R1: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 1:N:0: CGATGTTT\n",
      "                            ATCTAGTTCAGTACGGTGTTTTCGAATTAGAAAAATATGTATAGAGGAAATAGATATAAAGGCGTATTCGTTATTG\n",
      "                        R2: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 3:N:0: CGATGTTT\n",
      "                            CAATTTTGCAGTACAAAAATAATACCTCCTCTATTTATCCAAAATCACAAAAAACCACCCACTTAACTTTCCCTAA\n",
      "\n",
      "                        R1: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 1:N:0: CGATGTTT:R1:ATCTAGTT:R2:CAATTTTG:F1:CAGT:F2:CAGT\n",
      "                                         CGGTGTTTTCGAATTAGAAAAATATGTATAGAGGAAATAGATATAAAGGCGTATTCGTTATTG\n",
      "                        R2: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 3:N:0: CGATGTTT:R1:ATCTAGTT:R2:CAATTTTG:F1:CAGT:F2:CAGT\n",
      "                                         CAAAAATAATACCTCCTCTATTTATCCAAAATCACAAAAAACCACCCACTTAACTTTCCCTAA\n",
      "\n",
      "                        Following clock trimming, the resulting files (.clock_UMI.R1.fq(.gz) and .clock_UMI.R2.fq(.gz))\n",
      "                        should be adapter- and quality trimmed with Trim Galore as usual. In addition, reads need to be trimmed\n",
      "                        by 15bp from their 3' end to get rid of potential UMI and fixed sequences. The command is:\n",
      "\n",
      "                        trim_galore --paired --three_prime_clip_R1 15 --three_prime_clip_R2 15 *.clock_UMI.R1.fq.gz *.clock_UMI.R2.fq.gz\n",
      "\n",
      "                        Following this, reads should be aligned with Bismark and deduplicated with UmiBam\n",
      "                        in '--dual_index' mode (see here: https://github.com/FelixKrueger/Umi-Grinder). UmiBam recognises\n",
      "                        the UMIs within this pattern: R1:(ATCTAGTT):R2:(CAATTTTG): as (UMI R1) and (UMI R2).\n",
      "\n",
      "--polyA                 This is a new, still experimental, trimming mode to identify and remove poly-A tails from sequences.\n",
      "                        When --polyA is selected, Trim Galore attempts to identify from the first supplied sample whether\n",
      "                        sequences contain more often a stretch of either 'AAAAAAAAAA' or 'TTTTTTTTTT'. This determines\n",
      "                        if Read 1 of a paired-end end file, or single-end files, are trimmed for PolyA or PolyT. In case of\n",
      "                        paired-end sequencing, Read2 is trimmed for the complementary base from the start of the reads. The\n",
      "                        auto-detection uses a default of A{20} for Read1 (3'-end trimming) and T{150} for Read2 (5'-end trimming).\n",
      "                        These values may be changed manually using the options -a and -a2.\n",
      "\n",
      "                        In addition to trimming the sequences, white spaces are replaced with _ and it records in the read ID\n",
      "                        how many bases were trimmed so it can later be used to identify PolyA trimmed sequences. This is currently done\n",
      "                        by writing tags to both the start (\"32:A:\") and end (\"_PolyA:32\") of the reads in the following example:\n",
      "\n",
      "                        @READ-ID:1:1102:22039:36996 1:N:0:CCTAATCC\n",
      "                        GCCTAAGGAAACAAGTACACTCCACACATGCATAAAGGAAATCAAATGTTATTTTTAAGAAAATGGAAAATAAAAACTTTATAAACACCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      "\n",
      "                        @32:A:READ-ID:1:1102:22039:36996_1:N:0:CCTAATCC_PolyA:32\n",
      "                        GCCTAAGGAAACAAGTACACTCCACACATGCATAAAGGAAATCAAATGTTATTTTTAAGAAAATGGAAAATAAAAACTTTATAAACACC\n",
      "\n",
      "                        PLEASE NOTE: The poly-A trimming mode expects that sequences were both adapter and quality trimmed\n",
      "                        before looking for Poly-A tails, and it is the user's responsibility to carry out an initial round of\n",
      "                        trimming. The following sequence:\n",
      " \n",
      "                        1) trim_galore file.fastq.gz\n",
      "                        2) trim_galore --polyA file_trimmed.fq.gz\n",
      "                        3) zcat file_trimmed_trimmed.fq.gz | grep -A 3 PolyA | grep -v ^-- > PolyA_trimmed.fastq\n",
      "\n",
      "                        Will 1) trim qualities and Illumina adapter contamination, 2) find and remove PolyA contamination.\n",
      "                        Finally, if desired, 3) will specifically find PolyA trimmed sequences to a specific FastQ file of your choice.\n",
      "\n",
      "--implicon              This is a special mode of operation for paired-end data, such as required for the IMPLICON method, where a UMI sequence\n",
      "                        is getting transferred from the start of Read 2 to the readID of both reads. Following this, Trim Galore will exit.\n",
      "\n",
      "                        In it's current implementation, the UMI carrying reads come in the following format:\n",
      "\n",
      "                        Read 1  5' FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF 3'\n",
      "                        Read 2  3' UUUUUUUUFFFFFFFFFFFFFFFFFFFFFFFFFFFF 5'\n",
      "\n",
      "                        Where UUUUUUUU is a random 8-mer unique molecular identifier (UMI) and FFFFFFF... is the actual fragment to be\n",
      "                        sequenced. The UMI of Read 2 (R2) is written into the read ID of both reads and removed from the actual sequence.\n",
      "                        Here is an example:\n",
      "\n",
      "                        R1: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 1:N:0: CGATGTTT\n",
      "                            ATCTAGTTCAGTACGGTGTTTTCGAATTAGAAAAATATGTATAGAGGAAATAGATATAAAGGCGTATTCGTTATTG\n",
      "                        R2: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 3:N:0: CGATGTTT\n",
      "                            CAATTTTGCAGTACAAAAATAATACCTCCTCTATTTATCCAAAATCACAAAAAACCACCCACTTAACTTTCCCTAA\n",
      "                        \n",
      "                        After --implicon trimming:\n",
      "                        R1: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 1:N:0: CGATGTTT:CAATTTTG\n",
      "                            ATCTAGTTCAGTACGGTGTTTTCGAATTAGAAAAATATGTATAGAGGAAATAGATATAAAGGCGTATTCGTTATTG\n",
      "                        R2: @HWI-D00436:407:CCAETANXX:1:1101:4105:1905 3:N:0: CGATGTTT:CAATTTTG\n",
      "                                    CAGTACAAAAATAATACCTCCTCTATTTATCCAAAATCACAAAAAACCACCCACTTAACTTTCCCTAA\n",
      "\n",
      "RRBS-specific options (MspI digested material):\n",
      "\n",
      "--rrbs                  Specifies that the input file was an MspI digested RRBS sample (recognition\n",
      "                        site: CCGG). Single-end or Read 1 sequences (paired-end) which were adapter-trimmed\n",
      "                        will have a further 2 bp removed from their 3' end. Sequences which were merely\n",
      "                        trimmed because of poor quality will not be shortened further. Read 2 of paired-end\n",
      "                        libraries will in addition have the first 2 bp removed from the 5' end (by setting\n",
      "                        '--clip_r2 2'). This is to avoid using artificial methylation calls from the filled-in\n",
      "                        cytosine positions close to the 3' MspI site in sequenced fragments.\n",
      "                        This option is not recommended for users of the Tecan Ovation RRBS Methyl-Seq with TrueMethyl\n",
      "                        oxBS 1-16 kit (see below).\n",
      "\n",
      "--non_directional       Selecting this option for non-directional RRBS libraries will screen\n",
      "                        quality-trimmed sequences for 'CAA' or 'CGA' at the start of the read\n",
      "                        and, if found, removes the first two basepairs. Like with the option\n",
      "                        '--rrbs' this avoids using cytosine positions that were filled-in\n",
      "                        during the end-repair step. '--non_directional' requires '--rrbs' to\n",
      "                        be specified as well. Note that this option does not set '--clip_r2 2' in\n",
      "                        paired-end mode.\n",
      "\n",
      "--keep                  Keep the quality trimmed intermediate file. Default: off, which means\n",
      "                        the temporary file is being deleted after adapter trimming. Only has\n",
      "                        an effect for RRBS samples since other FastQ files are not trimmed\n",
      "                        for poor qualities separately.\n",
      "\n",
      "\n",
      "Note for RRBS using the Tecan Ovation RRBS Methyl-Seq with TrueMethyl oxBS 1-16 kit:\n",
      "\n",
      "Owing to the fact that the Tecan Ovation RRBS kit attaches a varying number of nucleotides (0-3) after each MspI\n",
      "site Trim Galore should be run WITHOUT the option --rrbs. This trimming is accomplished in a subsequent\n",
      "diversity trimming step afterwards (see their manual).\n",
      "\n",
      "\n",
      "\n",
      "Note for RRBS using MseI:\n",
      "\n",
      "If your DNA material was digested with MseI (recognition motif: TTAA) instead of MspI it is NOT necessary\n",
      "to specify --rrbs or --non_directional since virtually all reads should start with the sequence\n",
      "'TAA', and this holds true for both directional and non-directional libraries. As the end-repair of 'TAA'\n",
      "restricted sites does not involve any cytosines it does not need to be treated especially. Instead, simply\n",
      "run Trim Galore! in the standard (i.e. non-RRBS) mode.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Paired-end specific options:\n",
      "\n",
      "--paired                This option performs length trimming of quality/adapter/RRBS trimmed reads for\n",
      "                        paired-end files. To pass the validation test, both sequences of a sequence pair\n",
      "                        are required to have a certain minimum length which is governed by the option\n",
      "                        --length (see above). If only one read passes this length threshold the\n",
      "                        other read can be rescued (see option --retain_unpaired). Using this option lets\n",
      "                        you discard too short read pairs without disturbing the sequence-by-sequence order\n",
      "                        of FastQ files which is required by many aligners.\n",
      "\n",
      "                        Trim Galore! expects paired-end files to be supplied in a pairwise fashion, e.g.\n",
      "                        file1_1.fq file1_2.fq SRR2_1.fq.gz SRR2_2.fq.gz ... .\n",
      "\n",
      "\n",
      "--retain_unpaired       If only one of the two paired-end reads became too short, the longer\n",
      "                        read will be written to either '.unpaired_1.fq' or '.unpaired_2.fq'\n",
      "                        output files. The length cutoff for unpaired single-end reads is\n",
      "                        governed by the parameters -r1/--length_1 and -r2/--length_2. Default: OFF.\n",
      "\n",
      "-r1/--length_1 <INT>    Unpaired single-end read length cutoff needed for read 1 to be written to\n",
      "                        '.unpaired_1.fq' output file. These reads may be mapped in single-end mode.\n",
      "                        Default: 35 bp.\n",
      "\n",
      "-r2/--length_2 <INT>    Unpaired single-end read length cutoff needed for read 2 to be written to\n",
      "                        '.unpaired_2.fq' output file. These reads may be mapped in single-end mode.\n",
      "                        Default: 35 bp.\n",
      "\n",
      "Last modified on 02 02 2023.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Run Trim Galore\n",
    "!trim_galore --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "43223aed-128b-4ecb-b444-541496b933c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "a984e1d3-1f6d-4a14-adde-2b54a5dbcdfe",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null\n",
      "Analysis complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 1.18\n",
      "single-core operation.\n",
      "Proceeding with 'gzip' for decompression\n",
      "To decrease CPU usage of decompression, please install 'igzip' and run again\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /mnt/e/new_workflow/rerun_1/trimmed_output/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001.fastq <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t24\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "smallRNA\t13\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t4\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 24). Second best hit was smallRNA (count: 13)\n",
      "\n",
      "Writing report to '/mnt/e/new_workflow/rerun_1/trimmed_output/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001.fastq_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001.fastq\n",
      "Trimming mode: single-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 1.18\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 30\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length before a sequence gets removed: 25 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "\n",
      "Cutadapt seems to be reasonably up-to-date. Setting -j 1\n",
      "Writing final adapter and quality trimmed output to JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 30') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001.fastq <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 1.18 with Python 2.7.18\n",
      "Command line parameters: -j 1 -e 0.1 -q 30 -O 1 -a AGATCGGAAGAGC rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 402.29 s (8 us/read; 7.60 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              50,986,227\n",
      "Reads with adapters:                 7,000,084 (13.7%)\n",
      "Reads written (passing filters):    50,986,227 (100.0%)\n",
      "\n",
      "Total basepairs processed: 4,994,568,985 bp\n",
      "Quality-trimmed:              50,973,042 bp (1.0%)\n",
      "Total written (filtered):  4,933,528,013 bp (98.8%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 7000084 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 30.5%\n",
      "  C: 28.5%\n",
      "  G: 20.9%\n",
      "  T: 19.9%\n",
      "  none/other: 0.2%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t4867965\t12746556.8\t0\t4867965\n",
      "2\t1634117\t3186639.2\t0\t1634117\n",
      "3\t390200\t796659.8\t0\t390200\n",
      "4\t75582\t199164.9\t0\t75582\n",
      "5\t19879\t49791.2\t0\t19879\n",
      "6\t2220\t12447.8\t0\t2220\n",
      "7\t418\t3112.0\t0\t418\n",
      "8\t58\t778.0\t0\t58\n",
      "9\t360\t194.5\t0\t15 345\n",
      "10\t269\t48.6\t1\t21 248\n",
      "11\t599\t12.2\t1\t57 542\n",
      "12\t78\t3.0\t1\t1 77\n",
      "13\t39\t0.8\t1\t1 38\n",
      "14\t878\t0.8\t1\t1 877\n",
      "15\t98\t0.8\t1\t0 98\n",
      "16\t792\t0.8\t1\t14 778\n",
      "17\t169\t0.8\t1\t27 142\n",
      "18\t663\t0.8\t1\t238 425\n",
      "19\t663\t0.8\t1\t200 463\n",
      "20\t84\t0.8\t1\t4 80\n",
      "21\t82\t0.8\t1\t8 74\n",
      "22\t71\t0.8\t1\t10 61\n",
      "23\t106\t0.8\t1\t23 83\n",
      "24\t95\t0.8\t1\t33 62\n",
      "25\t126\t0.8\t1\t38 88\n",
      "26\t95\t0.8\t1\t10 85\n",
      "27\t138\t0.8\t1\t21 117\n",
      "28\t153\t0.8\t1\t26 127\n",
      "29\t49\t0.8\t1\t7 42\n",
      "30\t89\t0.8\t1\t7 82\n",
      "31\t60\t0.8\t1\t9 51\n",
      "32\t70\t0.8\t1\t7 63\n",
      "33\t56\t0.8\t1\t0 56\n",
      "34\t123\t0.8\t1\t9 114\n",
      "35\t95\t0.8\t1\t14 81\n",
      "36\t80\t0.8\t1\t6 74\n",
      "37\t55\t0.8\t1\t4 51\n",
      "38\t63\t0.8\t1\t9 54\n",
      "39\t46\t0.8\t1\t4 42\n",
      "40\t71\t0.8\t1\t6 65\n",
      "41\t65\t0.8\t1\t4 61\n",
      "42\t86\t0.8\t1\t0 86\n",
      "43\t37\t0.8\t1\t2 35\n",
      "44\t56\t0.8\t1\t5 51\n",
      "45\t151\t0.8\t1\t3 148\n",
      "46\t74\t0.8\t1\t2 72\n",
      "47\t70\t0.8\t1\t0 70\n",
      "48\t45\t0.8\t1\t1 44\n",
      "49\t49\t0.8\t1\t0 49\n",
      "50\t44\t0.8\t1\t0 44\n",
      "51\t53\t0.8\t1\t0 53\n",
      "52\t44\t0.8\t1\t1 43\n",
      "53\t46\t0.8\t1\t2 44\n",
      "54\t53\t0.8\t1\t0 53\n",
      "55\t59\t0.8\t1\t3 56\n",
      "56\t45\t0.8\t1\t0 45\n",
      "57\t35\t0.8\t1\t0 35\n",
      "58\t67\t0.8\t1\t3 64\n",
      "59\t26\t0.8\t1\t0 26\n",
      "60\t46\t0.8\t1\t0 46\n",
      "61\t52\t0.8\t1\t2 50\n",
      "62\t59\t0.8\t1\t0 59\n",
      "63\t56\t0.8\t1\t0 56\n",
      "64\t112\t0.8\t1\t0 112\n",
      "65\t51\t0.8\t1\t0 51\n",
      "66\t95\t0.8\t1\t1 94\n",
      "67\t44\t0.8\t1\t0 44\n",
      "68\t51\t0.8\t1\t0 51\n",
      "69\t72\t0.8\t1\t0 72\n",
      "70\t44\t0.8\t1\t0 44\n",
      "71\t36\t0.8\t1\t0 36\n",
      "72\t63\t0.8\t1\t1 62\n",
      "73\t57\t0.8\t1\t0 57\n",
      "74\t40\t0.8\t1\t0 40\n",
      "75\t72\t0.8\t1\t0 72\n",
      "76\t31\t0.8\t1\t0 31\n",
      "77\t70\t0.8\t1\t0 70\n",
      "78\t34\t0.8\t1\t0 34\n",
      "79\t47\t0.8\t1\t0 47\n",
      "80\t77\t0.8\t1\t0 77\n",
      "81\t32\t0.8\t1\t0 32\n",
      "82\t74\t0.8\t1\t0 74\n",
      "83\t29\t0.8\t1\t0 29\n",
      "84\t60\t0.8\t1\t0 60\n",
      "85\t45\t0.8\t1\t0 45\n",
      "86\t44\t0.8\t1\t0 44\n",
      "87\t28\t0.8\t1\t0 28\n",
      "88\t37\t0.8\t1\t0 37\n",
      "89\t39\t0.8\t1\t0 39\n",
      "90\t32\t0.8\t1\t0 32\n",
      "91\t36\t0.8\t1\t0 36\n",
      "92\t29\t0.8\t1\t0 29\n",
      "93\t16\t0.8\t1\t0 16\n",
      "94\t41\t0.8\t1\t0 41\n",
      "95\t26\t0.8\t1\t0 26\n",
      "96\t72\t0.8\t1\t1 71\n",
      "97\t31\t0.8\t1\t0 31\n",
      "98\t87\t0.8\t1\t0 87\n",
      "99\t32\t0.8\t1\t0 32\n",
      "100\t29\t0.8\t1\t0 29\n",
      "101\t97\t0.8\t1\t0 97\n",
      "\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001.fastq\n",
      "=============================================\n",
      "50986227 sequences processed in total\n",
      "Sequences removed because they became shorter than the length cutoff of 25 bp:\t374844 (0.7%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the data <<<\n",
      "\n",
      "Started analysis of JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 5% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 10% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 15% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 20% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 25% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 30% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 35% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 40% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 45% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 50% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 55% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 60% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 65% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 70% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 75% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 80% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 85% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 90% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n",
      "Approx 95% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001_trimmed.fq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null\n",
      "Analysis complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 1.18\n",
      "single-core operation.\n",
      "Proceeding with 'gzip' for decompression\n",
      "To decrease CPU usage of decompression, please install 'igzip' and run again\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /mnt/e/new_workflow/rerun_1/trimmed_output/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001.fastq <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t2317\tAGATCGGAAGAGC\t1000000\t0.23\n",
      "smallRNA\t6\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t2\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 2317). Second best hit was smallRNA (count: 6)\n",
      "\n",
      "Writing report to '/mnt/e/new_workflow/rerun_1/trimmed_output/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001.fastq_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001.fastq\n",
      "Trimming mode: single-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 1.18\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 30\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length before a sequence gets removed: 25 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "\n",
      "Cutadapt seems to be reasonably up-to-date. Setting -j 1\n",
      "Writing final adapter and quality trimmed output to JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 30') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001.fastq <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 1.18 with Python 2.7.18\n",
      "Command line parameters: -j 1 -e 0.1 -q 30 -O 1 -a AGATCGGAAGAGC rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 403.24 s (8 us/read; 7.59 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              50,986,227\n",
      "Reads with adapters:                 7,707,882 (15.1%)\n",
      "Reads written (passing filters):    50,986,227 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,002,602,577 bp\n",
      "Quality-trimmed:              87,266,371 bp (1.7%)\n",
      "Total written (filtered):  4,901,682,411 bp (98.0%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 7707882 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 32.3%\n",
      "  C: 27.5%\n",
      "  G: 21.9%\n",
      "  T: 17.7%\n",
      "  none/other: 0.6%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t5293352\t12746556.8\t0\t5293352\n",
      "2\t1790305\t3186639.2\t0\t1790305\n",
      "3\t389974\t796659.8\t0\t389974\n",
      "4\t78026\t199164.9\t0\t78026\n",
      "5\t23326\t49791.2\t0\t23326\n",
      "6\t6276\t12447.8\t0\t6276\n",
      "7\t1153\t3112.0\t0\t1153\n",
      "8\t344\t778.0\t0\t344\n",
      "9\t1315\t194.5\t0\t616 699\n",
      "10\t3272\t48.6\t1\t1123 2149\n",
      "11\t1159\t12.2\t1\t632 527\n",
      "12\t3165\t3.0\t1\t2164 1001\n",
      "13\t1483\t0.8\t1\t358 1125\n",
      "14\t2847\t0.8\t1\t900 1947\n",
      "15\t2387\t0.8\t1\t583 1804\n",
      "16\t2651\t0.8\t1\t896 1755\n",
      "17\t3698\t0.8\t1\t1924 1774\n",
      "18\t3432\t0.8\t1\t1809 1623\n",
      "19\t51205\t0.8\t1\t49113 2092\n",
      "20\t1201\t0.8\t1\t852 349\n",
      "21\t185\t0.8\t1\t70 115\n",
      "22\t254\t0.8\t1\t29 225\n",
      "23\t1083\t0.8\t1\t462 621\n",
      "24\t1750\t0.8\t1\t1081 669\n",
      "25\t460\t0.8\t1\t237 223\n",
      "26\t7340\t0.8\t1\t6709 631\n",
      "27\t2465\t0.8\t1\t2191 274\n",
      "28\t322\t0.8\t1\t211 111\n",
      "29\t272\t0.8\t1\t192 80\n",
      "30\t288\t0.8\t1\t209 79\n",
      "31\t1640\t0.8\t1\t1405 235\n",
      "32\t839\t0.8\t1\t677 162\n",
      "33\t5719\t0.8\t1\t5262 457\n",
      "34\t594\t0.8\t1\t478 116\n",
      "35\t2864\t0.8\t1\t2591 273\n",
      "36\t454\t0.8\t1\t347 107\n",
      "37\t2286\t0.8\t1\t2004 282\n",
      "38\t251\t0.8\t1\t178 73\n",
      "39\t1611\t0.8\t1\t1420 191\n",
      "40\t333\t0.8\t1\t261 72\n",
      "41\t823\t0.8\t1\t725 98\n",
      "42\t2132\t0.8\t1\t1939 193\n",
      "43\t776\t0.8\t1\t673 103\n",
      "44\t651\t0.8\t1\t554 97\n",
      "45\t3705\t0.8\t1\t3438 267\n",
      "46\t312\t0.8\t1\t244 68\n",
      "47\t335\t0.8\t1\t249 86\n",
      "48\t310\t0.8\t1\t235 75\n",
      "49\t940\t0.8\t1\t804 136\n",
      "50\t153\t0.8\t1\t98 55\n",
      "51\t775\t0.8\t1\t654 121\n",
      "52\t207\t0.8\t1\t127 80\n",
      "53\t340\t0.8\t1\t289 51\n",
      "54\t760\t0.8\t1\t658 102\n",
      "55\t261\t0.8\t1\t211 50\n",
      "56\t373\t0.8\t1\t305 68\n",
      "57\t480\t0.8\t1\t398 82\n",
      "58\t234\t0.8\t1\t187 47\n",
      "59\t159\t0.8\t1\t90 69\n",
      "60\t254\t0.8\t1\t195 59\n",
      "61\t183\t0.8\t1\t144 39\n",
      "62\t270\t0.8\t1\t222 48\n",
      "63\t143\t0.8\t1\t84 59\n",
      "64\t92\t0.8\t1\t52 40\n",
      "65\t54\t0.8\t1\t18 36\n",
      "66\t151\t0.8\t1\t87 64\n",
      "67\t67\t0.8\t1\t17 50\n",
      "68\t50\t0.8\t1\t20 30\n",
      "69\t84\t0.8\t1\t35 49\n",
      "70\t59\t0.8\t1\t21 38\n",
      "71\t37\t0.8\t1\t8 29\n",
      "72\t43\t0.8\t1\t7 36\n",
      "73\t49\t0.8\t1\t5 44\n",
      "74\t38\t0.8\t1\t5 33\n",
      "75\t47\t0.8\t1\t1 46\n",
      "76\t38\t0.8\t1\t1 37\n",
      "77\t40\t0.8\t1\t2 38\n",
      "78\t49\t0.8\t1\t0 49\n",
      "79\t58\t0.8\t1\t2 56\n",
      "80\t36\t0.8\t1\t1 35\n",
      "81\t53\t0.8\t1\t0 53\n",
      "82\t52\t0.8\t1\t1 51\n",
      "83\t34\t0.8\t1\t0 34\n",
      "84\t37\t0.8\t1\t0 37\n",
      "85\t45\t0.8\t1\t0 45\n",
      "86\t37\t0.8\t1\t0 37\n",
      "87\t29\t0.8\t1\t0 29\n",
      "88\t31\t0.8\t1\t0 31\n",
      "89\t19\t0.8\t1\t0 19\n",
      "90\t21\t0.8\t1\t1 20\n",
      "91\t40\t0.8\t1\t0 40\n",
      "92\t38\t0.8\t1\t0 38\n",
      "93\t28\t0.8\t1\t0 28\n",
      "94\t18\t0.8\t1\t0 18\n",
      "95\t12\t0.8\t1\t1 11\n",
      "96\t35\t0.8\t1\t1 34\n",
      "97\t38\t0.8\t1\t0 38\n",
      "98\t76\t0.8\t1\t0 76\n",
      "99\t33\t0.8\t1\t0 33\n",
      "100\t38\t0.8\t1\t0 38\n",
      "101\t19\t0.8\t1\t0 19\n",
      "\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001.fastq\n",
      "=============================================\n",
      "50986227 sequences processed in total\n",
      "Sequences removed because they became shorter than the length cutoff of 25 bp:\t720371 (1.4%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the data <<<\n",
      "\n",
      "Started analysis of JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 5% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 10% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 15% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 20% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 25% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 30% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 35% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 40% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 45% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 50% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 55% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 60% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 65% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 70% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 75% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 80% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 85% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 90% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n",
      "Approx 95% complete for JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001_trimmed.fq\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Set the number of cores to use\n",
    "num_cores=8  # You can adjust this based on your system's capabilities\n",
    "\n",
    "# Run Trim Galore! in parallel for all files in the folder\n",
    "find rerun_1/ -name '*.fastq' | parallel -j $num_cores \\\n",
    "'trim_galore \\\n",
    "--quality 30 \\\n",
    "--fastqc \\\n",
    "--length 25 \\\n",
    "--output_dir rerun_1/trimmed_output/ \\\n",
    "{}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49de8b0f-325f-420f-96d3-42882d6d2530",
   "metadata": {},
   "source": [
    "# Rerun test of seq files that yielded error "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "38dd128f-2d5d-4f89-9aac-f39c8b3d9ed1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Set the number of cores to use\n",
    "num_cores=8  # You can adjust this based on your system's capabilities\n",
    "\n",
    "# Run Trim Galore! in parallel for all files in the folder\n",
    "find input/ -name '*.fastq' | parallel -j $num_cores \\\n",
    "'trim_galore \\\n",
    "--quality 20 \\\n",
    "--fastqc \\\n",
    "--length 25 \\\n",
    "--output_dir results/2_trimmed_output/ \\\n",
    "{}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "eab60a4c-8414-492a-b926-1ed5d54efe8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/home/raotoo/miniconda/bin/STAR-avx2 --genomeDir /mnt/e/new_workflow/star_index/ --readFilesIn /mnt/e/new_workflow/rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001.fastq /mnt/e/new_workflow/rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001.fastq --runThreadN 8 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --outFileNamePrefix /mnt/e/new_workflow/rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_STAR_\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Jan 24 16:16:30 ..... started STAR run\n",
      "Jan 24 16:16:30 ..... loading genome\n",
      "Jan 24 16:19:08 ..... started mapping\n",
      "Jan 24 16:31:07 ..... finished mapping\n",
      "Jan 24 16:31:12 ..... started sorting BAM\n",
      "Jan 24 16:33:39 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "###--------------------------------------- Test Run: Runs one pair of read-----------------------------------------------------------\n",
    "#%%bash\n",
    "\n",
    "# Set the path to the STAR genome index\n",
    "#genome_dir=\"/mnt/e/new_workflow/star_index/\"  # Replace with the actual path\n",
    "\n",
    "# Set the path to the output directory\n",
    "#output_dir=\"/mnt/e/new_workflow/rerun_1/\"\n",
    "\n",
    "# Run STAR for the specific pair of reads\n",
    "!STAR \\\n",
    "    --genomeDir /mnt/e/new_workflow/star_index/ \\\n",
    "    --readFilesIn /mnt/e/new_workflow/rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R1_001.fastq \\\n",
    "                 /mnt/e/new_workflow/rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_R2_001.fastq \\\n",
    "    --runThreadN 8 \\\n",
    "    --outSAMtype BAM SortedByCoordinate \\\n",
    "    --quantMode GeneCounts \\\n",
    "    --outFileNamePrefix \"/mnt/e/new_workflow/rerun_1/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_STAR_\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "00f9279f-a4ce-4e1a-9238-b89558858271",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "753fbc85-8d1b-44db-97dd-fde14684f624",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_Aligned.sortedByCoord.out.bam' -> 'results/4_aligned_sequences/aligned_bam/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_Aligned.sortedByCoord.out.bam'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_STAR_Aligned.sortedByCoord.out.bam' -> 'results/4_aligned_sequences/aligned_bam/JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_STAR_Aligned.sortedByCoord.out.bam'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_STAR_Aligned.sortedByCoord.out.bam' -> 'results/4_aligned_sequences/aligned_bam/JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_STAR_Aligned.sortedByCoord.out.bam'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_STAR_Aligned.sortedByCoord.out.bam' -> 'results/4_aligned_sequences/aligned_bam/JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_STAR_Aligned.sortedByCoord.out.bam'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_STAR_Aligned.sortedByCoord.out.bam' -> 'results/4_aligned_sequences/aligned_bam/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_STAR_Aligned.sortedByCoord.out.bam'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_STAR_Aligned.sortedByCoord.out.bam' -> 'results/4_aligned_sequences/aligned_bam/JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_STAR_Aligned.sortedByCoord.out.bam'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_STAR_Aligned.sortedByCoord.out.bam' -> 'results/4_aligned_sequences/aligned_bam/JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_STAR_Aligned.sortedByCoord.out.bam'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_STAR_Aligned.sortedByCoord.out.bam' -> 'results/4_aligned_sequences/aligned_bam/JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_STAR_Aligned.sortedByCoord.out.bam'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_STAR_Aligned.sortedByCoord.out.bam' -> 'results/4_aligned_sequences/aligned_bam/JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_STAR_Aligned.sortedByCoord.out.bam'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_STAR_Aligned.sortedByCoord.out.bam' -> 'results/4_aligned_sequences/aligned_bam/JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_STAR_Aligned.sortedByCoord.out.bam'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_Log.final.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_Log.final.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_STAR_Log.final.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_STAR_Log.final.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_STAR_Log.final.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_STAR_Log.final.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_STAR_Log.final.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_STAR_Log.final.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_STAR_Log.final.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_STAR_Log.final.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_STAR_Log.final.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_STAR_Log.final.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_STAR_Log.final.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_STAR_Log.final.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_STAR_Log.final.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_STAR_Log.final.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_STAR_Log.final.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_STAR_Log.final.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_Log.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_Log.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_STAR_Log.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_STAR_Log.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_STAR_Log.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_STAR_Log.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_STAR_Log.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_STAR_Log.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_STAR_Log.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_STAR_Log.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_STAR_Log.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_STAR_Log.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_STAR_Log.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_STAR_Log.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_STAR_Log.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_STAR_Log.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_STAR_Log.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_STAR_Log.out'\n",
      "renamed 'results/4_aligned_sequences/JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_STAR_Log.out' -> 'results/4_aligned_sequences/aligned_logs/JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_STAR_Log.out'\n"
     ]
    }
   ],
   "source": [
    "#----------------------------------------------------For the whole batch--------------------------------------------------------------------------\n",
    "# Move the BAM file into the correct folder\n",
    "!mv -v results/4_aligned_sequences/*sortedByCoord.out.bam results/4_aligned_sequences/aligned_bam/\n",
    "\n",
    "# Move the logs into the correct folder\n",
    "!mv -v results/4_aligned_sequences/*Log.final.out results/4_aligned_sequences/aligned_logs/\n",
    "!mv -v results/4_aligned_sequences/*Log.out results/4_aligned_sequences/aligned_logs/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "1fd9090e-1d6a-4ffb-b444-c1086eda31f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mv: cannot stat 'rerun_1/*sortedByCoord.out.bam': No such file or directory\n",
      "mv: cannot stat 'rerun/*Log.final.out': No such file or directory\n",
      "mv: cannot stat 'rerun/*Log.out': No such file or directory\n"
     ]
    }
   ],
   "source": [
    "#---------------------------------------------------For the sample that was reanalyzed ------------------------------------------------------------------\n",
    "# Move the BAM file into the correct folder\n",
    "!mv -v rerun_1/*sortedByCoord.out.bam results/4_aligned_sequences/aligned_bam/\n",
    "\n",
    "# Move the logs into the correct folder\n",
    "!mv -v rerun/*Log.final.out results/4_aligned_sequences/aligned_logs/\n",
    "!mv -v rerun/*Log.out results/4_aligned_sequences/aligned_logs/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17abbe83-e45e-4fdd-a181-a8a1425d8b63",
   "metadata": {},
   "source": [
    "## 5. Summarizing Gene Counts with featureCounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "fd47be4c-4820-4a64-87af-2abc9b4f7a39",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving notices: ...working... done\n",
      "Channels:\n",
      " - bioconda\n",
      " - r\n",
      " - defaults\n",
      " - conda-forge\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/raotoo/miniconda\n",
      "\n",
      "  added / updated specs:\n",
      "    - subread\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    subread-2.0.6              |       he4a0461_0        26.1 MB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        26.1 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  subread            bioconda/linux-64::subread-2.0.6-he4a0461_0 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages:\n",
      "                                                                                \n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n"
     ]
    }
   ],
   "source": [
    "conda install -c bioconda subread --yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "7e9ea9ec-133a-4e76-bc75-df21fdcfffed",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "featureCounts: invalid option -- 'h'\n",
      "\n",
      "Version 2.0.6\n",
      "\n",
      "Usage: featureCounts [options] -a <annotation_file> -o <output_file> input_file1 [input_file2] ... \n",
      "\n",
      "## Mandatory arguments:\n",
      "\n",
      "  -a <string>         Name of an annotation file. GTF/GFF format by default. See\n",
      "                      -F option for more format information. Inbuilt annotations\n",
      "                      (SAF format) is available in 'annotation' directory of the\n",
      "                      package. Gzipped file is also accepted.\n",
      "\n",
      "  -o <string>         Name of output file including read counts. A separate file\n",
      "                      including summary statistics of counting results is also\n",
      "                      included in the output ('<string>.summary'). Both files\n",
      "                      are in tab delimited format.\n",
      "\n",
      "  input_file1 [input_file2] ...   A list of SAM or BAM format files. They can be\n",
      "                      either name or location sorted. If no files provided,\n",
      "                      <stdin> input is expected. Location-sorted paired-end reads\n",
      "                      are automatically sorted by read names.\n",
      "\n",
      "## Optional arguments:\n",
      "# Annotation\n",
      "\n",
      "  -F <string>         Specify format of the provided annotation file. Acceptable\n",
      "                      formats include 'GTF' (or compatible GFF format) and\n",
      "                      'SAF'. 'GTF' by default.  For SAF format, please refer to\n",
      "                      Users Guide.\n",
      "\n",
      "  -t <string>         Specify feature type(s) in a GTF annotation. If multiple\n",
      "                      types are provided, they should be separated by ',' with\n",
      "                      no space in between. 'exon' by default. Rows in the\n",
      "                      annotation with a matched feature will be extracted and\n",
      "                      used for read mapping. \n",
      "\n",
      "  -g <string>         Specify attribute type in GTF annotation. 'gene_id' by \n",
      "                      default. Meta-features used for read counting will be \n",
      "                      extracted from annotation using the provided value.\n",
      "\n",
      "  --extraAttributes   Extract extra attribute types from the provided GTF\n",
      "                      annotation and include them in the counting output. These\n",
      "                      attribute types will not be used to group features. If\n",
      "                      more than one attribute type is provided they should be\n",
      "                      separated by comma.\n",
      "\n",
      "  -A <string>         Provide a chromosome name alias file to match chr names in\n",
      "                      annotation with those in the reads. This should be a two-\n",
      "                      column comma-delimited text file. Its first column should\n",
      "                      include chr names in the annotation and its second column\n",
      "                      should include chr names in the reads. Chr names are case\n",
      "                      sensitive. No column header should be included in the\n",
      "                      file.\n",
      "\n",
      "# Level of summarization\n",
      "\n",
      "  -f                  Perform read counting at feature level (eg. counting \n",
      "                      reads for exons rather than genes).\n",
      "\n",
      "# Overlap between reads and features\n",
      "\n",
      "  -O                  Assign reads to all their overlapping meta-features (or \n",
      "                      features if -f is specified).\n",
      "\n",
      "  --minOverlap <int>  Minimum number of overlapping bases in a read that is\n",
      "                      required for read assignment. 1 by default. Number of\n",
      "                      overlapping bases is counted from both reads if paired\n",
      "                      end. If a negative value is provided, then a gap of up\n",
      "                      to specified size will be allowed between read and the\n",
      "                      feature that the read is assigned to.\n",
      "\n",
      "  --fracOverlap <float> Minimum fraction of overlapping bases in a read that is\n",
      "                      required for read assignment. Value should be within range\n",
      "                      [0,1]. 0 by default. Number of overlapping bases is\n",
      "                      counted from both reads if paired end. Both this option\n",
      "                      and '--minOverlap' option need to be satisfied for read\n",
      "                      assignment.\n",
      "\n",
      "  --fracOverlapFeature <float> Minimum fraction of overlapping bases in a\n",
      "                      feature that is required for read assignment. Value\n",
      "                      should be within range [0,1]. 0 by default.\n",
      "\n",
      "  --largestOverlap    Assign reads to a meta-feature/feature that has the \n",
      "                      largest number of overlapping bases.\n",
      "\n",
      "  --nonOverlap <int>  Maximum number of non-overlapping bases in a read (or a\n",
      "                      read pair) that is allowed when being assigned to a\n",
      "                      feature. No limit is set by default.\n",
      "\n",
      "  --nonOverlapFeature <int> Maximum number of non-overlapping bases in a feature\n",
      "                      that is allowed in read assignment. No limit is set by\n",
      "                      default.\n",
      "\n",
      "  --readExtension5 <int> Reads are extended upstream by <int> bases from their\n",
      "                      5' end.\n",
      "\n",
      "  --readExtension3 <int> Reads are extended upstream by <int> bases from their\n",
      "                      3' end.\n",
      "\n",
      "  --read2pos <5:3>    Reduce reads to their 5' most base or 3' most base. Read\n",
      "                      counting is then performed based on the single base the \n",
      "                      read is reduced to.\n",
      "\n",
      "# Multi-mapping reads\n",
      "\n",
      "  -M                  Multi-mapping reads will also be counted. For a multi-\n",
      "                      mapping read, all its reported alignments will be \n",
      "                      counted. The 'NH' tag in BAM/SAM input is used to detect \n",
      "                      multi-mapping reads.\n",
      "\n",
      "# Fractional counting\n",
      "\n",
      "  --fraction          Assign fractional counts to features. This option must\n",
      "                      be used together with '-M' or '-O' or both. When '-M' is\n",
      "                      specified, each reported alignment from a multi-mapping\n",
      "                      read (identified via 'NH' tag) will carry a fractional\n",
      "                      count of 1/x, instead of 1 (one), where x is the total\n",
      "                      number of alignments reported for the same read. When '-O'\n",
      "                      is specified, each overlapping feature will receive a\n",
      "                      fractional count of 1/y, where y is the total number of\n",
      "                      features overlapping with the read. When both '-M' and\n",
      "                      '-O' are specified, each alignment will carry a fractional\n",
      "                      count of 1/(x*y).\n",
      "\n",
      "# Read filtering\n",
      "\n",
      "  -Q <int>            The minimum mapping quality score a read must satisfy in\n",
      "                      order to be counted. For paired-end reads, at least one\n",
      "                      end should satisfy this criteria. 0 by default.\n",
      "\n",
      "  --splitOnly         Count split alignments only (ie. alignments with CIGAR\n",
      "                      string containing 'N'). An example of split alignments is\n",
      "                      exon-spanning reads in RNA-seq data.\n",
      "\n",
      "  --nonSplitOnly      If specified, only non-split alignments (CIGAR strings do\n",
      "                      not contain letter 'N') will be counted. All the other\n",
      "                      alignments will be ignored.\n",
      "\n",
      "  --primary           Count primary alignments only. Primary alignments are \n",
      "                      identified using bit 0x100 in SAM/BAM FLAG field.\n",
      "\n",
      "  --ignoreDup         Ignore duplicate reads in read counting. Duplicate reads \n",
      "                      are identified using bit Ox400 in BAM/SAM FLAG field. The \n",
      "                      whole read pair is ignored if one of the reads is a \n",
      "                      duplicate read for paired end data.\n",
      "\n",
      "# Strandness\n",
      "\n",
      "  -s <int or string>  Perform strand-specific read counting. A single integer\n",
      "                      value (applied to all input files) or a string of comma-\n",
      "                      separated values (applied to each corresponding input\n",
      "                      file) should be provided. Possible values include:\n",
      "                      0 (unstranded), 1 (stranded) and 2 (reversely stranded).\n",
      "                      Default value is 0 (ie. unstranded read counting carried\n",
      "                      out for all input files).\n",
      "\n",
      "# Exon-exon junctions\n",
      "\n",
      "  -J                  Count number of reads supporting each exon-exon junction.\n",
      "                      Junctions were identified from all the exon-spanning reads\n",
      "                      in the input (containing 'N' in CIGAR string). Counting\n",
      "                      results are saved to a file named '<output_file>.jcounts'\n",
      "\n",
      "  -G <string>         Provide the name of a FASTA-format file that contains the\n",
      "                      reference sequences used in read mapping that produced the\n",
      "                      provided SAM/BAM files. This optional argument can be used\n",
      "                      with '-J' option to improve read counting for junctions.\n",
      "\n",
      "# Parameters specific to paired end reads\n",
      "\n",
      "  -p                  If specified, libraries are assumed to contain paired-end\n",
      "                      reads. For any library that contains paired-end reads, the\n",
      "                      'countReadPairs' parameter controls if read pairs or reads\n",
      "                      should be counted.\n",
      "\n",
      "  --countReadPairs    If specified, fragments (or templates) will be counted\n",
      "                      instead of reads. This option is only applicable for\n",
      "                      paired-end reads. For single-end data, it is ignored.\n",
      "\n",
      "  -B                  Only count read pairs that have both ends aligned.\n",
      "\n",
      "  -P                  Check validity of paired-end distance when counting read \n",
      "                      pairs. Use -d and -D to set thresholds.\n",
      "\n",
      "  -d <int>            Minimum fragment/template length, 50 by default.\n",
      "\n",
      "  -D <int>            Maximum fragment/template length, 600 by default.\n",
      "\n",
      "  -C                  Do not count read pairs that have their two ends mapping \n",
      "                      to different chromosomes or mapping to same chromosome \n",
      "                      but on different strands.\n",
      "\n",
      "  --donotsort         Do not sort reads in BAM/SAM input. Note that reads from \n",
      "                      the same pair are required to be located next to each \n",
      "                      other in the input.\n",
      "\n",
      "# Number of CPU threads\n",
      "\n",
      "  -T <int>            Number of the threads. 1 by default.\n",
      "\n",
      "# Read groups\n",
      "\n",
      "  --byReadGroup       Assign reads by read group. \"RG\" tag is required to be\n",
      "                      present in the input BAM/SAM files.\n",
      "                      \n",
      "\n",
      "# Long reads\n",
      "\n",
      "  -L                  Count long reads such as Nanopore and PacBio reads. Long\n",
      "                      read counting can only run in one thread and only reads\n",
      "                      (not read-pairs) can be counted. There is no limitation on\n",
      "                      the number of 'M' operations allowed in a CIGAR string in\n",
      "                      long read counting.\n",
      "\n",
      "# Assignment results for each read\n",
      "\n",
      "  -R <format>         Output detailed assignment results for each read or read-\n",
      "                      pair. Results are saved to a file that is in one of the\n",
      "                      following formats: CORE, SAM and BAM. See Users Guide for\n",
      "                      more info about these formats.\n",
      "\n",
      "  --Rpath <string>    Specify a directory to save the detailed assignment\n",
      "                      results. If unspecified, the directory where counting\n",
      "                      results are saved is used.\n",
      "\n",
      "# Miscellaneous\n",
      "\n",
      "  --tmpDir <string>   Directory under which intermediate files are saved (later\n",
      "                      removed). By default, intermediate files will be saved to\n",
      "                      the directory specified in '-o' argument.\n",
      "\n",
      "  --maxMOp <int>      Maximum number of 'M' operations allowed in a CIGAR\n",
      "                      string. 10 by default. Both 'X' and '=' are treated as 'M'\n",
      "                      and adjacent 'M' operations are merged in the CIGAR\n",
      "                      string.\n",
      "\n",
      "  --verbose           Output verbose information for debugging, such as un-\n",
      "                      matched chromosome/contig names.\n",
      "\n",
      "  -v                  Output version of the program.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Help\n",
    "!featureCounts -h\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d23e27bf-cbe0-45ee-9e89-d085eb025af3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow/results/4_aligned_sequences/aligned_bam\n"
     ]
    }
   ],
   "source": [
    "# Change directory into the aligned .BAM folder\n",
    "%cd results/4_aligned_sequences/aligned_bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e1bf8b7f-acf5-4972-9a40-76518f5c2e27",
   "metadata": {},
   "outputs": [],
   "source": [
    "!chmod -R u+w /mnt/e/new_workflow/results/5_final_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d194e790-4e25-4a12-ba5e-f2ef899c4870",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_STAR_Aligned.sortedByCoord.out.bam ./JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_STAR_Aligned.sortedByCoord.out.bam ./JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_STAR_Aligned.sortedByCoord.out.bam ./JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_STAR_Aligned.sortedByCoord.out.bam ./JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_STAR_Aligned.sortedByCoord.out.bam ./JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_STAR_Aligned.sortedByCoord.out.bam ./JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_STAR_Aligned.sortedByCoord.out.bam ./JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_STAR_Aligned.sortedByCoord.out.bam ./JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_STAR_Aligned.sortedByCoord.out.bam ./JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_Aligned.sortedByCoord.out.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "        ==========     _____ _    _ ____  _____  ______          _____  \n",
      "        =====         / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "          =====      | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "            ====      \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "              ====    ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "        ==========   |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\n",
      "\t  v2.0.6\n",
      "\n",
      "//========================== featureCounts setting ===========================\\\\\n",
      "||                                                                            ||\n",
      "||             Input files : 10 BAM files                                     ||\n",
      "||                                                                            ||\n",
      "||                           JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496 ... ||\n",
      "||                           JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496 ... ||\n",
      "||                           JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496 ... ||\n",
      "||                           JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496 ... ||\n",
      "||                           JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496 ... ||\n",
      "||                           JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496 ... ||\n",
      "||                           JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496 ... ||\n",
      "||                           JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496 ... ||\n",
      "||                           JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496 ... ||\n",
      "||                           JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496 ... ||\n",
      "||                                                                            ||\n",
      "||             Output file : final_counts.txt                                 ||\n",
      "||                 Summary : final_counts.txt.summary                         ||\n",
      "||              Paired-end : yes                                              ||\n",
      "||        Count read pairs : no                                               ||\n",
      "||              Annotation : gencode.vM34.annotation.gtf (GTF)                ||\n",
      "||      Dir for temp files : /mnt/e/new_workflow/results/5_final_counts       ||\n",
      "||                                                                            ||\n",
      "||                 Threads : 4                                                ||\n",
      "||                   Level : meta-feature level                               ||\n",
      "||      Multimapping reads : not counted                                      ||\n",
      "|| Multi-overlapping reads : not counted                                      ||\n",
      "||   Min overlapping bases : 1                                                ||\n",
      "||                                                                            ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//================================= Running ==================================\\\\\n",
      "||                                                                            ||\n",
      "|| Load annotation file gencode.vM34.annotation.gtf ...                       ||\n",
      "||    Features : 863151                                                       ||\n",
      "||    Meta-features : 56893                                                   ||\n",
      "||    Chromosomes/contigs : 22                                                ||\n",
      "||                                                                            ||\n",
      "|| Process BAM file JvRad_Rhipp_204_TSMRLT_efeebde0_P0330-SB0496_S5_STAR_ ... ||\n",
      "||    Paired-end reads are included.                                          ||\n",
      "||    The reads are assigned on the single-end mode.                          ||\n",
      "||    Total alignments : 93684326                                             ||\n",
      "||    Successfully assigned alignments : 64469817 (68.8%)                     ||\n",
      "||    Running time : 0.33 minutes                                             ||\n",
      "||                                                                            ||\n",
      "|| Process BAM file JvRad_Rhipp_504_TSMRLT_cfb7f9d5_P0330-SB0496_S10_STAR ... ||\n",
      "||    Paired-end reads are included.                                          ||\n",
      "||    The reads are assigned on the single-end mode.                          ||\n",
      "||    Total alignments : 724782                                               ||\n",
      "||    Successfully assigned alignments : 440443 (60.8%)                       ||\n",
      "||    Running time : 0.01 minutes                                             ||\n",
      "||                                                                            ||\n",
      "|| Process BAM file JvRad_Rhipp_503_TSMRLT_09cb4f24_P0330-SB0496_S9_STAR_ ... ||\n",
      "||    Paired-end reads are included.                                          ||\n",
      "||    The reads are assigned on the single-end mode.                          ||\n",
      "||    Total alignments : 2074965                                              ||\n",
      "||    Successfully assigned alignments : 1302925 (62.8%)                      ||\n",
      "||    Running time : 0.01 minutes                                             ||\n",
      "||                                                                            ||\n",
      "|| Process BAM file JvRad_Rhipp_402_TSMRLT_6615aec4_P0330-SB0496_S8_STAR_ ... ||\n",
      "||    Paired-end reads are included.                                          ||\n",
      "||    The reads are assigned on the single-end mode.                          ||\n",
      "||    Total alignments : 1142099                                              ||\n",
      "||    Successfully assigned alignments : 706456 (61.9%)                       ||\n",
      "||    Running time : 0.01 minutes                                             ||\n",
      "||                                                                            ||\n",
      "|| Process BAM file JvRad_Rhipp_401_TSMRLT_6f156c58_P0330-SB0496_S7_STAR_ ... ||\n",
      "||    Paired-end reads are included.                                          ||\n",
      "||    The reads are assigned on the single-end mode.                          ||\n",
      "||    Total alignments : 944211                                               ||\n",
      "||    Successfully assigned alignments : 590043 (62.5%)                       ||\n",
      "||    Running time : 0.01 minutes                                             ||\n",
      "||                                                                            ||\n",
      "|| Process BAM file JvRad_Rhipp_400_TSMRLT_c5c04043_P0330-SB0496_S6_STAR_ ... ||\n",
      "||    Paired-end reads are included.                                          ||\n",
      "||    The reads are assigned on the single-end mode.                          ||\n",
      "||    Total alignments : 1340865                                              ||\n",
      "||    Successfully assigned alignments : 858776 (64.0%)                       ||\n",
      "||    Running time : 0.01 minutes                                             ||\n",
      "||                                                                            ||\n",
      "|| Process BAM file JvRad_Rhipp_203_TSMRLT_e32edba2_P0330-SB0496_S4_STAR_ ... ||\n",
      "||    Paired-end reads are included.                                          ||\n",
      "||    The reads are assigned on the single-end mode.                          ||\n",
      "||    Total alignments : 506573                                               ||\n",
      "||    Successfully assigned alignments : 286294 (56.5%)                       ||\n",
      "||    Running time : 0.00 minutes                                             ||\n",
      "||                                                                            ||\n",
      "|| Process BAM file JvRad_Rhipp_202_TSMRLT_b3c033d6_P0330-SB0496_S3_STAR_ ... ||\n",
      "||    Paired-end reads are included.                                          ||\n",
      "||    The reads are assigned on the single-end mode.                          ||\n",
      "||    Total alignments : 745447                                               ||\n",
      "||    Successfully assigned alignments : 433478 (58.2%)                       ||\n",
      "||    Running time : 0.01 minutes                                             ||\n",
      "||                                                                            ||\n",
      "|| Process BAM file JvRad_Rhipp_201_TSMRLT_a429b07e_P0330-SB0496_S2_STAR_ ... ||\n",
      "||    Paired-end reads are included.                                          ||\n",
      "||    The reads are assigned on the single-end mode.                          ||\n",
      "||    Total alignments : 531551                                               ||\n",
      "||    Successfully assigned alignments : 284458 (53.5%)                       ||\n",
      "||    Running time : 0.00 minutes                                             ||\n",
      "||                                                                            ||\n",
      "|| Process BAM file JvRad_Rhipp_200_TSMRLT_d438483b_P0330-SB0496_S1_STAR_ ... ||\n",
      "||    Paired-end reads are included.                                          ||\n",
      "||    The reads are assigned on the single-end mode.                          ||\n",
      "||    Total alignments : 565938                                               ||\n",
      "||    Successfully assigned alignments : 335053 (59.2%)                       ||\n",
      "||    Running time : 0.01 minutes                                             ||\n",
      "||                                                                            ||\n",
      "|| Write the final count table.                                               ||\n",
      "|| Write the read assignment summary.                                         ||\n",
      "||                                                                            ||\n",
      "|| Summary of counting results can be found in file \"/mnt/e/new_workflow/res  ||\n",
      "|| ults/5_final_counts/final_counts.txt.summary\"                              ||\n",
      "||                                                                            ||\n",
      "\\\\============================================================================//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Store list of files as a variable\n",
    "dirlist=$(ls -t ./*.bam | tr '\\n' ' ')\n",
    "echo $dirlist\n",
    "\n",
    "# Run featureCounts on all of the samples (~10 minutes)\n",
    "featureCounts \\\n",
    "-p \\\n",
    "-a /mnt/e/new_workflow/annotation/* \\\n",
    "-o /mnt/e/new_workflow/results/5_final_counts/final_counts.txt \\\n",
    "-g 'gene_name' \\\n",
    "-T 4 \\\n",
    "$dirlist\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "435d2dc1-4583-4ed3-8020-8400ccccb84c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow/results/4_aligned_sequences/aligned_bam\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c8ac33a1-4a6a-4406-a7aa-71c21f1d993f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow/results/5_final_counts\n"
     ]
    }
   ],
   "source": [
    "# Change directory back to main folder\n",
    "%cd /mnt/e/new_workflow/results/5_final_counts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c3b40efb-fc0e-4981-8302-c48ac93d01cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final_counts.txt\t  normalized_counts.txt\n",
      "final_counts.txt.summary  normalized_counts_significant.txt\n",
      "mmu04070.xml\t\t  results_gene_annotated.txt\n",
      "mmu05012.xml\t\t  results_gene_annotated_significant.txt\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "706bbc84-2b52-41fa-be0e-c1f205237d5f",
   "metadata": {},
   "source": [
    "## 6. Generating report with MultiQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "7fe48988-63c1-412c-84c9-152799ec5457",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Channels:\n",
      " - bioconda\n",
      " - r\n",
      " - defaults\n",
      " - conda-forge\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/raotoo/miniconda\n",
      "\n",
      "  added / updated specs:\n",
      "    - multiqc\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    blas-1.0                   |         openblas          46 KB\n",
      "    brotli-1.0.9               |       h5eee18b_7          18 KB\n",
      "    brotli-bin-1.0.9           |       h5eee18b_7          19 KB\n",
      "    click-8.1.7                |  py311h06a4308_0         221 KB\n",
      "    coloredlogs-15.0.1         |  py311h06a4308_1          76 KB\n",
      "    colormath-3.0.0            |             py_2          34 KB  conda-forge\n",
      "    contourpy-1.2.0            |  py311hdb19cb5_0         263 KB\n",
      "    cycler-0.11.0              |     pyhd3eb1b0_0          12 KB\n",
      "    fonttools-4.25.0           |     pyhd3eb1b0_0         632 KB\n",
      "    future-0.18.3              |  py311h06a4308_0         824 KB\n",
      "    giflib-5.2.1               |       h5eee18b_3          80 KB\n",
      "    humanfriendly-10.0         |  py311h06a4308_1         148 KB\n",
      "    humanize-3.10.0            |     pyhd3eb1b0_0          45 KB\n",
      "    importlib-metadata-7.0.1   |  py311h06a4308_0          49 KB\n",
      "    kiwisolver-1.4.4           |  py311h6a678d5_0          70 KB\n",
      "    lcms2-2.12                 |       h3be6417_0         312 KB\n",
      "    libbrotlicommon-1.0.9      |       h5eee18b_7          70 KB\n",
      "    libbrotlidec-1.0.9         |       h5eee18b_7          31 KB\n",
      "    libbrotlienc-1.0.9         |       h5eee18b_7         264 KB\n",
      "    libgfortran-ng-11.2.0      |       h00389a5_1          20 KB\n",
      "    libgfortran5-11.2.0        |       h1234567_1         2.0 MB\n",
      "    libopenblas-0.3.21         |       h043d6bf_0         5.4 MB\n",
      "    libtiff-4.2.0              |       hecacb30_2         466 KB\n",
      "    libwebp-1.3.2              |       h11a3e52_0          87 KB\n",
      "    libwebp-base-1.3.2         |       h5eee18b_0         387 KB\n",
      "    lzstring-1.0.4             |          py_1001           6 KB  conda-forge\n",
      "    markdown-3.4.1             |  py311h06a4308_0         164 KB\n",
      "    markdown-it-py-2.2.0       |  py311h06a4308_1         145 KB\n",
      "    matplotlib-base-3.8.0      |  py311ha02d727_0         7.7 MB\n",
      "    mdurl-0.1.0                |  py311h06a4308_0          22 KB\n",
      "    multiqc-1.19               |     pyhdfd78af_0         1.0 MB  bioconda\n",
      "    munkres-1.0.7              |             py_1          10 KB  bioconda\n",
      "    networkx-3.1               |  py311h06a4308_0         3.3 MB\n",
      "    numpy-1.26.3               |  py311h24aa872_0          10 KB\n",
      "    numpy-base-1.26.3          |  py311hbfb1bba_0         8.3 MB\n",
      "    openjpeg-2.4.0             |       h3ad879b_0         331 KB\n",
      "    pillow-10.0.1              |  py311ha6cbd5a_0         896 KB\n",
      "    pyaml-env-1.1.0            |     pyhd8ed1ab_0          11 KB  conda-forge\n",
      "    pyparsing-3.0.9            |  py311h06a4308_0         204 KB\n",
      "    pyyaml-5.4.1               |  py311h5eee18b_3         212 KB\n",
      "    rich-13.3.5                |  py311h06a4308_0         560 KB\n",
      "    rich-click-1.7.3           |     pyhd8ed1ab_0          32 KB  conda-forge\n",
      "    simplejson-3.17.6          |  py311h5eee18b_0         128 KB\n",
      "    spectra-0.0.11             |             py_1          19 KB  conda-forge\n",
      "    zipp-3.17.0                |  py311h06a4308_0          24 KB\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        34.5 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  blas               pkgs/main/linux-64::blas-1.0-openblas \n",
      "  brotli             pkgs/main/linux-64::brotli-1.0.9-h5eee18b_7 \n",
      "  brotli-bin         pkgs/main/linux-64::brotli-bin-1.0.9-h5eee18b_7 \n",
      "  click              pkgs/main/linux-64::click-8.1.7-py311h06a4308_0 \n",
      "  coloredlogs        pkgs/main/linux-64::coloredlogs-15.0.1-py311h06a4308_1 \n",
      "  colormath          conda-forge/noarch::colormath-3.0.0-py_2 \n",
      "  contourpy          pkgs/main/linux-64::contourpy-1.2.0-py311hdb19cb5_0 \n",
      "  cycler             pkgs/main/noarch::cycler-0.11.0-pyhd3eb1b0_0 \n",
      "  fonttools          pkgs/main/noarch::fonttools-4.25.0-pyhd3eb1b0_0 \n",
      "  future             pkgs/main/linux-64::future-0.18.3-py311h06a4308_0 \n",
      "  giflib             pkgs/main/linux-64::giflib-5.2.1-h5eee18b_3 \n",
      "  humanfriendly      pkgs/main/linux-64::humanfriendly-10.0-py311h06a4308_1 \n",
      "  humanize           pkgs/main/noarch::humanize-3.10.0-pyhd3eb1b0_0 \n",
      "  importlib-metadata pkgs/main/linux-64::importlib-metadata-7.0.1-py311h06a4308_0 \n",
      "  jpeg               pkgs/main/linux-64::jpeg-9e-h5eee18b_1 \n",
      "  kiwisolver         pkgs/main/linux-64::kiwisolver-1.4.4-py311h6a678d5_0 \n",
      "  lcms2              pkgs/main/linux-64::lcms2-2.12-h3be6417_0 \n",
      "  libbrotlicommon    pkgs/main/linux-64::libbrotlicommon-1.0.9-h5eee18b_7 \n",
      "  libbrotlidec       pkgs/main/linux-64::libbrotlidec-1.0.9-h5eee18b_7 \n",
      "  libbrotlienc       pkgs/main/linux-64::libbrotlienc-1.0.9-h5eee18b_7 \n",
      "  libgfortran-ng     pkgs/main/linux-64::libgfortran-ng-11.2.0-h00389a5_1 \n",
      "  libgfortran5       pkgs/main/linux-64::libgfortran5-11.2.0-h1234567_1 \n",
      "  libopenblas        pkgs/main/linux-64::libopenblas-0.3.21-h043d6bf_0 \n",
      "  libtiff            pkgs/main/linux-64::libtiff-4.2.0-hecacb30_2 \n",
      "  libwebp            pkgs/main/linux-64::libwebp-1.3.2-h11a3e52_0 \n",
      "  libwebp-base       pkgs/main/linux-64::libwebp-base-1.3.2-h5eee18b_0 \n",
      "  lzstring           conda-forge/noarch::lzstring-1.0.4-py_1001 \n",
      "  markdown           pkgs/main/linux-64::markdown-3.4.1-py311h06a4308_0 \n",
      "  markdown-it-py     pkgs/main/linux-64::markdown-it-py-2.2.0-py311h06a4308_1 \n",
      "  matplotlib-base    pkgs/main/linux-64::matplotlib-base-3.8.0-py311ha02d727_0 \n",
      "  mdurl              pkgs/main/linux-64::mdurl-0.1.0-py311h06a4308_0 \n",
      "  multiqc            bioconda/noarch::multiqc-1.19-pyhdfd78af_0 \n",
      "  munkres            bioconda/noarch::munkres-1.0.7-py_1 \n",
      "  networkx           pkgs/main/linux-64::networkx-3.1-py311h06a4308_0 \n",
      "  numpy              pkgs/main/linux-64::numpy-1.26.3-py311h24aa872_0 \n",
      "  numpy-base         pkgs/main/linux-64::numpy-base-1.26.3-py311hbfb1bba_0 \n",
      "  openjpeg           pkgs/main/linux-64::openjpeg-2.4.0-h3ad879b_0 \n",
      "  pillow             pkgs/main/linux-64::pillow-10.0.1-py311ha6cbd5a_0 \n",
      "  pyaml-env          conda-forge/noarch::pyaml-env-1.1.0-pyhd8ed1ab_0 \n",
      "  pyparsing          pkgs/main/linux-64::pyparsing-3.0.9-py311h06a4308_0 \n",
      "  rich               pkgs/main/linux-64::rich-13.3.5-py311h06a4308_0 \n",
      "  rich-click         conda-forge/noarch::rich-click-1.7.3-pyhd8ed1ab_0 \n",
      "  simplejson         pkgs/main/linux-64::simplejson-3.17.6-py311h5eee18b_0 \n",
      "  spectra            conda-forge/noarch::spectra-0.0.11-py_1 \n",
      "  zipp               pkgs/main/linux-64::zipp-3.17.0-py311h06a4308_0 \n",
      "\n",
      "The following packages will be DOWNGRADED:\n",
      "\n",
      "  pyyaml                              6.0.1-py311h5eee18b_0 --> 5.4.1-py311h5eee18b_3 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages:\n",
      "numpy-base-1.26.3    | 8.3 MB    |                                       |   0% \n",
      "matplotlib-base-3.8. | 7.7 MB    |                                       |   0% \u001b[A\n",
      "\n",
      "libopenblas-0.3.21   | 5.4 MB    |                                       |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "networkx-3.1         | 3.3 MB    |                                       |   0% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libgfortran5-11.2.0  | 2.0 MB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "multiqc-1.19         | 1.0 MB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pillow-10.0.1        | 896 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "future-0.18.3        | 824 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "fonttools-4.25.0     | 632 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "rich-13.3.5          | 560 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libtiff-4.2.0        | 466 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libwebp-base-1.3.2   | 387 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "openjpeg-2.4.0       | 331 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "lcms2-2.12           | 312 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libbrotlienc-1.0.9   | 264 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "contourpy-1.2.0      | 263 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "click-8.1.7          | 221 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pyyaml-5.4.1         | 212 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pyparsing-3.0.9      | 204 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "markdown-3.4.1       | 164 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "humanfriendly-10.0   | 148 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "markdown-it-py-2.2.0 | 145 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "simplejson-3.17.6    | 128 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "numpy-base-1.26.3    | 8.3 MB    | #1                                    |   3% [A\u001b[A\u001b[A\u001b[A\n",
      "matplotlib-base-3.8. | 7.7 MB    | ##                                    |   5% \u001b[A\n",
      "\n",
      "\n",
      "networkx-3.1         | 3.3 MB    | ######6                               |  18% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libgfortran5-11.2.0  | 2.0 MB    | ####8                                 |  13% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "numpy-base-1.26.3    | 8.3 MB    | #####2                                |  14% \u001b[A\u001b[A\n",
      "matplotlib-base-3.8. | 7.7 MB    | ######4                               |  17% \u001b[A\n",
      "\n",
      "\n",
      "networkx-3.1         | 3.3 MB    | ##################################7   |  94% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libgfortran5-11.2.0  | 2.0 MB    | #################################5    |  91% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "libopenblas-0.3.21   | 5.4 MB    | #########8                            |  27% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pillow-10.0.1        | 896 KB    | 6                                     |   2% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "matplotlib-base-3.8. | 7.7 MB    | #######################2              |  63% \u001b[A\n",
      "\n",
      "numpy-base-1.26.3    | 8.3 MB    | #############                         |  35% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "numpy-base-1.26.3    | 8.3 MB    | ###########################8          |  75% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "multiqc-1.19         | 1.0 MB    | 5                                     |   1% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "fonttools-4.25.0     | 632 KB    | 9                                     |   3% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libtiff-4.2.0        | 466 KB    | #2                                    |   3% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "rich-13.3.5          | 560 KB    | #                                     |   3% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libwebp-base-1.3.2   | 387 KB    | #5                                    |   4% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "openjpeg-2.4.0       | 331 KB    | #7                                    |   5% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "lcms2-2.12           | 312 KB    | #9                                    |   5% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libbrotlienc-1.0.9   | 264 KB    | ##2                                   |   6% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "contourpy-1.2.0      | 263 KB    | ##2                                   |   6% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "click-8.1.7          | 221 KB    | ##6                                   |   7% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pyparsing-3.0.9      | 204 KB    | ##9                                   |   8% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "markdown-3.4.1       | 164 KB    | ###6                                  |  10% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pyyaml-5.4.1         | 212 KB    | ##7                                   |   8% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "markdown-it-py-2.2.0 | 145 KB    | ####                                  |  11% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "humanfriendly-10.0   | 148 KB    | ###9                                  |  11% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "simplejson-3.17.6    | 128 KB    | ####6                                 |  13% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libgfortran5-11.2.0  | 2.0 MB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pillow-10.0.1        | 896 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pillow-10.0.1        | 896 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "libopenblas-0.3.21   | 5.4 MB    | ##################################### | 100% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "future-0.18.3        | 824 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "future-0.18.3        | 824 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "fonttools-4.25.0     | 632 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "fonttools-4.25.0     | 632 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "multiqc-1.19         | 1.0 MB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "multiqc-1.19         | 1.0 MB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libtiff-4.2.0        | 466 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libtiff-4.2.0        | 466 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "rich-13.3.5          | 560 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "rich-13.3.5          | 560 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "networkx-3.1         | 3.3 MB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libwebp-base-1.3.2   | 387 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libwebp-base-1.3.2   | 387 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "openjpeg-2.4.0       | 331 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "openjpeg-2.4.0       | 331 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libbrotlienc-1.0.9   | 264 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libbrotlienc-1.0.9   | 264 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "lcms2-2.12           | 312 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "lcms2-2.12           | 312 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "contourpy-1.2.0      | 263 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "contourpy-1.2.0      | 263 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "click-8.1.7          | 221 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "click-8.1.7          | 221 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pyparsing-3.0.9      | 204 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pyparsing-3.0.9      | 204 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "markdown-3.4.1       | 164 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "markdown-3.4.1       | 164 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "markdown-it-py-2.2.0 | 145 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "markdown-it-py-2.2.0 | 145 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pyyaml-5.4.1         | 212 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pyyaml-5.4.1         | 212 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "matplotlib-base-3.8. | 7.7 MB    | ##################################### | 100% \u001b[A\n",
      "matplotlib-base-3.8. | 7.7 MB    | ##################################### | 100% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "humanfriendly-10.0   | 148 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "humanfriendly-10.0   | 148 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "simplejson-3.17.6    | 128 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "numpy-base-1.26.3    | 8.3 MB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \n",
      "                                                                                \u001b[A\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\u001b[A\n",
      "\n",
      "\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n"
     ]
    }
   ],
   "source": [
    "!conda install -c bioconda multiqc --yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "26784216-2462-464d-9748-a669987ad476",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/e/new_workflow\n"
     ]
    }
   ],
   "source": [
    "%cd /mnt/e/new_workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "ab76c83e-9025-4d8d-8b9b-902175d55e85",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                \n",
      " \u001b[91m///\u001b[0m \u001b]8;id=901523;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.19\u001b[0m                                                         \n",
      "                                                                                \n",
      " \u001b[33mUsage:\u001b[0m \u001b[1mmultiqc\u001b[0m [\u001b[1;36mOPTIONS\u001b[0m] [\u001b[1;36mANALYSIS\u001b[0m \u001b[1;36mDIRECTORY\u001b[0m]                                  \n",
      "                                                                                \n",
      " MultiQC aggregates results from bioinformatics analyses across many samples    \n",
      " into a single report.                                                          \n",
      " \u001b[2mIt searches a given directory for analysis logs and compiles a HTML report. \u001b[0m   \n",
      " \u001b[2mIt's a general use tool, perfect for summarising the output from numerous \u001b[0m     \n",
      " \u001b[2mbioinformatics tools.\u001b[0m                                                          \n",
      " \u001b[2mTo run, supply with one or more directory to scan for analysis results. For \u001b[0m   \n",
      " \u001b[2mexample, to run in the current working directory, use '\u001b[0m\u001b[1;2;34mmultiqc .\u001b[0m\u001b[2m'\u001b[0m              \n",
      "                                                                                \n",
      "\u001b[2m\u001b[0m\u001b[2m Main options \u001b[0m\u001b[2m\u001b[0m\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-force\u001b[0m            \u001b[1;32m-f\u001b[0m  Overwrite any existing reports                        \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-config\u001b[0m           \u001b[1;32m-c\u001b[0m  Specific config file to load, after those in MultiQC  \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        dir / home dir / working dir.                         \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-cl\u001b[0m\u001b[1;36m-config\u001b[0m            Specify MultiQC config YAML on the command line       \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        \u001b[2;33m(TEXT)                                         \u001b[0m       \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-filename\u001b[0m         \u001b[1;32m-n\u001b[0m  Report filename. Use '\u001b[3;33mstdout\u001b[0m' to print to standard    \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        out.                                                  \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-outdir\u001b[0m           \u001b[1;32m-o\u001b[0m  Create report in the specified output directory.      \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        \u001b[2;33m(TEXT)                                          \u001b[0m      \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ignore\u001b[0m           \u001b[1;32m-x\u001b[0m  Ignore analysis files \u001b[2;33m(GLOB EXPRESSION)\u001b[0m               \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ignore\u001b[0m\u001b[1;36m-samples\u001b[0m       Ignore sample names \u001b[2;33m(GLOB EXPRESSION)\u001b[0m                 \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ignore\u001b[0m\u001b[1;36m-symlinks\u001b[0m      Ignore symlinked directories and files                \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-file\u001b[0m\u001b[1;36m-list\u001b[0m        \u001b[1;32m-l\u001b[0m  Supply a file containing a list of file paths to be   \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        searched, one per row                                 \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\u001b[2m Choosing modules to run \u001b[0m\u001b[2m\u001b[0m\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-module\u001b[0m     \u001b[1;32m-m\u001b[0m  Use only this module. Can specify multiple times.           \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                  \u001b[2;33m(MODULE NAME)                                    \u001b[0m           \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-exclude\u001b[0m    \u001b[1;32m-e\u001b[0m  Do not use this module. Can specify multiple times.         \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                  \u001b[2;33m(MODULE NAME)                                      \u001b[0m         \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-tag\u001b[0m            Use only modules which tagged with this keyword \u001b[2;33m(TEXT)\u001b[0m      \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-view\u001b[0m\u001b[1;36m-tags\u001b[0m      View the available tags and which modules they load         \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\u001b[2m Sample handling \u001b[0m\u001b[2m\u001b[0m\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-dirs\u001b[0m           \u001b[1;32m-d\u001b[0m   Prepend directory to sample names                      \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-dirs\u001b[0m\u001b[1;36m-depth\u001b[0m     \u001b[1;32m-dd\u001b[0m  Prepend \u001b[3;33mn\u001b[0m directories to sample names. Negative number \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                       to take from start of path.                            \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                       \u001b[2;33m(INTEGER)                                             \u001b[0m \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-fullnames\u001b[0m      \u001b[1;32m-s\u001b[0m   Do not clean the sample names \u001b[3m(leave as full file \u001b[0m     \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                       \u001b[3mname)\u001b[0m                                                  \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-fn_as_s_name\u001b[0m        Use the log filename as the sample name                \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-replace\u001b[0m\u001b[1;36m-names\u001b[0m       TSV file to rename sample names during report          \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                       generation                                             \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                       \u001b[2;33m(PATH)                                                \u001b[0m \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\u001b[2m Report customisation \u001b[0m\u001b[2m\u001b[0m\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-title\u001b[0m            \u001b[1;32m-i\u001b[0m  Report title. Printed as page header, used for        \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        filename if not otherwise specified.                  \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-comment\u001b[0m          \u001b[1;32m-b\u001b[0m  Custom comment, will be printed at the top of the     \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        report.                                               \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-template\u001b[0m         \u001b[1;32m-t\u001b[0m  Report template to use.                               \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        \u001b[2;33m(default|default_dev|gathered|geo|sections|simple)\u001b[0m    \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-sample\u001b[0m\u001b[1;36m-names\u001b[0m         TSV file containing alternative sample names for      \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        renaming buttons in the report                        \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-sample\u001b[0m\u001b[1;36m-filters\u001b[0m       TSV file containing show/hide patterns for the report \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-custom\u001b[0m\u001b[1;36m-css-file\u001b[0m      Custom CSS file to add to the final report \u001b[2;33m(PATH)\u001b[0m     \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\u001b[2m Output files \u001b[0m\u001b[2m\u001b[0m\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-flat\u001b[0m          \u001b[1;32m-fp\u001b[0m  Use only flat plots \u001b[3m(static images)\u001b[0m                     \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-interactive\u001b[0m   \u001b[1;32m-ip\u001b[0m  Use only interactive plots \u001b[3m(in-browser Javascript)\u001b[0m      \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-export\u001b[0m        \u001b[1;32m-p\u001b[0m   Export plots as static images in addition to the report \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-data\u001b[0m\u001b[1;36m-dir\u001b[0m           Force the parsed data directory to be created.          \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-no\u001b[0m\u001b[1;36m-data-dir\u001b[0m        Prevent the parsed data directory from being created.   \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-data\u001b[0m\u001b[1;36m-format\u001b[0m   \u001b[1;32m-k\u001b[0m   Output parsed data in a different format.               \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                      \u001b[2;33m(tsv|json|yaml)                          \u001b[0m               \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-zip\u001b[0m\u001b[1;36m-data-dir\u001b[0m  \u001b[1;32m-z\u001b[0m   Compress the data directory.                            \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-no\u001b[0m\u001b[1;36m-report\u001b[0m          Do not generate a report, only export data and plots    \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-pdf\u001b[0m                Creates PDF report with the \u001b[3m'simple'\u001b[0m template. Requires \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                      \u001b]8;id=468761;https://pandoc.org/\u001b\\Pandoc\u001b]8;;\u001b\\ to be installed.                                 \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\u001b[2m MultiQC behaviour \u001b[0m\u001b[2m\u001b[0m\u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-verbose\u001b[0m           \u001b[1;32m-v\u001b[0m  Increase output verbosity. \u001b[2;33m(INTEGER RANGE)\u001b[0m           \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-quiet\u001b[0m             \u001b[1;32m-q\u001b[0m  Only show log warnings                               \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-strict\u001b[0m                Don't catch exceptions, run additional code checks   \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                         to help development.                                 \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-require\u001b[0m\u001b[1;36m-logs\u001b[0m          Require all explicitly requested modules to have log \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                         files. If not, MultiQC will exit with an error.      \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-profile\u001b[0m\u001b[1;36m-runtime\u001b[0m       Add analysis of how long MultiQC takes to run to the \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                         report                                               \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-no\u001b[0m\u001b[1;36m-megaqc-upload\u001b[0m      Don't upload generated report to MegaQC, even if     \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m                         MegaQC options are found                             \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-no\u001b[0m\u001b[1;36m-ansi\u001b[0m               Disable coloured log output                          \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-version\u001b[0m               Show the version and exit.                           \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-help\u001b[0m              \u001b[1;32m-h\u001b[0m  Show this message and exit.                          \u001b[2m\u001b[0m\n",
      "\u001b[2m\u001b[0m\n",
      "                                                                                \n",
      " See \u001b]8;id=570028;http://multiqc.info\u001b\\http://multiqc.info\u001b]8;;\u001b\\ for more details.                                      \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Help\n",
    "!multiqc -h "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e1c3c683-6cc9-4388-b43f-e31ffcb22e77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=606326;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\  \u001b[2m| v1.19\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /mnt/e/new_workflow/results\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m833/833\u001b[0m  iQC/multiqc_report_1.html\u001b[0mtkeep\u001b[0m\n",
      "\u001b[?25h\u001b[34m|    feature_counts\u001b[0m | Found 10 reports\n",
      "\u001b[34m|              star\u001b[0m | Found 9 reports and 9 gene count files\n",
      "\u001b[34m|          cutadapt\u001b[0m | Found 20 reports\n",
      "\u001b[34m|            fastqc\u001b[0m | Found 20 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Existing reports found, adding suffix to filenames. Use '--force' to overwrite.\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : results/6_multiQC/multiqc_report_2.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : results/6_multiQC/multiqc_data_2\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "# Run multiqc and output results into final folder\n",
    "!multiqc results \\\n",
    "--outdir results/6_multiQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d77609dc-e7fb-4dd7-87e0-285e5e36de41",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rna_seq",
   "language": "python",
   "name": "rna_seq"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
